{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a01d0f96",
   "metadata": {},
   "source": [
    "# 1. read metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d101cc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cyvcf2 import VCF, Writer\n",
    "import subprocess\n",
    "import os\n",
    "import random\n",
    "\n",
    "os.chdir('/mnt/qmtang/EvoFill_data/20251211_chr22/')\n",
    "\n",
    "KGP_vcf = \"/mnt/qmtang/EvoFill_data/20251204_chr22/1kGP_hg38_chr22_maf0.01_snps.vcf.gz\"\n",
    "\n",
    "VCF2DIS = \"/home/qmtang/GitHub/VCF2Dis/bin/VCF2Dis\"\n",
    "KGP_ped = \"/mnt/NAS/Omics/DNA/1kGP/meta/20130606_g1k.ped\"\n",
    "\n",
    "\n",
    "save_dir = \"./data\"\n",
    "if not os.path.exists(save_dir):\n",
    "        os.makedirs(save_dir)\n",
    "random.seed(42)\n",
    "\n",
    "pop_anno: dict[str, dict[str, str]] = {\n",
    "    \"GWD\":  {\"CHN\":\"冈比亚西部人\",        \"Region\":\"非洲/冈比亚\",            \"说明\":\"西非撒哈拉以南，班图语支\"},\n",
    "    \"CHS\":  {\"CHN\":\"中国南方汉族\",        \"Region\":\"亚洲/中国南方\",         \"说明\":\"中国汉族南方群体\"},\n",
    "    \"TSI\":  {\"CHN\":\"托斯卡纳人\",          \"Region\":\"欧洲/意大利\",           \"说明\":\"意大利中部托斯卡纳地区\"},\n",
    "    \"PUR\":  {\"CHN\":\"波多黎各人\",          \"Region\":\"美洲/波多黎各\",         \"说明\":\"拉丁裔，欧洲-非洲-美洲土著混合\"},\n",
    "    \"JPT\":  {\"CHN\":\"日本东京人\",          \"Region\":\"亚洲/日本\",             \"说明\":\"本州岛东部，大和族\"},\n",
    "    \"IBS\":  {\"CHN\":\"西班牙伊比利亚人\",    \"Region\":\"欧洲/西班牙\",           \"说明\":\"伊比利亚半岛土著\"},\n",
    "    \"YRI\":  {\"CHN\":\"尼日利亚约鲁巴人\",    \"Region\":\"非洲/尼日利亚\",         \"说明\":\"西非尼日尔-刚果语系\"},\n",
    "    \"GIH\":  {\"CHN\":\"印度古吉拉特人\",      \"Region\":\"亚洲/印度（古吉拉特）\", \"说明\":\"印度西部，印欧语系\"},\n",
    "    \"FIN\":  {\"CHN\":\"芬兰人\",              \"Region\":\"欧洲/芬兰\",             \"说明\":\"北欧乌拉尔语系，遗传瓶颈明显\"},\n",
    "    \"CEU\":  {\"CHN\":\"北欧白人（犹他）\",    \"Region\":\"欧洲/西北欧→美国\",      \"说明\":\"西北欧移民后裔（美国犹他州）\"},\n",
    "    \"ITU\":  {\"CHN\":\"印度泰卢固人\",        \"Region\":\"亚洲/印度（安得拉）\",   \"说明\":\"印度东南部，德拉维语系\"},\n",
    "    \"STU\":  {\"CHN\":\"印度泰米尔人\",        \"Region\":\"亚洲/印度（泰米尔纳德）\",\"说明\":\"印度最南端，德拉维语系\"},\n",
    "    \"KHV\":  {\"CHN\":\"越南京族人\",          \"Region\":\"亚洲/越南\",             \"说明\":\"越南北部，南亚语系→越芒语支\"},\n",
    "    \"CHB\":  {\"CHN\":\"中国北方汉族\",        \"Region\":\"亚洲/中国北方\",         \"说明\":\"中国汉族北方群体\"},\n",
    "    \"LWK\":  {\"CHN\":\"肯尼亚卢赫雅人\",      \"Region\":\"非洲/肯尼亚\",           \"说明\":\"东非班图语支\"},\n",
    "    \"ESN\":  {\"CHN\":\"尼日利亚埃桑人\",      \"Region\":\"非洲/尼日利亚\",         \"说明\":\"尼日尔-刚果语系，西非\"},\n",
    "    \"ACB\":  {\"CHN\":\"加勒比非裔\",          \"Region\":\"美洲/加勒比→非洲\",      \"说明\":\"非洲奴隶贸易后裔（加勒比地区）\"},\n",
    "    \"PJL\":  {\"CHN\":\"印度旁遮普人\",        \"Region\":\"亚洲/印度（旁遮普）\",   \"说明\":\"印度西北部，印欧语系\"},\n",
    "    \"GBR\":  {\"CHN\":\"英国英格兰人\",        \"Region\":\"欧洲/英国\",             \"说明\":\"大不列颠岛土著\"},\n",
    "    \"CLM\":  {\"CHN\":\"哥伦比亚麦德林人\",    \"Region\":\"美洲/哥伦比亚\",         \"说明\":\"拉丁裔，欧洲-美洲土著-非洲混合\"},\n",
    "    \"CDX\":  {\"CHN\":\"中国西双版纳傣族人\",  \"Region\":\"亚洲/中国云南\",         \"说明\":\"傣泰民族，侗台语系\"},\n",
    "    \"MSL\":  {\"CHN\":\"塞拉利昂门德人\",      \"Region\":\"非洲/塞拉利昂\",         \"说明\":\"西非曼德语支\"},\n",
    "    \"PEL\":  {\"CHN\":\"秘鲁利马人\",          \"Region\":\"美洲/秘鲁\",             \"说明\":\"拉丁裔，克丘亚-欧洲混合\"},\n",
    "    \"BEB\":  {\"CHN\":\"孟加拉国人\",          \"Region\":\"亚洲/孟加拉国\",         \"说明\":\"孟加拉湾沿岸，印欧语系\"},\n",
    "    \"MXL\":  {\"CHN\":\"墨西哥洛杉矶人\",      \"Region\":\"美洲/墨西哥→美国\",      \"说明\":\"拉丁裔，美洲土著-欧洲混合\"},\n",
    "    \"ASW\":  {\"CHN\":\"非裔美国人\",          \"Region\":\"美洲/美国\",             \"说明\":\"非洲奴隶贸易后裔（美国南部）\"},\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa36cf21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. 读取 ped 文件\n",
    "ped_df = pd.read_csv(KGP_ped, sep='\\t',)\n",
    "\n",
    "# 3. 读取 VCF 样本列表\n",
    "vcf = VCF(KGP_vcf)\n",
    "vcf_samples = set(vcf.samples)\n",
    "vcf.close()\n",
    "\n",
    "# 4. 筛选 ped 中也在 VCF 中出现的样本\n",
    "ped_df = ped_df[ped_df['Individual ID'].isin(vcf_samples)]\n",
    "\n",
    "# 5. 统计各个 population 的样本数量\n",
    "pop_counts = ped_df['Population'].value_counts()\n",
    "(pop_counts\n",
    " .rename_axis(\"Population\")\n",
    " .to_frame(\"Samples\")\n",
    " .assign(CHN=lambda x: x.index.map(lambda p: pop_anno[p][\"CHN\"]),\n",
    "         Region=lambda x: x.index.map(lambda p: pop_anno[p][\"Region\"]))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f952ab1",
   "metadata": {},
   "source": [
    "# 2. splitting samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e7870dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "minor_pop_list = ['CDX'] # 选择需要分割的人群"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aad482bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "major_samples = ped_df[~ped_df['Population'].isin(minor_pop_list)]['Individual ID'].tolist()\n",
    "minor_samples = ped_df[ ped_df['Population'].isin(minor_pop_list)]['Individual ID'].tolist()\n",
    "\n",
    "random.shuffle(major_samples)\n",
    "major_val   = major_samples[:100]          # 100 个验证集\n",
    "major_train = major_samples[100:]          # 剩下的训练集\n",
    "\n",
    "major_val_list   = os.path.join(save_dir, \"major_pops_val.samples.txt\")\n",
    "major_train_list = os.path.join(save_dir, \"major_pops_train.samples.txt\")\n",
    "with open(major_val_list, 'w') as f:\n",
    "    for s in major_val:\n",
    "        f.write(s + '\\n')\n",
    "with open(major_train_list, 'w') as f:\n",
    "    for s in major_train:\n",
    "        f.write(s + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6f9e75b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sample_vcf(in_vcf, sample_list, out_vcf, threads=8):\n",
    "    \"\"\"用 bcftools 按样本列表切子集\"\"\"\n",
    "    subprocess.run([\n",
    "        \"bcftools\", \"view\", \"-S\", sample_list,\n",
    "        \"-Oz\", \"-o\", out_vcf, in_vcf, \"--threads\", str(threads)\n",
    "    ], check=True)\n",
    "    subprocess.run([\"tabix\", \"-p\", \"vcf\", out_vcf], check=True)\n",
    "\n",
    "major_val_vcf   = os.path.join(save_dir, \"major_pops_val.vcf.gz\")\n",
    "major_train_vcf = os.path.join(save_dir, \"major_pops_train.vcf.gz\")\n",
    "sample_vcf(KGP_vcf, major_val_list,   major_val_vcf)\n",
    "sample_vcf(KGP_vcf, major_train_list, major_train_vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fde8518b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mask_sites(in_vcf, out_vcf, missing_rate=0.9, seed=42):\n",
    "    \"\"\"\n",
    "    按位点（variant）维度随机 mask missing_rate 比例，\n",
    "    即每个位点以 missing_rate 概率被整列设成 ./.\n",
    "    \"\"\"\n",
    "    random.seed(seed)\n",
    "    tmp = out_vcf.replace(\".gz\", \"\")\n",
    "    vcf = VCF(in_vcf)\n",
    "    w   = Writer(tmp, vcf)\n",
    "    for var in vcf:\n",
    "        if random.random() < missing_rate:\n",
    "            # 整列置空\n",
    "            gts = [[-1, -1, True] for _ in range(len(vcf.samples))]\n",
    "        else:\n",
    "            gts = var.genotypes\n",
    "        var.genotypes = gts\n",
    "        w.write_record(var)\n",
    "    w.close(); vcf.close()\n",
    "    subprocess.run([\"bgzip\", \"-f\", tmp], check=True)\n",
    "    subprocess.run([\"tabix\", \"-p\", \"vcf\", out_vcf], check=True)\n",
    "\n",
    "major_val_mask_vcf = os.path.join(save_dir, \"major_pops_val.mask90p.vcf.gz\")\n",
    "mask_sites(major_val_vcf, major_val_mask_vcf, missing_rate=0.9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dcbb209d",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.shuffle(minor_samples)\n",
    "n_minor     = len(minor_samples)\n",
    "fewshot_n   = max(1, int(n_minor * 0.1))\n",
    "fewshot_sam = minor_samples[:fewshot_n]\n",
    "remain_sam  = minor_samples[fewshot_n:]\n",
    "\n",
    "# 样本列表文件路径\n",
    "fewshot_list = os.path.join(save_dir, \"minor_fewshot.samples.txt\")\n",
    "remain_list  = os.path.join(save_dir,  \"minor_remain.samples.txt\")\n",
    "minor_all_list = os.path.join(save_dir, \"minor_pops_all.samples.txt\")\n",
    "\n",
    "# 写 txt\n",
    "def write_samples(sam_list, file_path):\n",
    "    with open(file_path, 'w') as f:\n",
    "        for s in sam_list:\n",
    "            f.write(s + '\\n')\n",
    "\n",
    "write_samples(fewshot_sam, fewshot_list)\n",
    "write_samples(remain_sam,  remain_list)\n",
    "write_samples(minor_samples, minor_all_list)\n",
    "\n",
    "# 4.1 全部 minor 样本\n",
    "minor_all_vcf = os.path.join(save_dir, \"minor_pops_all.vcf.gz\")\n",
    "sample_vcf(KGP_vcf, minor_all_list, minor_all_vcf)\n",
    "\n",
    "# 4.2 全部 minor 90% mask\n",
    "minor_all_mask_vcf = os.path.join(save_dir, \"minor_pops_all.mask90p.vcf.gz\")\n",
    "mask_sites(minor_all_vcf, minor_all_mask_vcf, missing_rate=0.9)\n",
    "\n",
    "# 4.3 few-shot 子集（完整位点）\n",
    "fewshot_vcf = os.path.join(save_dir,\"minor_pops_fewshot.vcf.gz\")\n",
    "sample_vcf(KGP_vcf, fewshot_list, fewshot_vcf)\n",
    "\n",
    "# 4.4 remain 子集再 90% mask\n",
    "remain_mask_vcf = os.path.join(save_dir,  \"minor_pops_90pct.mask90p.vcf.gz\")\n",
    "mask_sites(minor_all_vcf, remain_mask_vcf, missing_rate=0.9)   # 位点全集已存在，直接用"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f73e1c4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Sample Number to construct p-distance matrix is [ 3009 ]\n",
      "Start To Cal ...\n",
      "Start To Create P_distance ...\n",
      "P_distance is created done ...\n",
      "\twarning: can't find the [Rscript ] in your $PATH ; no tree and pdf Out\n",
      "\tRscript\tvistreecode.r\t./data/evo_mat_major_pops_train.tsv\n",
      "\tor see more at\n",
      "\t\t[ https://github.com/hewm2008/VCF2Dis/blob/main/OtherWay2ConstructTree.md ]\t./data/evo_mat_major_pops_train.tsv\n",
      "Total Sample Number to construct p-distance matrix is [ 9 ]\n",
      "Start To Cal ...\n",
      "Start To Create P_distance ...\n",
      "P_distance is created done ...\n",
      "\twarning: can't find the [Rscript ] in your $PATH ; no tree and pdf Out\n",
      "\tRscript\tvistreecode.r\t./data/evo_mat_minor_fewshot.tsv\n",
      "\tor see more at\n",
      "\t\t[ https://github.com/hewm2008/VCF2Dis/blob/main/OtherWay2ConstructTree.md ]\t./data/evo_mat_minor_fewshot.tsv\n"
     ]
    }
   ],
   "source": [
    "def calc_evo_matrix(vcf_path, out_tsv, VCF2DIS_BIN):\n",
    "    \"\"\"计算并对称化距离矩阵\"\"\"\n",
    "    subprocess.run([VCF2DIS_BIN, \"-InPut\", vcf_path, \"-OutPut\", out_tsv], check=True)\n",
    "    dist = pd.read_csv(out_tsv, sep='\\t', index_col=0, skiprows=1, header=None)\n",
    "    idx  = [i.strip() for i in dist.index]\n",
    "    dist.columns, dist.index = idx, idx\n",
    "    dist.to_csv(out_tsv, sep='\\t', header=True, index=True)\n",
    "\n",
    "calc_evo_matrix(major_train_vcf,\n",
    "                os.path.join(save_dir, \"evo_mat_major_pops_train.tsv\"),\n",
    "                VCF2DIS)\n",
    "\n",
    "calc_evo_matrix(fewshot_vcf,\n",
    "                os.path.join(save_dir, \"evo_mat_minor_fewshot.tsv\"),\n",
    "                VCF2DIS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4592ba8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Sample Number to construct p-distance matrix is [ 8953 ]\n",
      "Start To Cal ...\n",
      "Start To Create P_distance ...\n",
      "P_distance is created done ...\n",
      "\twarning: can't find the [Rscript ] in your $PATH ; no tree and pdf Out\n",
      "\tRscript\tvistreecode.r\t/mnt/qmtang/EvoFill_data/20251211_chr22/augment/evo_mat_major_pops_train.tsv\n",
      "\tor see more at\n",
      "\t\t[ https://github.com/hewm2008/VCF2Dis/blob/main/OtherWay2ConstructTree.md ]\t/mnt/qmtang/EvoFill_data/20251211_chr22/augment/evo_mat_major_pops_train.tsv\n"
     ]
    }
   ],
   "source": [
    "def calc_evo_matrix(vcf_path, out_tsv, VCF2DIS_BIN):\n",
    "    \"\"\"计算并对称化距离矩阵\"\"\"\n",
    "    subprocess.run([VCF2DIS_BIN, \"-InPut\", vcf_path, \"-OutPut\", out_tsv], check=True)\n",
    "    dist = pd.read_csv(out_tsv, sep='\\t', index_col=0, skiprows=1, header=None)\n",
    "    idx  = [i.strip() for i in dist.index]\n",
    "    dist.columns, dist.index = idx, idx\n",
    "    dist.to_csv(out_tsv, sep='\\t', header=True, index=True)\n",
    "\n",
    "calc_evo_matrix(\"/mnt/qmtang/EvoFill_data/20251211_chr22/augment/hg38_chr22_aDNA_augment_pop.vcf.gz\",\n",
    "                \"/mnt/qmtang/EvoFill_data/20251211_chr22/augment/evo_mat_aDNA.tsv\",\n",
    "                VCF2DIS)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
