{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aa6bc4ec",
   "metadata": {},
   "source": [
    "【TODO】\n",
    "1. chunks 改单chunk 【OK】\n",
    "2. 位点相对坐标改绝对坐标\n",
    "3. 演化距离D嵌入"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d4bb00d",
   "metadata": {},
   "source": [
    "## Dependency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "757312f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os; os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"1\" # 设置用GPU1\n",
    "import gzip\n",
    "import json\n",
    "import logging\n",
    "import shutil\n",
    "from typing import Union\n",
    "from argparse import Namespace\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import datatable as dt\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from torch.autograd import grad\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "\n",
    "from tqdm import tqdm\n",
    "from joblib import Parallel, delayed\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from mamba_ssm import Mamba2\n",
    "from mamba_ssm.modules.mamba2_simple import Mamba2Simple as Mamba2Block # 原Mamba2Block\n",
    "from torch_optimizer import Lamb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "975df815",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e5eeded0",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPPORTED_FILE_FORMATS = {\"vcf\", \"csv\", \"tsv\"}\n",
    "class DataReader:\n",
    "    def __init__(self):\n",
    "        self.target_is_gonna_be_phased = None\n",
    "        self.target_set = None\n",
    "        self.target_sample_value_index = 2\n",
    "        self.ref_sample_value_index = 2\n",
    "        self.target_file_extension = None\n",
    "        self.allele_count = 2\n",
    "        self.genotype_vals = None\n",
    "        self.ref_is_phased = None\n",
    "        self.reference_panel = None\n",
    "        self.VARIANT_COUNT = 0\n",
    "        self.is_phased = False\n",
    "        self.MISSING_VALUE = None\n",
    "        self.ref_is_hap = False\n",
    "        self.target_is_hap = False\n",
    "        self.ref_n_header_lines = []\n",
    "        self.ref_n_data_header = \"\"\n",
    "        self.target_n_header_lines = []\n",
    "        self.target_n_data_header = \"\"\n",
    "        self.ref_separator = None\n",
    "        self.map_values_1_vec = np.vectorize(self.__map_hap_2_ind_parent_1)\n",
    "        self.map_values_2_vec = np.vectorize(self.__map_hap_2_ind_parent_2)\n",
    "        self.map_haps_to_vec = np.vectorize(self.__map_haps_2_ind)\n",
    "        self.delimiter_dictionary = {\"vcf\": \"\\t\", \"csv\": \",\", \"tsv\": \"\\t\", \"infer\": \"\\t\"}\n",
    "        self.ref_file_extension = \"vcf\"\n",
    "        self.test_file_extension = \"vcf\"\n",
    "        self.target_is_phased = True\n",
    "\n",
    "    def __read_csv(self, file_path, is_vcf=False, is_reference=False, separator=\"\\t\", first_column_is_index=True,\n",
    "                   comments=\"##\") -> pd.DataFrame:\n",
    "        \"\"\"Read CSV/VCF files\"\"\"\n",
    "        print(\"Reading the file...\")\n",
    "        data_header = None\n",
    "        path_sep = \"/\" if \"/\" in file_path else os.path.sep\n",
    "        line_counter = 0\n",
    "        root, ext = os.path.splitext(file_path)\n",
    "        with gzip.open(file_path, 'rt') if ext == '.gz' else open(file_path, 'rt') as f_in:\n",
    "            while True:\n",
    "                line = f_in.readline()\n",
    "                if line.startswith(comments):\n",
    "                    line_counter += 1\n",
    "                    if is_reference:\n",
    "                        self.ref_n_header_lines.append(line)\n",
    "                    else:\n",
    "                        self.target_n_header_lines.append(line)\n",
    "                else:\n",
    "                    data_header = line\n",
    "                    break\n",
    "        if data_header is None:\n",
    "            raise IOError(\"The file only contains comments!\")\n",
    "        df = dt.fread(file=file_path, sep=separator, header=True, skip_to_line=line_counter + 1)\n",
    "        df = df.to_pandas()\n",
    "        if first_column_is_index:\n",
    "            df.set_index(df.columns[0], inplace=True)\n",
    "        return df\n",
    "\n",
    "    def __find_file_extension(self, file_path, file_format, delimiter):\n",
    "        separator = \"\\t\"\n",
    "        found_file_format = None\n",
    "        if file_format not in [\"infer\"] + list(SUPPORTED_FILE_FORMATS):\n",
    "            raise ValueError(\"File extension must be one of {'vcf', 'csv', 'tsv', 'infer'}.\")\n",
    "        if file_format == 'infer':\n",
    "            file_name_tokenized = file_path.split(\".\")\n",
    "            for possible_extension in file_name_tokenized[::-1]:\n",
    "                if possible_extension in SUPPORTED_FILE_FORMATS:\n",
    "                    found_file_format = possible_extension\n",
    "                    separator = self.delimiter_dictionary[possible_extension] if delimiter is None else delimiter\n",
    "                    break\n",
    "            if found_file_format is None:\n",
    "                logging.warning(\"Could not infer the file type. Using tsv as the last resort.\")\n",
    "                found_file_format = \"tsv\"\n",
    "        else:\n",
    "            found_file_format = file_format\n",
    "            separator = self.delimiter_dictionary[file_format] if delimiter is None else delimiter\n",
    "        return found_file_format, separator\n",
    "\n",
    "    def assign_training_set(self, file_path: str, target_is_gonna_be_phased_or_haps: bool,\n",
    "                            variants_as_columns: bool = False, delimiter=None, file_format=\"infer\",\n",
    "                            first_column_is_index=True, comments=\"##\") -> None:\n",
    "        self.target_is_gonna_be_phased = target_is_gonna_be_phased_or_haps\n",
    "        self.ref_file_extension, self.ref_separator = self.__find_file_extension(file_path, file_format, delimiter)\n",
    "\n",
    "        self.reference_panel = self.__read_csv(file_path, is_reference=True, is_vcf=False, separator=self.ref_separator,\n",
    "                                               first_column_is_index=first_column_is_index,\n",
    "                                               comments=comments) if self.ref_file_extension != 'vcf' else self.__read_csv(\n",
    "            file_path, is_reference=True, is_vcf=True, separator='\\t', first_column_is_index=False, comments=\"##\")\n",
    "\n",
    "        if self.ref_file_extension != \"vcf\":\n",
    "            if variants_as_columns:\n",
    "                self.reference_panel = self.reference_panel.transpose()\n",
    "            self.reference_panel.reset_index(drop=False, inplace=True)\n",
    "            self.reference_panel.rename(columns={self.reference_panel.columns[0]: \"ID\"}, inplace=True)\n",
    "        else:\n",
    "            self.ref_sample_value_index += 8\n",
    "\n",
    "        self.ref_is_hap = not (\"|\" in self.reference_panel.iloc[0, self.ref_sample_value_index - 1] or \"/\" in\n",
    "                               self.reference_panel.iloc[0, self.ref_sample_value_index - 1])\n",
    "        self.ref_is_phased = \"|\" in self.reference_panel.iloc[0, self.ref_sample_value_index - 1]\n",
    "\n",
    "        if self.ref_is_hap and not target_is_gonna_be_phased_or_haps:\n",
    "            raise ValueError(\n",
    "                \"Reference contains haploids while target will be unphased diploids. Model cannot predict target.\")\n",
    "\n",
    "        if not (self.ref_is_phased or self.ref_is_hap) and target_is_gonna_be_phased_or_haps:\n",
    "            raise ValueError(\n",
    "                \"Reference contains unphased diploids while target will be phased/haploid. Model cannot predict target.\")\n",
    "\n",
    "        self.VARIANT_COUNT = self.reference_panel.shape[0]\n",
    "        print(\n",
    "            f\"{self.reference_panel.shape[1] - (self.ref_sample_value_index - 1)} {'haploid' if self.ref_is_hap else 'diploid'} samples with {self.VARIANT_COUNT} variants found!\")\n",
    "\n",
    "        self.is_phased = target_is_gonna_be_phased_or_haps and (self.ref_is_phased or self.ref_is_hap)\n",
    "\n",
    "        original_allele_sep = \"|\" if self.ref_is_phased or self.ref_is_hap else \"/\"\n",
    "        final_allele_sep = \"|\" if self.is_phased else \"/\"\n",
    "\n",
    "        def get_diploid_alleles(genotype_vals):\n",
    "            allele_set = set()\n",
    "            for genotype_val in genotype_vals:\n",
    "                if genotype_val not in [\".\", \".|.\", \"./.\"]:\n",
    "                    if final_allele_sep in genotype_val:\n",
    "                        v1, v2 = genotype_val.split(final_allele_sep)\n",
    "                        allele_set.update([v1, v2])\n",
    "                    else:\n",
    "                        allele_set.add(genotype_val)  # For haploids\n",
    "            return np.array(list(allele_set))\n",
    "\n",
    "        genotype_vals = pd.unique(self.reference_panel.iloc[:, self.ref_sample_value_index - 1:].values.ravel('K'))\n",
    "        print(f\"DEBUG: Unique genotypes in dataset: {genotype_vals[:10]}...\")  # Show first 10\n",
    "\n",
    "        if self.ref_is_phased and not target_is_gonna_be_phased_or_haps:\n",
    "            phased_to_unphased_dict = {}\n",
    "            for i in range(genotype_vals.shape[0]):\n",
    "                key = genotype_vals[i]\n",
    "                if \"|\" in key and key not in [\".\", \".|.\"]:\n",
    "                    v1, v2 = [int(s) for s in genotype_vals[i].split(original_allele_sep)]\n",
    "                    genotype_vals[i] = f\"{min(v1, v2)}/{max(v1, v2)}\"\n",
    "                    phased_to_unphased_dict[key] = genotype_vals[i]\n",
    "            if phased_to_unphased_dict:\n",
    "                self.reference_panel.iloc[:, self.ref_sample_value_index - 1:].replace(phased_to_unphased_dict,\n",
    "                                                                                       inplace=True)\n",
    "\n",
    "        self.genotype_vals = np.unique(genotype_vals)\n",
    "        self.alleles = get_diploid_alleles(self.genotype_vals) if not self.ref_is_hap else self.genotype_vals\n",
    "        self.allele_count = len(self.alleles)\n",
    "        self.MISSING_VALUE = self.allele_count if self.is_phased else len(self.genotype_vals)\n",
    "\n",
    "        print(f\"DEBUG: self.genotype_vals: {self.genotype_vals}\")\n",
    "        print(f\"DEBUG: self.alleles: {self.alleles}\")\n",
    "        print(f\"DEBUG: is_phased: {self.is_phased}\")\n",
    "\n",
    "        if self.is_phased:\n",
    "            self.hap_map = {str(v): i for i, v in enumerate(list(sorted(self.alleles)))}\n",
    "            self.hap_map.update({\".\": self.MISSING_VALUE})\n",
    "            self.r_hap_map = {i: k for k, i in self.hap_map.items()}\n",
    "            self.map_preds_2_allele = np.vectorize(lambda x: self.r_hap_map[x])\n",
    "            print(f\"DEBUG: hap_map: {self.hap_map}\")\n",
    "        else:\n",
    "            unphased_missing_genotype = \"./.\"\n",
    "            self.replacement_dict = {g: i for i, g in enumerate(list(sorted(self.genotype_vals)))}\n",
    "            self.replacement_dict[unphased_missing_genotype] = self.MISSING_VALUE\n",
    "            self.reverse_replacement_dict = {v: k for k, v in self.replacement_dict.items()}\n",
    "            print(f\"DEBUG: replacement_dict: {self.replacement_dict}\")\n",
    "\n",
    "        self.SEQ_DEPTH = self.allele_count + 1 if self.is_phased else len(self.genotype_vals) + 1\n",
    "        print(f\"DEBUG: self.SEQ_DEPTH: {self.SEQ_DEPTH}\")\n",
    "\n",
    "    def assign_test_set(self, file_path, variants_as_columns=False, delimiter=None,\n",
    "                        file_format=\"infer\", first_column_is_index=True, comments=\"##\") -> None:\n",
    "        \"\"\"Assign test set for imputation\"\"\"\n",
    "        if self.reference_panel is None:\n",
    "            raise RuntimeError(\"First you need to use 'DataReader.assign_training_set(...) to assign a training set.'\")\n",
    "\n",
    "        self.target_file_extension, separator = self.__find_file_extension(file_path, file_format, delimiter)\n",
    "\n",
    "        test_df = self.__read_csv(file_path, is_reference=False, is_vcf=False, separator=separator,\n",
    "                                  first_column_is_index=first_column_is_index,\n",
    "                                  comments=comments) if self.target_file_extension != 'vcf' else self.__read_csv(\n",
    "            file_path, is_reference=False, is_vcf=True, separator='\\t', first_column_is_index=False, comments=\"##\")\n",
    "\n",
    "        if self.target_file_extension != \"vcf\":\n",
    "            if variants_as_columns:\n",
    "                test_df = test_df.transpose()\n",
    "            test_df.reset_index(drop=False, inplace=True)\n",
    "            test_df.rename(columns={test_df.columns[0]: \"ID\"}, inplace=True)\n",
    "        else:\n",
    "            self.target_sample_value_index += 8\n",
    "\n",
    "        self.target_is_hap = not (\"|\" in test_df.iloc[0, self.target_sample_value_index - 1] or \"/\" in\n",
    "                                  test_df.iloc[0, self.target_sample_value_index - 1])\n",
    "        is_phased = \"|\" in test_df.iloc[0, self.target_sample_value_index - 1]\n",
    "        test_var_count = test_df.shape[0]\n",
    "        print(f\"{test_var_count} {'haplotype' if self.target_is_hap else 'diplotype'} variants found!\")\n",
    "\n",
    "        # Validate compatibility\n",
    "        if (self.target_is_hap or is_phased) and not (self.ref_is_phased or self.ref_is_hap):\n",
    "            raise RuntimeError(\"The training set contains unphased data. The target must be unphased as well.\")\n",
    "        if self.ref_is_hap and not (self.target_is_hap or is_phased):\n",
    "            raise RuntimeError(\"The training set contains haploids. Target set should be phased or haploids.\")\n",
    "\n",
    "        # Merge with reference panel to align variants\n",
    "        self.target_set = test_df.merge(right=self.reference_panel[[\"ID\"]], on='ID', how='right')\n",
    "        if self.target_file_extension == \"vcf\" == self.ref_file_extension:\n",
    "            self.target_set[self.reference_panel.columns[:9]] = self.reference_panel[self.reference_panel.columns[:9]]\n",
    "\n",
    "        self.target_set = self.target_set.astype('str')\n",
    "        missing_value = \".\" if self.target_is_hap else \".|.\" if self.is_phased else \"./.\"\n",
    "        self.target_set.fillna(missing_value, inplace=True)\n",
    "        self.target_set.replace(\"nan\", missing_value, inplace=True)\n",
    "        print(\"Target set assignment done!\")\n",
    "\n",
    "    def __map_hap_2_ind_parent_1(self, x) -> int:\n",
    "        return self.hap_map[x.split('|')[0]]\n",
    "\n",
    "    def __map_hap_2_ind_parent_2(self, x) -> int:\n",
    "        return self.hap_map[x.split('|')[1]]\n",
    "\n",
    "    def __map_haps_2_ind(self, x) -> int:\n",
    "        return self.hap_map[x]\n",
    "\n",
    "    def get_ref_set(self, starting_var_index=0, ending_var_index=0) -> np.ndarray:\n",
    "        if 0 <= starting_var_index < ending_var_index:\n",
    "            data = self.reference_panel.iloc[starting_var_index:ending_var_index, self.ref_sample_value_index - 1:]\n",
    "        else:\n",
    "            data = self.reference_panel.iloc[:, self.ref_sample_value_index - 1:]\n",
    "\n",
    "        if self.is_phased:\n",
    "            is_haps = \"|\" not in data.iloc[0, 0]\n",
    "            if not is_haps:\n",
    "                # diploids to hap vecs\n",
    "                _x = np.empty((data.shape[1] * 2, data.shape[0]), dtype=np.int32)\n",
    "                _x[0::2] = self.map_values_1_vec(data.values.T)\n",
    "                _x[1::2] = self.map_values_2_vec(data.values.T)\n",
    "                return _x\n",
    "            else:\n",
    "                return self.map_haps_to_vec(data.values.T)\n",
    "        else:\n",
    "            return data.replace(self.replacement_dict).values.T.astype(np.int32)\n",
    "\n",
    "    def get_target_set(self, starting_var_index=0, ending_var_index=0) -> np.ndarray:\n",
    "        \"\"\"Get target data for imputation\"\"\"\n",
    "        if 0 <= starting_var_index < ending_var_index:\n",
    "            data = self.target_set.iloc[starting_var_index:ending_var_index, self.target_sample_value_index - 1:]\n",
    "        else:\n",
    "            data = self.target_set.iloc[:, self.target_sample_value_index - 1:]\n",
    "\n",
    "        if self.is_phased:\n",
    "            is_haps = \"|\" not in data.iloc[0, 0]\n",
    "            if not is_haps:\n",
    "                # diploids to hap vecs\n",
    "                _x = np.empty((data.shape[1] * 2, data.shape[0]), dtype=np.int32)\n",
    "                _x[0::2] = self.map_values_1_vec(data.values.T)\n",
    "                _x[1::2] = self.map_values_2_vec(data.values.T)\n",
    "                return _x\n",
    "            else:\n",
    "                return self.map_haps_to_vec(data.values.T)\n",
    "        else:\n",
    "            return data.replace(self.replacement_dict).values.T.astype(np.int32)\n",
    "\n",
    "    def __convert_unphased_probs_to_genotypes(self, allele_probs) -> np.ndarray:\n",
    "        \"\"\"Convert unphased probabilities to genotypes\"\"\"\n",
    "        n_samples, n_variants, n_alleles = allele_probs.shape\n",
    "        genotypes = np.zeros((n_samples, n_variants), dtype=object)\n",
    "\n",
    "        for i in tqdm(range(n_samples)):\n",
    "            for j in range(n_variants):\n",
    "                unphased_probs = allele_probs[i, j]\n",
    "                variant_genotypes = np.vectorize(self.reverse_replacement_dict.get)(\n",
    "                    np.argmax(unphased_probs, axis=-1)).flatten()\n",
    "                genotypes[i, j] = variant_genotypes\n",
    "        return genotypes\n",
    "\n",
    "    def __convert_hap_probs_to_diploid_genotypes(self, allele_probs) -> np.ndarray:\n",
    "        \"\"\"Convert haplotype probabilities to diploid genotypes\"\"\"\n",
    "        n_haploids, n_variants, n_alleles = allele_probs.shape\n",
    "\n",
    "        if n_haploids % 2 != 0:\n",
    "            raise ValueError(\"Number of haploids should be even.\")\n",
    "\n",
    "        n_samples = n_haploids // 2\n",
    "        genotypes = np.empty((n_samples, n_variants), dtype=object)\n",
    "        haploids_as_diploids = allele_probs.reshape((n_samples, 2, n_variants, -1))\n",
    "        variant_genotypes = self.map_preds_2_allele(np.argmax(haploids_as_diploids, axis=-1))\n",
    "\n",
    "        def process_variant_in_sample(haps_for_sample_at_variant, variant_genotypes_for_sample_at_variant):\n",
    "            if n_alleles > 2:\n",
    "                return '|'.join(variant_genotypes_for_sample_at_variant)\n",
    "            else:\n",
    "                # Output GP (genotype probabilities)\n",
    "                phased_probs = np.outer(haps_for_sample_at_variant[0], haps_for_sample_at_variant[1]).flatten()\n",
    "                unphased_probs = np.array([phased_probs[0], phased_probs[1] + phased_probs[2], phased_probs[-1]])\n",
    "                unphased_probs_str = \",\".join([f\"{v:.6f}\" for v in unphased_probs])\n",
    "                alt_dosage = np.dot(unphased_probs, [0, 1, 2])\n",
    "                return '|'.join(variant_genotypes_for_sample_at_variant) + f\":{unphased_probs_str}:{alt_dosage:.3f}\"\n",
    "\n",
    "        def process_sample(i):\n",
    "            return np.array([\n",
    "                process_variant_in_sample(haploids_as_diploids[i, :, j, :], variant_genotypes[i, :, j])\n",
    "                for j in range(n_variants)\n",
    "            ])\n",
    "\n",
    "        # Parallel processing\n",
    "        genotypes = Parallel(n_jobs=-1)(delayed(process_sample)(i) for i in tqdm(range(n_samples)))\n",
    "        return np.array(genotypes)\n",
    "\n",
    "    def __convert_hap_probs_to_hap_genotypes(self, allele_probs) -> np.ndarray:\n",
    "        \"\"\"Convert hap probabilities to hap genotypes\"\"\"\n",
    "        return np.argmax(allele_probs, axis=1).astype(str)\n",
    "\n",
    "    def __get_headers_for_output(self, contain_probs, chr=22):\n",
    "        \"\"\"Get VCF headers for output file\"\"\"\n",
    "        headers = [\n",
    "            \"##fileformat=VCFv4.2\",\n",
    "            '''##source=BiMamba v1.0.0''',\n",
    "            '''##INFO=<ID=AF,Number=A,Type=Float,Description=\"Estimated Alternate Allele Frequency\">''',\n",
    "            '''##INFO=<ID=MAF,Number=1,Type=Float,Description=\"Estimated Minor Allele Frequency\">''',\n",
    "            '''##INFO=<ID=AVG_CS,Number=1,Type=Float,Description=\"Average Call Score\">''',\n",
    "            '''##INFO=<ID=IMPUTED,Number=0,Type=Flag,Description=\"Marker was imputed\">''',\n",
    "            '''##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">''',\n",
    "        ]\n",
    "        probs_headers = [\n",
    "            '''##FORMAT=<ID=DS,Number=A,Type=Float,Description=\"Estimated Alternate Allele Dosage : [P(0/1)+2*P(1/1)]\">''',\n",
    "            '''##FORMAT=<ID=GP,Number=G,Type=Float,Description=\"Estimated Posterior Probabilities for Genotypes 0/0, 0/1 and 1/1\">'''\n",
    "        ]\n",
    "        if contain_probs:\n",
    "            headers.extend(probs_headers)\n",
    "        return headers\n",
    "\n",
    "    def __convert_genotypes_to_vcf(self, genotypes, pred_format=\"GT:GP:DS\"):\n",
    "        \"\"\"Convert genotypes to VCF format\"\"\"\n",
    "        new_vcf = self.target_set.copy()\n",
    "        new_vcf[new_vcf.columns[self.target_sample_value_index - 1:]] = genotypes\n",
    "        new_vcf[\"FORMAT\"] = pred_format\n",
    "        new_vcf[\"QUAL\"] = \".\"\n",
    "        new_vcf[\"FILTER\"] = \".\"\n",
    "        new_vcf[\"INFO\"] = \"IMPUTED\"\n",
    "        return new_vcf\n",
    "\n",
    "    def preds_to_genotypes(self, predictions: Union[str, np.ndarray]) -> pd.DataFrame:\n",
    "        \"\"\"Convert predictions to genotypes\"\"\"\n",
    "        if isinstance(predictions, str):\n",
    "            preds = np.load(predictions)\n",
    "        else:\n",
    "            preds = predictions\n",
    "\n",
    "        target_df = self.target_set.copy()\n",
    "        if not self.is_phased:\n",
    "            target_df[\n",
    "                target_df.columns[self.target_sample_value_index - 1:]] = self.__convert_unphased_probs_to_genotypes(\n",
    "                preds).T\n",
    "        elif self.target_is_hap:\n",
    "            target_df[\n",
    "                target_df.columns[self.target_sample_value_index - 1:]] = self.__convert_hap_probs_to_hap_genotypes(\n",
    "                preds).T\n",
    "        else:\n",
    "            pred_format = \"GT:GP:DS\" if preds.shape[-1] == 2 else \"GT\"\n",
    "            target_df = self.__convert_genotypes_to_vcf(self.__convert_hap_probs_to_diploid_genotypes(preds).T,\n",
    "                                                        pred_format)\n",
    "        return target_df\n",
    "\n",
    "    def write_ligated_results_to_file(self, df: pd.DataFrame, file_name: str, compress=True) -> str:\n",
    "        \"\"\"Write results to file\"\"\"\n",
    "        to_write_format = self.ref_file_extension\n",
    "        file_path = f\"{file_name}.{to_write_format}.gz\" if compress else f\"{file_name}.{to_write_format}\"\n",
    "\n",
    "        with gzip.open(file_path, 'wt') if compress else open(file_path, 'wt') as f_out:\n",
    "            # Write headers\n",
    "            if self.ref_file_extension == \"vcf\":\n",
    "                f_out.write(\n",
    "                    \"\\n\".join(self.__get_headers_for_output(contain_probs=\"GP\" in df[\"FORMAT\"].values[0])) + \"\\n\")\n",
    "            else:\n",
    "                f_out.write(\"\\n\".join(self.ref_n_header_lines))\n",
    "\n",
    "        # Append data\n",
    "        df.to_csv(file_path, sep=self.ref_separator, mode='a', index=False)\n",
    "        return file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fbb96506",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomicDataset(Dataset):\n",
    "    \"\"\"Dataset class for genomic data with masking for training\"\"\"\n",
    "\n",
    "    def __init__(self, data, targets, seq_depth, offset_before=0, offset_after=0,\n",
    "                 training=True, masking_rates=(0.5, 0.99)):\n",
    "        self.data = data\n",
    "        self.targets = targets\n",
    "        self.seq_depth = seq_depth\n",
    "        self.offset_before = offset_before\n",
    "        self.offset_after = offset_after\n",
    "        self.training = training\n",
    "        self.masking_rates = masking_rates\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.data)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        x = self.data[idx].copy()\n",
    "        y = self.targets[idx]\n",
    "\n",
    "        if self.training:\n",
    "            # Apply masking\n",
    "            seq_len = len(x)\n",
    "            masking_rate = np.random.uniform(*self.masking_rates)\n",
    "            mask_size = int(seq_len * masking_rate)\n",
    "            mask_indices = np.random.choice(seq_len, mask_size, replace=False)\n",
    "            x[mask_indices] = self.seq_depth - 1  # Missing value token\n",
    "\n",
    "        # Convert to one-hot\n",
    "        x_onehot = np.eye(self.seq_depth)[x]\n",
    "        y_onehot = np.eye(self.seq_depth - 1)[y]\n",
    "\n",
    "        return torch.FloatTensor(x_onehot), torch.FloatTensor(y_onehot)\n",
    "\n",
    "class ImputationDataset(Dataset):\n",
    "    \"\"\"Dataset for imputation (no masking needed)\"\"\"\n",
    "\n",
    "    def __init__(self, data, seq_depth):\n",
    "        self.data = data\n",
    "        self.seq_depth = seq_depth\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.data)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        x = self.data[idx]\n",
    "        # Convert to one-hot without masking\n",
    "        x_onehot = np.eye(self.seq_depth)[x]\n",
    "        return torch.FloatTensor(x_onehot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c86e007",
   "metadata": {},
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5339189",
   "metadata": {},
   "source": [
    "EvoFill 单chunk module 版"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ac21e2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "class BiMambaBlock(nn.Module):\n",
    "    \"\"\"Bidirectional Mamba block for genomic sequence processing\"\"\"\n",
    "\n",
    "    def __init__(self, d_model, d_state=16, d_conv=4, expand=2):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "\n",
    "        # Forward and backward Mamba blocks\n",
    "        self.mamba_forward = Mamba2(\n",
    "            d_model=d_model,\n",
    "            d_state=d_state,\n",
    "            d_conv=d_conv,\n",
    "            expand=expand\n",
    "        )\n",
    "\n",
    "        self.mamba_backward = Mamba2(\n",
    "            d_model=d_model,\n",
    "            d_state=d_state,\n",
    "            d_conv=d_conv,\n",
    "            expand=expand\n",
    "        )\n",
    "\n",
    "        # Layer normalization\n",
    "        self.norm1 = nn.LayerNorm(d_model)\n",
    "        self.norm2 = nn.LayerNorm(d_model)\n",
    "\n",
    "        # FFN\n",
    "        self.ffn = nn.Sequential(\n",
    "            nn.Linear(d_model * 2, d_model * 4),\n",
    "            nn.GELU(),\n",
    "            nn.Linear(d_model * 4, d_model),\n",
    "            nn.GELU()\n",
    "        )\n",
    "\n",
    "        self.dropout = nn.Dropout(0.1)\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, d_model)\n",
    "        residual = x\n",
    "\n",
    "        # Bidirectional processing\n",
    "        x_norm = self.norm1(x)\n",
    "\n",
    "        # Forward direction\n",
    "        forward_out = self.mamba_forward(x_norm)\n",
    "\n",
    "        # Backward direction (flip sequence)\n",
    "        x_backward = torch.flip(x_norm, dims=[1])\n",
    "        backward_out = self.mamba_backward(x_backward)\n",
    "        backward_out = torch.flip(backward_out, dims=[1])\n",
    "\n",
    "        # Concatenate bidirectional outputs\n",
    "        bi_out = torch.cat([forward_out, backward_out], dim=-1)\n",
    "\n",
    "        # FFN\n",
    "        ffn_out = self.ffn(bi_out)\n",
    "        ffn_out = self.dropout(ffn_out)\n",
    "\n",
    "        # Residual connection\n",
    "        out = self.norm2(residual + ffn_out)\n",
    "\n",
    "        return out\n",
    "\n",
    "class ConvBlock(nn.Module):\n",
    "    \"\"\"Convolutional block for local pattern extraction\"\"\"\n",
    "\n",
    "    def __init__(self, d_model):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "\n",
    "        self.conv1 = nn.Conv1d(d_model, d_model, kernel_size=3, padding=1)\n",
    "        self.conv2 = nn.Conv1d(d_model, d_model, kernel_size=5, padding=2)\n",
    "        self.conv3 = nn.Conv1d(d_model, d_model, kernel_size=7, padding=3)\n",
    "\n",
    "        self.conv_large1 = nn.Conv1d(d_model, d_model, kernel_size=7, padding=3)\n",
    "        self.conv_large2 = nn.Conv1d(d_model, d_model, kernel_size=15, padding=7)\n",
    "\n",
    "        self.conv_final = nn.Conv1d(d_model, d_model, kernel_size=3, padding=1)\n",
    "        self.conv_reduce = nn.Conv1d(d_model, d_model, kernel_size=1)\n",
    "\n",
    "        self.bn1 = nn.BatchNorm1d(d_model)\n",
    "        self.bn2 = nn.BatchNorm1d(d_model)\n",
    "\n",
    "        self.gelu = nn.GELU()\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, d_model)\n",
    "        x = x.transpose(1, 2)  # (batch, d_model, seq_len)\n",
    "\n",
    "        xa = self.gelu(self.conv1(x))\n",
    "\n",
    "        xb = self.gelu(self.conv2(xa))\n",
    "        xb = self.gelu(self.conv3(xb))\n",
    "\n",
    "        xc = self.gelu(self.conv_large1(xa))\n",
    "        xc = self.gelu(self.conv_large2(xc))\n",
    "\n",
    "        xa = xb + xc\n",
    "        xa = self.gelu(self.conv_final(xa))\n",
    "        xa = self.bn1(xa)\n",
    "        xa = self.gelu(self.conv_reduce(xa))\n",
    "        xa = self.bn2(xa)\n",
    "        xa = self.gelu(xa)\n",
    "\n",
    "        return xa.transpose(1, 2)  # (batch, seq_len, d_model)\n",
    "\n",
    "class CrossAttentionLayer(nn.Module):\n",
    "    \"\"\"Cross attention for integrating local and global features\"\"\"\n",
    "    def __init__(self, d_model, n_heads=8):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "        self.n_heads = n_heads\n",
    "\n",
    "        self.norm1 = nn.LayerNorm(d_model)\n",
    "        self.norm2 = nn.LayerNorm(d_model)\n",
    "        self.norm3 = nn.LayerNorm(d_model)\n",
    "\n",
    "        self.cross_attention = nn.MultiheadAttention(\n",
    "            embed_dim=d_model,\n",
    "            num_heads=n_heads,\n",
    "            batch_first=True\n",
    "        )\n",
    "\n",
    "        self.ffn = nn.Sequential(\n",
    "            nn.Linear(d_model, d_model // 2),\n",
    "            nn.GELU(),\n",
    "            nn.Linear(d_model // 2, d_model),\n",
    "            nn.GELU()\n",
    "        )\n",
    "\n",
    "    def forward(self, local_repr, global_repr, start_offset=0, end_offset=0):\n",
    "        local_norm = self.norm1(local_repr)\n",
    "        global_norm = self.norm2(global_repr)\n",
    "\n",
    "        # Apply offsets\n",
    "        if start_offset > 0 or end_offset > 0:\n",
    "            query = local_norm[:, start_offset:local_norm.shape[1] - end_offset]\n",
    "        else:\n",
    "            query = local_norm\n",
    "\n",
    "        key = value = global_norm\n",
    "\n",
    "        # Cross attention\n",
    "        attn_output, _ = self.cross_attention(query, key, value)\n",
    "\n",
    "        # Pad attn_output back to original length if offsets were applied\n",
    "        if start_offset > 0 or end_offset > 0:\n",
    "            pad_left = start_offset\n",
    "            pad_right = end_offset\n",
    "            attn_output = torch.nn.functional.pad(attn_output, (0, 0, pad_left, pad_right), mode='constant', value=0)\n",
    "\n",
    "        # Skip connection\n",
    "        attn_output = attn_output + local_norm  # Changed from +query to +local_norm\n",
    "\n",
    "        # FFN\n",
    "        attn_output = self.norm3(attn_output)\n",
    "        ffn_output = self.ffn(attn_output)\n",
    "        output = ffn_output + attn_output\n",
    "\n",
    "        return output\n",
    "\n",
    "class Mamba2CrossBlock(nn.Module):\n",
    "    \"\"\"\n",
    "    用 Mamba2Simple 替代 MultiheadAttention 的交叉块。\n",
    "    接口与原来 CrossAttentionLayer 保持一致，可直接替换。\n",
    "    \"\"\"\n",
    "    def __init__(\n",
    "        self,\n",
    "        d_model,\n",
    "        n_heads=8,            # 仅保留接口，无实际意义\n",
    "        d_state=64,\n",
    "        d_conv=4,\n",
    "        expand=2,\n",
    "        headdim=128,\n",
    "        ngroups=1,\n",
    "        chunk_size=256,\n",
    "        dropout=0.0,\n",
    "        device=None,\n",
    "        dtype=None,\n",
    "    ):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "\n",
    "        # 1. 归一化\n",
    "        self.norm1 = nn.LayerNorm(d_model)\n",
    "        self.norm2 = nn.LayerNorm(d_model)\n",
    "        self.norm3 = nn.LayerNorm(d_model)\n",
    "\n",
    "        # 2. Mamba2Simple 作为交叉建模核心\n",
    "        self.ssd = Mamba2Block(\n",
    "            d_model=d_model,\n",
    "            d_state=d_state,\n",
    "            d_conv=d_conv,\n",
    "            expand=expand,\n",
    "            headdim=headdim,\n",
    "            ngroups=ngroups,\n",
    "            chunk_size=chunk_size,\n",
    "            use_mem_eff_path=True,\n",
    "            device=device,\n",
    "            dtype=dtype,\n",
    "        )\n",
    "\n",
    "        # 3. FFN 保持不变\n",
    "        self.ffn = nn.Sequential(\n",
    "            nn.Linear(d_model, d_model // 2),\n",
    "            nn.GELU(),\n",
    "            nn.Linear(d_model // 2, d_model),\n",
    "        )\n",
    "\n",
    "    def forward(self, local_repr, global_repr, start_offset=0, end_offset=0):\n",
    "        \"\"\"\n",
    "        local_repr: [B, L, D]\n",
    "        global_repr: [B, G, D]\n",
    "        输出:       [B, L, D]  （长度与 local_repr 保持一致）\n",
    "        \"\"\"\n",
    "        B, L, D = local_repr.shape\n",
    "\n",
    "        # 1. 归一化\n",
    "        local_norm  = self.norm1(local_repr)\n",
    "        global_norm = self.norm2(global_repr)\n",
    "\n",
    "        # 2. 如果用了 offset，先截断 query 长度\n",
    "        if start_offset > 0 or end_offset > 0:\n",
    "            query = local_norm[:, start_offset:L - end_offset]\n",
    "        else:\n",
    "            query = local_norm\n",
    "\n",
    "        # 3. 拼接：global 在前，local 在后，让 SSD 扫描时 local 能看到 global\n",
    "        x = torch.cat([global_norm, query], dim=1)   # [B, G + L', D]\n",
    "        x = self.ssd(x)                              # [B, G + L', D]\n",
    "        x = x[:, global_norm.shape[1]:, :]           # 只取 local 对应部分 [B, L', D]\n",
    "\n",
    "        # 4. pad 回原始长度（如果之前截断过）\n",
    "        if start_offset > 0 or end_offset > 0:\n",
    "            x = F.pad(x, (0, 0, start_offset, end_offset))  # [B, L, D]\n",
    "\n",
    "        # 5. 残差连接\n",
    "        x = x + local_norm\n",
    "\n",
    "        # 6. FFN\n",
    "        x = self.norm3(x)\n",
    "        x = self.ffn(x) + x\n",
    "\n",
    "        return x\n",
    "\n",
    "class GenoEmbedding(nn.Module):\n",
    "    \"\"\"Genomic embedding layer with positional encoding\"\"\"\n",
    "\n",
    "    def __init__(self, n_alleles, n_snps, d_model):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "        self.n_alleles = n_alleles\n",
    "        self.n_snps = n_snps\n",
    "\n",
    "        # Allele embedding\n",
    "        self.allele_embedding = nn.Parameter(torch.randn(n_alleles, d_model))\n",
    "\n",
    "        # Positional embedding\n",
    "        self.position_embedding = nn.Embedding(n_snps, d_model)\n",
    "\n",
    "        # Initialize parameters\n",
    "        nn.init.xavier_uniform_(self.allele_embedding)\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, n_alleles) - one-hot encoded\n",
    "        batch_size, seq_len, _ = x.shape\n",
    "\n",
    "        # Allele embedding\n",
    "        embedded = torch.einsum('bsn,nd->bsd', x, self.allele_embedding)\n",
    "\n",
    "        # Positional embedding\n",
    "        positions = torch.arange(seq_len, device=x.device)\n",
    "        pos_emb = self.position_embedding(positions).unsqueeze(0)\n",
    "\n",
    "        return embedded + pos_emb\n",
    "\n",
    "class ChunkModule(nn.Module):\n",
    "    \"\"\"Single chunk processing module with BiMamba\"\"\"\n",
    "\n",
    "    def __init__(self, d_model, n_heads, start_offset=0, end_offset=0, dropout_rate=0.1):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "        self.start_offset = start_offset\n",
    "        self.end_offset = end_offset\n",
    "\n",
    "        # BiMamba block\n",
    "        self.bimamba_block = BiMambaBlock(d_model)\n",
    "\n",
    "        # Convolutional blocks\n",
    "        self.conv_block1 = ConvBlock(d_model)\n",
    "        self.conv_block2 = ConvBlock(d_model)\n",
    "        self.conv_block3 = ConvBlock(d_model)\n",
    "\n",
    "        # Cross attention\n",
    "        # self.cross_attention = CrossAttentionLayer(d_model, n_heads)\n",
    "        self.cross_attention = Mamba2CrossBlock(\n",
    "            d_model=d_model,\n",
    "            n_heads=n_heads,      # 仅保留接口，无实际用途\n",
    "            d_state=64,\n",
    "            d_conv=4,\n",
    "            expand=2,\n",
    "            headdim=128,\n",
    "            ngroups=1,\n",
    "            chunk_size=256,\n",
    "        )\n",
    "\n",
    "        # Additional layers\n",
    "        self.dense = nn.Linear(d_model, d_model)\n",
    "        self.dropout = nn.Dropout(dropout_rate)\n",
    "        self.gelu = nn.GELU()\n",
    "\n",
    "    def forward(self, x):\n",
    "        # BiMamba processing\n",
    "        xa0 = self.bimamba_block(x)\n",
    "\n",
    "        # First conv block\n",
    "        xa = self.conv_block1(xa0)\n",
    "        xa_skip = self.conv_block2(xa)\n",
    "\n",
    "        # Dense layer\n",
    "        xa = self.gelu(self.dense(xa))\n",
    "        xa = self.conv_block2(xa)\n",
    "\n",
    "        # Cross attention\n",
    "        xa = self.cross_attention(xa, xa0, self.start_offset, self.end_offset)\n",
    "        xa = self.dropout(xa)\n",
    "\n",
    "        # Final conv block\n",
    "        xa = self.conv_block3(xa)\n",
    "\n",
    "        # Concatenate with skip connection\n",
    "        xa = torch.cat([xa_skip, xa], dim=-1)\n",
    "\n",
    "        return xa\n",
    "\n",
    "class EvoFill(nn.Module):\n",
    "    \"\"\"Main BiMamba model for genomic imputation\"\"\"\n",
    "\n",
    "    def __init__(self,\n",
    "                 d_model,\n",
    "                 n_heads,\n",
    "                 chunk_size=2048,\n",
    "                 attention_range=64,\n",
    "                 offset_before=0,\n",
    "                 offset_after=0,\n",
    "                 dropout_rate=0.1):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "        self.n_heads = n_heads\n",
    "        self.chunk_size = chunk_size\n",
    "        self.attention_range = attention_range\n",
    "        self.offset_before = offset_before\n",
    "        self.offset_after = offset_after\n",
    "        self.dropout_rate = dropout_rate\n",
    "\n",
    "        # Will be set during build\n",
    "        self.seq_len = None\n",
    "        self.n_alleles = None\n",
    "        self.embedding = None\n",
    "        self.chunk_modules = nn.ModuleList()\n",
    "        self.final_conv = None\n",
    "        self.output_conv = None\n",
    "\n",
    "    def build(self, seq_len, n_alleles):\n",
    "        \"\"\"Build the model with specific sequence length and allele count\"\"\"\n",
    "        self.seq_len = seq_len\n",
    "        self.n_alleles = n_alleles\n",
    "\n",
    "        # Embedding layer\n",
    "        self.embedding = GenoEmbedding(n_alleles, seq_len, self.d_model)\n",
    "\n",
    "        # Calculate chunks\n",
    "        chunk_starts = list(range(0, seq_len, self.chunk_size))\n",
    "        chunk_ends = [min(cs + self.chunk_size, seq_len) for cs in chunk_starts]\n",
    "        mask_starts = [max(0, cs - self.attention_range) for cs in chunk_starts]\n",
    "        mask_ends = [min(ce + self.attention_range, seq_len) for ce in chunk_ends]\n",
    "\n",
    "        # Create chunk modules\n",
    "        self.chunk_module = ChunkModule(\n",
    "            d_model=self.d_model,\n",
    "            n_heads=self.n_heads,\n",
    "            start_offset=0,          # 不再给具体 offset\n",
    "            end_offset=0,\n",
    "            dropout_rate=self.dropout_rate\n",
    "        )\n",
    "\n",
    "        # Store chunk information\n",
    "        self.chunk_starts = chunk_starts\n",
    "        self.chunk_ends = chunk_ends\n",
    "        self.mask_starts = mask_starts\n",
    "        self.mask_ends = mask_ends\n",
    "\n",
    "        # Final layers\n",
    "        self.final_conv = nn.Conv1d(self.d_model * 2, self.d_model // 2,\n",
    "                                    kernel_size=5, padding=2)\n",
    "        self.output_conv = nn.Conv1d(self.d_model // 2, n_alleles - 1,\n",
    "                                     kernel_size=5, padding=2)\n",
    "        # self.output_proj = nn.Linear(self.d_model * 2, n_alleles - 1)\n",
    "\n",
    "        self.gelu = nn.GELU()\n",
    "        self.softmax = nn.Softmax(dim=-1)\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, n_alleles)\n",
    "        if self.embedding is None:\n",
    "            raise RuntimeError(\"Model not built. Call build() first.\")\n",
    "\n",
    "        # Embedding\n",
    "        x_embedded = self.embedding(x)\n",
    "\n",
    "        # Process chunks\n",
    "        chunk_outputs = []\n",
    "        for i in range(len(self.chunk_starts)):\n",
    "            pad_left  = self.chunk_starts[i] - self.mask_starts[i]\n",
    "            pad_right = self.mask_ends[i] - self.chunk_ends[i]\n",
    "            chunk_input = x_embedded[:, self.mask_starts[i]:self.mask_ends[i]]\n",
    "            chunk_output = self.chunk_module(chunk_input)   # 共享权重\n",
    "            if pad_left or pad_right:\n",
    "                chunk_output = F.pad(chunk_output, (0, 0, pad_left, pad_right))\n",
    "            chunk_outputs.append(chunk_output)\n",
    "        # Concatenate chunks along sequence dimension\n",
    "        x_concat = torch.cat(chunk_outputs, dim=1)\n",
    "\n",
    "        # # Final processing\n",
    "        x_concat = x_concat.transpose(1, 2)  # (batch, features, seq_len)\n",
    "        x_final = self.gelu(self.final_conv(x_concat))\n",
    "        x_output = self.output_conv(x_final)\n",
    "        x_output = x_output.transpose(1, 2)  # (batch, seq_len, n_alleles-1)\n",
    "        # x_output = self.output_proj(x_concat) \n",
    "\n",
    "        # Apply offsets\n",
    "        if self.offset_before > 0 or self.offset_after > 0:\n",
    "            x_output = x_output[:, self.offset_before:self.seq_len - self.offset_after]\n",
    "\n",
    "        x_output = self.softmax(x_output)\n",
    "\n",
    "        return x_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7f053a76",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ 含缺失数据前向通过，输出形状: torch.Size([2, 5120, 3])\n"
     ]
    }
   ],
   "source": [
    "n_alleles = 4  # 包含missing\n",
    "model = EvoFill(\n",
    "    d_model=256,\n",
    "    n_heads=8,\n",
    "    chunk_size=5120,\n",
    "    attention_range=64, \n",
    "    offset_before=0,\n",
    "    offset_after=0,\n",
    "    dropout_rate=0.1,\n",
    ").cuda()\n",
    "\n",
    "B, L = 2, 5120\n",
    "model.build(seq_len=L, n_alleles=n_alleles)\n",
    "model = model.cuda()  \n",
    "\n",
    "# 1. 生成输入\n",
    "x = torch.randint(0, n_alleles, (B, L)).long().cuda()   # {0,1,2,3} 3=missing\n",
    "\n",
    "# 2. -1 -> 3，并构造 one-hot（4 维）\n",
    "x_map = x.clone()\n",
    "x_onehot = torch.zeros(B, L, n_alleles, device='cuda')\n",
    "x_onehot.scatter_(2, x_map.unsqueeze(-1), 1)\n",
    "\n",
    "# 3. 前向\n",
    "with torch.no_grad():\n",
    "    probs = model(x_onehot)          # shape: (B,L,3)\n",
    "\n",
    "# 4. 简单校验\n",
    "assert torch.allclose(probs.sum(dim=-1), torch.ones(B, L, device='cuda'), atol=1e-5), \\\n",
    "    \"概率未归一\"\n",
    "print(\"✅ 含缺失数据前向通过，输出形状:\", probs.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ce8c794",
   "metadata": {},
   "source": [
    "## Loss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8f6f2206",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ImputationLoss(nn.Module):\n",
    "    \"\"\"Custom loss function for genomic imputation\"\"\"\n",
    "\n",
    "    def __init__(self, use_r2=True, \n",
    "                 use_focal=False, #  all dummy \n",
    "                 group_size=None,\n",
    "                 gamma=None,\n",
    "                 alpha=None,\n",
    "                 eps=None,\n",
    "                 use_gradnorm=None,\n",
    "                 gn_alpha=None,\n",
    "                 gn_lr_w=None,):\n",
    "        super().__init__()\n",
    "        self.use_r2_loss = use_r2\n",
    "        self.ce_loss = nn.CrossEntropyLoss(reduction='sum')\n",
    "        self.kl_loss = nn.KLDivLoss(reduction='sum')\n",
    "\n",
    "    def calculate_minimac_r2(self, pred_alt_allele_probs, gt_alt_af):\n",
    "        \"\"\"Calculate Minimac-style RÂ² metric\"\"\"\n",
    "        mask = torch.logical_or(torch.eq(gt_alt_af, 0.0), torch.eq(gt_alt_af, 1.0))\n",
    "        gt_alt_af = torch.where(mask, 0.5, gt_alt_af)\n",
    "        denom = gt_alt_af * (1.0 - gt_alt_af)\n",
    "        denom = torch.where(denom < 0.01, 0.01, denom)\n",
    "        r2 = torch.mean(torch.square(pred_alt_allele_probs - gt_alt_af), dim=0) / denom\n",
    "        r2 = torch.where(mask, torch.zeros_like(r2), r2)\n",
    "        return r2\n",
    "\n",
    "    def forward(self, y_pred, y_true):\n",
    "        y_true = y_true.float()\n",
    "\n",
    "        # Convert to proper format for losses\n",
    "        y_true_ce = torch.argmax(y_true, dim=-1)  # For CrossEntropy\n",
    "        y_pred_log = torch.log(y_pred + 1e-8)  # For KL divergence\n",
    "\n",
    "        # Basic losses\n",
    "        ce_loss = self.ce_loss(y_pred.view(-1, y_pred.size(-1)), y_true_ce.view(-1))\n",
    "        kl_loss = self.kl_loss(y_pred_log.view(-1, y_pred.size(-1)),\n",
    "                               y_true.view(-1, y_true.size(-1)))\n",
    "\n",
    "        total_loss = ce_loss + kl_loss\n",
    "\n",
    "        if self.use_r2_loss:\n",
    "            batch_size = y_true.size(0)\n",
    "            group_size = 4\n",
    "            num_full_groups = batch_size // group_size\n",
    "\n",
    "            if num_full_groups > 0:\n",
    "                y_true_grouped = y_true[:num_full_groups * group_size].view(\n",
    "                    num_full_groups, group_size, *y_true.shape[1:])\n",
    "                y_pred_grouped = y_pred[:num_full_groups * group_size].view(\n",
    "                    num_full_groups, group_size, *y_pred.shape[1:])\n",
    "\n",
    "                r2_loss = 0.0\n",
    "                for i in range(num_full_groups):\n",
    "                    gt_alt_af = torch.count_nonzero(\n",
    "                        torch.argmax(y_true_grouped[i], dim=-1), dim=0\n",
    "                    ).float() / group_size\n",
    "\n",
    "                    pred_alt_allele_probs = torch.sum(y_pred_grouped[i][:, :, 1:], dim=-1)\n",
    "                    r2_loss += -torch.sum(self.calculate_minimac_r2(\n",
    "                        pred_alt_allele_probs, gt_alt_af)) * group_size\n",
    "\n",
    "                total_loss += r2_loss\n",
    "\n",
    "        return total_loss, None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d0ff63",
   "metadata": {},
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bb648d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_similar_rows(array):\n",
    "    \"\"\"Remove duplicate haploids from training set\"\"\"\n",
    "    print(\"Finding duplicate haploids in training set.\")\n",
    "    unique_array = np.unique(array, axis=0)\n",
    "    print(f\"Removed {len(array) - len(unique_array)} rows. {len(unique_array)} training samples remaining.\")\n",
    "    return unique_array\n",
    "\n",
    "def create_directories(save_dir, models_dir=\"models\", outputs=\"out\") -> None:\n",
    "    \"\"\"Create necessary directories\"\"\"\n",
    "    for dd in [save_dir, f\"{save_dir}/{models_dir}\", f\"{save_dir}/{outputs}\"]:\n",
    "        if not os.path.exists(dd):\n",
    "            os.makedirs(dd)\n",
    "\n",
    "def clear_dir(path) -> None:\n",
    "    \"\"\"Clear directory if it exists\"\"\"\n",
    "    if os.path.exists(path):\n",
    "        shutil.rmtree(path)\n",
    "\n",
    "def load_chunk_info(save_dir, break_points):\n",
    "    \"\"\"Load chunk training status information\"\"\"\n",
    "    chunk_info = {ww: False for ww in list(range(len(break_points) - 1))}\n",
    "    if os.path.isfile(f\"{save_dir}/models/chunks_info.json\"):\n",
    "        with open(f\"{save_dir}/models/chunks_info.json\", 'r') as f:\n",
    "            loaded_chunks_info = json.load(f)\n",
    "            if isinstance(loaded_chunks_info, dict) and len(loaded_chunks_info) == len(chunk_info):\n",
    "                print(\"Resuming the training...\")\n",
    "                chunk_info = {int(k): v for k, v in loaded_chunks_info.items()}\n",
    "    return chunk_info\n",
    "\n",
    "def save_chunk_status(save_dir, chunk_info) -> None:\n",
    "    \"\"\"Save chunk training status information\"\"\"\n",
    "    with open(f\"{save_dir}/models/chunks_info.json\", \"w\") as outfile:\n",
    "        json.dump(chunk_info, outfile)\n",
    "\n",
    "def create_model(args, seq_len, n_alleles):\n",
    "    \"\"\"Create BiMamba model\"\"\"\n",
    "    model = EvoFill(\n",
    "        d_model=args.embed_dim,\n",
    "        n_heads=args.na_heads,\n",
    "        chunk_size=args.cs,\n",
    "        attention_range=args.co,\n",
    "        offset_before=getattr(args, 'offset_before', 0),\n",
    "        offset_after=getattr(args, 'offset_after', 0),\n",
    "        dropout_rate=0.1\n",
    "    )\n",
    "\n",
    "    # Build the model\n",
    "    model.build(seq_len, n_alleles)\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7c302bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAF_BINS = [(0.00, 0.05), (0.05, 0.10), (0.10, 0.20),\n",
    "            (0.20, 0.30), (0.30, 0.40), (0.40, 0.50)]\n",
    "\n",
    "def precompute_maf(gts_np, mask_int=-1):\n",
    "    \"\"\"\n",
    "    gts_np: (N, L)  int64\n",
    "    return:\n",
    "        maf: (L,) float32\n",
    "        bin_cnt: list[int] 长度 6，对应 6 个 bin 的位点数量\n",
    "    \"\"\"\n",
    "    L = gts_np.shape[1]\n",
    "    maf = np.zeros(L, dtype=np.float32)\n",
    "    bin_cnt = [0] * 6\n",
    "\n",
    "    for l in range(L):\n",
    "        alleles = gts_np[:, l]\n",
    "        alleles = alleles[alleles != mask_int]   # 去掉缺失\n",
    "        if alleles.size == 0:\n",
    "            maf[l] = 0.0\n",
    "            continue\n",
    "\n",
    "        uniq, cnt = np.unique(alleles, return_counts=True)\n",
    "        total = cnt.sum()\n",
    "        freq = cnt / total\n",
    "        freq[::-1].sort()\n",
    "        maf_val = freq[1] if len(freq) > 1 else 0.0\n",
    "        maf[l] = maf_val\n",
    "\n",
    "        # 统计 bin\n",
    "        for i, (lo, hi) in enumerate(MAF_BINS):\n",
    "            if lo <= maf_val < hi:\n",
    "                bin_cnt[i] += 1\n",
    "                break\n",
    "\n",
    "    return maf, bin_cnt\n",
    "\n",
    "def imputation_maf_accuracy_epoch(all_logits, all_gts, global_maf, mask=None):\n",
    "    \"\"\"\n",
    "    all_logits: (N, L, C)\n",
    "    all_gts:    (N, L, C) one-hot\n",
    "    global_maf: (L,)\n",
    "    mask:       (N, L) 或 None\n",
    "    return:     list[float] 长度 6\n",
    "    \"\"\"\n",
    "    # 1. 预测 vs 真实\n",
    "    all_gts = all_gts.argmax(dim=-1)      # (N, L)\n",
    "    preds   = all_logits.argmax(dim=-1)   # (N, L)\n",
    "\n",
    "    # 2. 如果没有外部 mask，就默认全 1\n",
    "    if mask is None:\n",
    "        mask = torch.ones_like(all_gts, dtype=torch.bool)   # (N, L)\n",
    "    correct = (preds == all_gts) & mask                   # (N, L)\n",
    "\n",
    "    # 3. MAF 条件 -> (1, L) 再广播到 (N, L)\n",
    "    maf = global_maf.unsqueeze(0)                         # (1, L)\n",
    "\n",
    "    # 4. 分 bin 计算\n",
    "    accs = []\n",
    "    for lo, hi in MAF_BINS:\n",
    "        maf_bin = mask & (maf >= lo) & (maf < hi)                # (1, L)\n",
    "        n_cor = (correct & maf_bin).sum()\n",
    "        n_tot = maf_bin.sum()\n",
    "        accs.append(100*(n_cor / n_tot).item() if n_tot > 0 else 0.0)\n",
    "    return accs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d21fb7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using device: cuda\n",
      "Reading the file...\n",
      "2000 diploid samples with 56145 variants found!\n",
      "DEBUG: Unique genotypes in dataset: ['0|0' '0|1' '1|0' '1|1' '2|0' '2|1' '2|2' '1|2' '0|2']...\n",
      "DEBUG: self.genotype_vals: ['0|0' '0|1' '0|2' '1|0' '1|1' '1|2' '2|0' '2|1' '2|2']\n",
      "DEBUG: self.alleles: ['2' '1' '0']\n",
      "DEBUG: is_phased: True\n",
      "DEBUG: hap_map: {'0': 0, '1': 1, '2': 2, '.': 3}\n",
      "DEBUG: self.SEQ_DEPTH: 4\n",
      "Training on chunk 1/7\n",
      "Data shape: (4000, 8320)\n",
      "Finding duplicate haploids in training set.\n",
      "Removed 1175 rows. 2569 training samples remaining.\n",
      "Chunk MAF-bin counts: [7421, 224, 210, 129, 108, 100]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                          \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 1/100, Train Loss: 211528.0077, Val Loss: 205887.8457\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)   7421 99.01 99.38\n",
      "(0.05, 0.10)    224 87.24 86.80\n",
      "(0.10, 0.20)    210 76.32 72.27\n",
      "(0.20, 0.30)    129 65.30 59.32\n",
      "(0.30, 0.40)    108 55.53 52.71\n",
      "(0.40, 0.50)    100 50.32 49.57\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                          \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 2/100, Train Loss: 185111.4103, Val Loss: 195619.4180\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)   7421 99.58 99.34\n",
      "(0.05, 0.10)    224 90.81 87.16\n",
      "(0.10, 0.20)    210 84.15 75.79\n",
      "(0.20, 0.30)    129 73.72 64.01\n",
      "(0.30, 0.40)    108 63.37 55.60\n",
      "(0.40, 0.50)    100 54.67 55.13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                          \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 3/100, Train Loss: 173497.1838, Val Loss: 178869.9492\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)   7421 99.58 99.37\n",
      "(0.05, 0.10)    224 90.96 89.17\n",
      "(0.10, 0.20)    210 84.52 80.72\n",
      "(0.20, 0.30)    129 74.70 70.47\n",
      "(0.30, 0.40)    108 64.38 60.50\n",
      "(0.40, 0.50)    100 56.05 55.79\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                          \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 4/100, Train Loss: 164481.6905, Val Loss: 172987.3320\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)   7421 99.58 99.38\n",
      "(0.05, 0.10)    224 90.91 89.20\n",
      "(0.10, 0.20)    210 84.79 81.55\n",
      "(0.20, 0.30)    129 75.38 72.13\n",
      "(0.30, 0.40)    108 64.69 62.40\n",
      "(0.40, 0.50)    100 56.30 57.23\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Epoch 5/100:  17%|█▋        | 14/81 [00:05<00:22,  3.00it/s, loss=1.63e+5]"
     ]
    }
   ],
   "source": [
    "# ---------------- 以下即命令行参数对应的行内变量 ----------------\n",
    "mode                 = 'train'\n",
    "restart_training     = True          # 对应命令行 1\n",
    "ref                  = '/home/qmtang/mnt_qmtang/EvoFill/data/sim_s2_250901/YRI_CEU_chr22.vcf.gz'\n",
    "tihp                 = True          # 对应命令行 1\n",
    "which_chunk          = -1            # All chunkss\n",
    "save_dir             = '/home/qmtang/mnt_qmtang/EvoFill/data/mam_YRI_1chunkmod'\n",
    "co                   = 64\n",
    "cs                   = 1024\n",
    "sites_per_model      = 8192\n",
    "max_mr               = 0.7\n",
    "min_mr               = 0.3\n",
    "epochs               = 100\n",
    "na_heads             = 8\n",
    "embed_dim            = 64\n",
    "lr                   = 0.001\n",
    "weight_decay         = 1e-5\n",
    "batch_size_per_gpu   = 32\n",
    "use_r2               = True\n",
    "use_focal            = True\n",
    "earlystop_patience   = 9\n",
    "verbose              = 1\n",
    "# -------------------------------------------------------------\n",
    "\n",
    "# 组装成 Namespace\n",
    "args = Namespace(\n",
    "    restart_training=restart_training,\n",
    "    ref=ref,\n",
    "    tihp=tihp,\n",
    "    which_chunk=which_chunk,\n",
    "    save_dir=save_dir,\n",
    "    co=co,\n",
    "    cs=cs,\n",
    "    sites_per_model=sites_per_model,\n",
    "    max_mr=max_mr,\n",
    "    min_mr=min_mr,\n",
    "    epochs=epochs,\n",
    "    na_heads=na_heads,\n",
    "    embed_dim=embed_dim,\n",
    "    lr=lr,\n",
    "    weight_decay=weight_decay,\n",
    "    batch_size_per_gpu=batch_size_per_gpu,\n",
    "    use_r2=use_r2,\n",
    "    earlystop_patience=earlystop_patience,\n",
    "    verbose=verbose,\n",
    ")\n",
    "\n",
    "assert args.max_mr > 0\n",
    "assert args.min_mr > 0\n",
    "assert args.max_mr >= args.min_mr\n",
    "\n",
    "# Setup device\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "print(f\"Using device: {device}\")\n",
    "\n",
    "# Create directories\n",
    "create_directories(args.save_dir)\n",
    "with open(f\"{args.save_dir}/commandline_args.json\", 'w') as f:\n",
    "    json.dump(vars(args), f, indent=4)\n",
    "\n",
    "# Load data\n",
    "dr = DataReader()\n",
    "dr.assign_training_set(\n",
    "    file_path=args.ref,\n",
    "    target_is_gonna_be_phased_or_haps=args.tihp,\n",
    "    variants_as_columns=getattr(args, 'ref_vac', False),\n",
    "    delimiter=getattr(args, 'ref_sep', None),\n",
    "    file_format=getattr(args, 'ref_file_format', 'infer'),\n",
    "    first_column_is_index=getattr(args, 'ref_fcai', True),\n",
    "    comments=getattr(args, 'ref_comment', '##')\n",
    ")\n",
    "\n",
    "# Split data for validation\n",
    "n_samples = dr.get_ref_set(0, 1).shape[0]\n",
    "val_n_samples = args.batch_size_per_gpu * getattr(args, 'val_n_batches', 8)\n",
    "x_train_indices, x_valid_indices = train_test_split(\n",
    "    range(n_samples),\n",
    "    test_size=val_n_samples,\n",
    "    random_state=getattr(args, 'random_seed', 2022),\n",
    "    shuffle=True\n",
    ")\n",
    "\n",
    "# Process chunks\n",
    "break_points = list(np.arange(0, dr.VARIANT_COUNT, args.sites_per_model)) + [dr.VARIANT_COUNT]\n",
    "chunks_done = load_chunk_info(args.save_dir, break_points)\n",
    "\n",
    "for w in range(len(break_points) - 1):\n",
    "    if chunks_done[w]:\n",
    "        print(f\"Skipping chunk {w + 1}/{len(break_points) - 1} due to previous training.\")\n",
    "        continue\n",
    "    if args.which_chunk != -1 and w + 1 != args.which_chunk:\n",
    "        print(f\"Skipping chunk {w + 1}/{len(break_points) - 1} due to your request using --which-chunk.\")\n",
    "        continue\n",
    "\n",
    "    print(f\"Training on chunk {w + 1}/{len(break_points) - 1}\")\n",
    "\n",
    "    # Calculate chunk boundaries\n",
    "    final_start_pos = max(0, break_points[w] - 2 * args.co)\n",
    "    final_end_pos = min(dr.VARIANT_COUNT, break_points[w + 1] + 2 * args.co)\n",
    "    offset_before = break_points[w] - final_start_pos\n",
    "    offset_after = final_end_pos - break_points[w + 1]\n",
    "\n",
    "    # Get data for this chunk\n",
    "    ref_set = dr.get_ref_set(final_start_pos, final_end_pos).astype(np.int32)\n",
    "    print(f\"Data shape: {ref_set.shape}\")\n",
    "\n",
    "    # Remove duplicates from training\n",
    "    ref_set_train = remove_similar_rows(ref_set[x_train_indices])\n",
    "    ref_set_val = ref_set[x_valid_indices]\n",
    "\n",
    "    # MAF bins counts\n",
    "    valid_slice = slice(offset_before,\n",
    "                        ref_set_train.shape[1] - offset_after)\n",
    "    chunk_maf, chunk_bin_cnt = precompute_maf(\n",
    "        ref_set_train[:, valid_slice], \n",
    "        mask_int=-1\n",
    "    )\n",
    "    chunk_maf = torch.from_numpy(chunk_maf).to(device)          # (L_chunk,)\n",
    "    if args.verbose:\n",
    "        print('Chunk MAF-bin counts:', chunk_bin_cnt)\n",
    "\n",
    "    # Create targets (same as input for reconstruction)\n",
    "    target_train = ref_set_train[:, offset_before:ref_set_train.shape[1] - offset_after]\n",
    "    target_val = ref_set_val[:, offset_before:ref_set_val.shape[1] - offset_after]\n",
    "\n",
    "    # Create datasets\n",
    "    train_dataset = GenomicDataset(\n",
    "        ref_set_train, target_train, dr.SEQ_DEPTH,\n",
    "        offset_before, offset_after, training=True,\n",
    "        masking_rates=(args.min_mr, args.max_mr)\n",
    "    )\n",
    "\n",
    "    val_dataset = GenomicDataset(\n",
    "        ref_set_val, target_val, dr.SEQ_DEPTH,\n",
    "        offset_before, offset_after, training=False,\n",
    "        masking_rates=(args.min_mr, args.max_mr)\n",
    "    )\n",
    "\n",
    "    # Create data loaders\n",
    "    train_loader = DataLoader(train_dataset, batch_size=args.batch_size_per_gpu,\n",
    "                                shuffle=True, num_workers=4, pin_memory=True)\n",
    "    val_loader = DataLoader(val_dataset, batch_size=args.batch_size_per_gpu,\n",
    "                            shuffle=False, num_workers=4, pin_memory=True)\n",
    "\n",
    "    # Create model\n",
    "    seq_len = ref_set.shape[1]\n",
    "    model = create_model(args, seq_len, dr.SEQ_DEPTH)\n",
    "    model.offset_before = offset_before\n",
    "    model.offset_after = offset_after\n",
    "    model.to(device)\n",
    "\n",
    "    # Loss and optimizer\n",
    "    criterion = ImputationLoss(use_r2=getattr(args, 'use_r2', True),use_focal=getattr(args, 'use_focal', True))\n",
    "    optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)\n",
    "    # optimizer = Lamb(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)\n",
    "    \n",
    "    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(\n",
    "        optimizer, mode='min', factor=0.5, patience=3, min_lr=1e-7\n",
    "    )\n",
    "\n",
    "    # Training loop\n",
    "    best_loss = float('inf')\n",
    "    patience = args.earlystop_patience\n",
    "    patience_counter = 0\n",
    "\n",
    "    for epoch in range(args.epochs):\n",
    "        model.train()\n",
    "        train_loss = 0.0\n",
    "        train_logits, train_gts, train_mask = [], [], []\n",
    "\n",
    "        train_pbar = tqdm(train_loader, desc=f'Epoch {epoch + 1}/{args.epochs}', leave=False)\n",
    "        for batch_idx, (data, target) in enumerate(train_pbar):\n",
    "            data, target = data.to(device), target.to(device)\n",
    "\n",
    "            optimizer.zero_grad()\n",
    "            output = model(data)\n",
    "            loss, logs = criterion(output, target)\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "\n",
    "            train_loss += loss.item()\n",
    "            train_pbar.set_postfix({'loss': loss.item()})\n",
    "\n",
    "            # === 收集训练结果 ===\n",
    "            mask = data[..., -1].bool()         # 只关心被 mask 的位点\n",
    "            train_logits.append(output.detach())\n",
    "            train_gts.append(target.detach())\n",
    "            train_mask.append(mask)\n",
    "\n",
    "        # 训练集 MAF-acc\n",
    "        train_logits = torch.cat(train_logits, dim=0)\n",
    "        train_gts    = torch.cat(train_gts,    dim=0)\n",
    "        train_mask   = torch.cat(train_mask,   dim=0)\n",
    "        # 只保留有效位点（去掉 offset  padding）\n",
    "        if model.offset_before > 0 or model.offset_after > 0:\n",
    "            train_mask   = train_mask  [:, model.offset_before : train_mask.shape[1]  -model.offset_after]\n",
    "        train_maf_accs = imputation_maf_accuracy_epoch(train_logits, train_gts, chunk_maf, mask=train_mask)\n",
    "\n",
    "        # ----------- 验证循环同理 ------------\n",
    "        model.eval()\n",
    "        val_loss = 0.0\n",
    "        val_logits, val_gts = [], []\n",
    "        with torch.no_grad():\n",
    "            for data, target in val_loader:\n",
    "                data, target = data.to(device), target.to(device)\n",
    "                output = model(data)\n",
    "                loss, logs = criterion(output, target)\n",
    "                val_loss += loss.item()\n",
    "\n",
    "                mask = data[..., -1].bool()\n",
    "                val_logits.append(output)\n",
    "                val_gts.append(target)\n",
    "\n",
    "        val_logits = torch.cat(val_logits, dim=0)\n",
    "        val_gts    = torch.cat(val_gts,    dim=0)\n",
    "        val_maf_accs = imputation_maf_accuracy_epoch(\n",
    "            val_logits, val_gts, chunk_maf,  mask=None,)\n",
    "\n",
    "        avg_train_loss = train_loss / len(train_loader)\n",
    "        avg_val_loss   = val_loss   / len(val_loader)\n",
    "\n",
    "        if args.verbose >= 1:\n",
    "            avg_train_loss = train_loss / len(train_loader)\n",
    "            avg_val_loss   = val_loss   / len(val_loader)\n",
    "            print(f'Chunk {w + 1}, Epoch {epoch + 1}/{args.epochs}, '\n",
    "                  f'Train Loss: {avg_train_loss:.4f}, Val Loss: {avg_val_loss:.4f}')\n",
    "\n",
    "            # 用 DataFrame 打印 MAF-bin 结果\n",
    "            maf_df = pd.DataFrame({\n",
    "                'MAF_bin': ['(0.00, 0.05)', '(0.05, 0.10)', '(0.10, 0.20)',\n",
    "                            '(0.20, 0.30)', '(0.30, 0.40)', '(0.40, 0.50)'],\n",
    "                'Counts':  [f\"{c}\" for c in chunk_bin_cnt],\n",
    "                'Train':   [f\"{acc:.2f}\" for acc in train_maf_accs],\n",
    "                'Val':     [f\"{acc:.2f}\" for acc in val_maf_accs]\n",
    "            })\n",
    "            print(maf_df.to_string(index=False))\n",
    "\n",
    "        scheduler.step(avg_val_loss)\n",
    "\n",
    "        # Early stopping\n",
    "        if avg_val_loss < best_loss:\n",
    "            best_loss = avg_val_loss\n",
    "            patience_counter = 0\n",
    "            # Save best model\n",
    "            import os\n",
    "            os.makedirs(f'{args.save_dir}/models', exist_ok=True)\n",
    "            torch.save(model.state_dict(), f'{args.save_dir}/models/w_{w}.pth')\n",
    "        else:\n",
    "            patience_counter += 1\n",
    "            if patience_counter >= patience:\n",
    "                if args.verbose >= 1:\n",
    "                    print('Early stopping triggered')\n",
    "                break\n",
    "\n",
    "    # Mark chunk as done\n",
    "    chunks_done[w] = True\n",
    "    save_chunk_status(args.save_dir, chunks_done)\n",
    "\n",
    "    # Clean up\n",
    "    del ref_set, train_dataset, val_dataset, train_loader, val_loader, model\n",
    "    torch.cuda.empty_cache() if torch.cuda.is_available() else None"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
