{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "aa6bc4ec",
   "metadata": {},
   "source": [
    "ver1: chunk module 共享权重\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d4bb00d",
   "metadata": {},
   "source": [
    "## Dependency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "757312f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os; os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"0\" # 设置用GPU1\n",
    "import gzip\n",
    "import json\n",
    "import logging\n",
    "import shutil\n",
    "from typing import Union\n",
    "from argparse import Namespace\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import datatable as dt\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from torch.autograd import grad\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "\n",
    "from tqdm import tqdm\n",
    "from joblib import Parallel, delayed\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "from mamba_ssm import Mamba2\n",
    "from mamba_ssm.modules.mamba2_simple import Mamba2Simple as Mamba2Block # 原Mamba2Block\n",
    "from torch_optimizer import Lamb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "975df815",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e5eeded0",
   "metadata": {},
   "outputs": [],
   "source": [
    "SUPPORTED_FILE_FORMATS = {\"vcf\", \"csv\", \"tsv\"}\n",
    "class DataReader:\n",
    "    def __init__(self):\n",
    "        self.target_is_gonna_be_phased = None\n",
    "        self.target_set = None\n",
    "        self.target_sample_value_index = 2\n",
    "        self.ref_sample_value_index = 2\n",
    "        self.target_file_extension = None\n",
    "        self.allele_count = 2\n",
    "        self.genotype_vals = None\n",
    "        self.ref_is_phased = None\n",
    "        self.reference_panel = None\n",
    "        self.VARIANT_COUNT = 0\n",
    "        self.is_phased = False\n",
    "        self.MISSING_VALUE = None\n",
    "        self.ref_is_hap = False\n",
    "        self.target_is_hap = False\n",
    "        self.ref_n_header_lines = []\n",
    "        self.ref_n_data_header = \"\"\n",
    "        self.target_n_header_lines = []\n",
    "        self.target_n_data_header = \"\"\n",
    "        self.ref_separator = None\n",
    "        self.map_values_1_vec = np.vectorize(self.__map_hap_2_ind_parent_1)\n",
    "        self.map_values_2_vec = np.vectorize(self.__map_hap_2_ind_parent_2)\n",
    "        self.map_haps_to_vec = np.vectorize(self.__map_haps_2_ind)\n",
    "        self.delimiter_dictionary = {\"vcf\": \"\\t\", \"csv\": \",\", \"tsv\": \"\\t\", \"infer\": \"\\t\"}\n",
    "        self.ref_file_extension = \"vcf\"\n",
    "        self.test_file_extension = \"vcf\"\n",
    "        self.target_is_phased = True\n",
    "\n",
    "    def __read_csv(self, file_path, is_vcf=False, is_reference=False, separator=\"\\t\", first_column_is_index=True,\n",
    "                   comments=\"##\") -> pd.DataFrame:\n",
    "        \"\"\"Read CSV/VCF files\"\"\"\n",
    "        print(\"Reading the file...\")\n",
    "        data_header = None\n",
    "        path_sep = \"/\" if \"/\" in file_path else os.path.sep\n",
    "        line_counter = 0\n",
    "        root, ext = os.path.splitext(file_path)\n",
    "        with gzip.open(file_path, 'rt') if ext == '.gz' else open(file_path, 'rt') as f_in:\n",
    "            while True:\n",
    "                line = f_in.readline()\n",
    "                if line.startswith(comments):\n",
    "                    line_counter += 1\n",
    "                    if is_reference:\n",
    "                        self.ref_n_header_lines.append(line)\n",
    "                    else:\n",
    "                        self.target_n_header_lines.append(line)\n",
    "                else:\n",
    "                    data_header = line\n",
    "                    break\n",
    "        if data_header is None:\n",
    "            raise IOError(\"The file only contains comments!\")\n",
    "        df = dt.fread(file=file_path, sep=separator, header=True, skip_to_line=line_counter + 1)\n",
    "        df = df.to_pandas()\n",
    "        if first_column_is_index:\n",
    "            df.set_index(df.columns[0], inplace=True)\n",
    "        return df\n",
    "\n",
    "    def __find_file_extension(self, file_path, file_format, delimiter):\n",
    "        separator = \"\\t\"\n",
    "        found_file_format = None\n",
    "        if file_format not in [\"infer\"] + list(SUPPORTED_FILE_FORMATS):\n",
    "            raise ValueError(\"File extension must be one of {'vcf', 'csv', 'tsv', 'infer'}.\")\n",
    "        if file_format == 'infer':\n",
    "            file_name_tokenized = file_path.split(\".\")\n",
    "            for possible_extension in file_name_tokenized[::-1]:\n",
    "                if possible_extension in SUPPORTED_FILE_FORMATS:\n",
    "                    found_file_format = possible_extension\n",
    "                    separator = self.delimiter_dictionary[possible_extension] if delimiter is None else delimiter\n",
    "                    break\n",
    "            if found_file_format is None:\n",
    "                logging.warning(\"Could not infer the file type. Using tsv as the last resort.\")\n",
    "                found_file_format = \"tsv\"\n",
    "        else:\n",
    "            found_file_format = file_format\n",
    "            separator = self.delimiter_dictionary[file_format] if delimiter is None else delimiter\n",
    "        return found_file_format, separator\n",
    "\n",
    "    def assign_training_set(self, file_path: str, target_is_gonna_be_phased_or_haps: bool,\n",
    "                            variants_as_columns: bool = False, delimiter=None, file_format=\"infer\",\n",
    "                            first_column_is_index=True, comments=\"##\") -> None:\n",
    "        self.target_is_gonna_be_phased = target_is_gonna_be_phased_or_haps\n",
    "        self.ref_file_extension, self.ref_separator = self.__find_file_extension(file_path, file_format, delimiter)\n",
    "\n",
    "        self.reference_panel = self.__read_csv(file_path, is_reference=True, is_vcf=False, separator=self.ref_separator,\n",
    "                                               first_column_is_index=first_column_is_index,\n",
    "                                               comments=comments) if self.ref_file_extension != 'vcf' else self.__read_csv(\n",
    "            file_path, is_reference=True, is_vcf=True, separator='\\t', first_column_is_index=False, comments=\"##\")\n",
    "\n",
    "        if self.ref_file_extension != \"vcf\":\n",
    "            if variants_as_columns:\n",
    "                self.reference_panel = self.reference_panel.transpose()\n",
    "            self.reference_panel.reset_index(drop=False, inplace=True)\n",
    "            self.reference_panel.rename(columns={self.reference_panel.columns[0]: \"ID\"}, inplace=True)\n",
    "        else:\n",
    "            self.ref_sample_value_index += 8\n",
    "\n",
    "        self.ref_is_hap = not (\"|\" in self.reference_panel.iloc[0, self.ref_sample_value_index - 1] or \"/\" in\n",
    "                               self.reference_panel.iloc[0, self.ref_sample_value_index - 1])\n",
    "        self.ref_is_phased = \"|\" in self.reference_panel.iloc[0, self.ref_sample_value_index - 1]\n",
    "\n",
    "        if self.ref_is_hap and not target_is_gonna_be_phased_or_haps:\n",
    "            raise ValueError(\n",
    "                \"Reference contains haploids while target will be unphased diploids. Model cannot predict target.\")\n",
    "\n",
    "        if not (self.ref_is_phased or self.ref_is_hap) and target_is_gonna_be_phased_or_haps:\n",
    "            raise ValueError(\n",
    "                \"Reference contains unphased diploids while target will be phased/haploid. Model cannot predict target.\")\n",
    "\n",
    "        self.VARIANT_COUNT = self.reference_panel.shape[0]\n",
    "        print(\n",
    "            f\"{self.reference_panel.shape[1] - (self.ref_sample_value_index - 1)} {'haploid' if self.ref_is_hap else 'diploid'} samples with {self.VARIANT_COUNT} variants found!\")\n",
    "\n",
    "        self.is_phased = target_is_gonna_be_phased_or_haps and (self.ref_is_phased or self.ref_is_hap)\n",
    "\n",
    "        original_allele_sep = \"|\" if self.ref_is_phased or self.ref_is_hap else \"/\"\n",
    "        final_allele_sep = \"|\" if self.is_phased else \"/\"\n",
    "\n",
    "        def get_diploid_alleles(genotype_vals):\n",
    "            allele_set = set()\n",
    "            for genotype_val in genotype_vals:\n",
    "                if genotype_val not in [\".\", \".|.\", \"./.\"]:\n",
    "                    if final_allele_sep in genotype_val:\n",
    "                        v1, v2 = genotype_val.split(final_allele_sep)\n",
    "                        allele_set.update([v1, v2])\n",
    "                    else:\n",
    "                        allele_set.add(genotype_val)  # For haploids\n",
    "            return np.array(list(allele_set))\n",
    "\n",
    "        genotype_vals = pd.unique(self.reference_panel.iloc[:, self.ref_sample_value_index - 1:].values.ravel('K'))\n",
    "        print(f\"DEBUG: Unique genotypes in dataset: {genotype_vals[:10]}...\")  # Show first 10\n",
    "\n",
    "        if self.ref_is_phased and not target_is_gonna_be_phased_or_haps:\n",
    "            phased_to_unphased_dict = {}\n",
    "            for i in range(genotype_vals.shape[0]):\n",
    "                key = genotype_vals[i]\n",
    "                if \"|\" in key and key not in [\".\", \".|.\"]:\n",
    "                    v1, v2 = [int(s) for s in genotype_vals[i].split(original_allele_sep)]\n",
    "                    genotype_vals[i] = f\"{min(v1, v2)}/{max(v1, v2)}\"\n",
    "                    phased_to_unphased_dict[key] = genotype_vals[i]\n",
    "            if phased_to_unphased_dict:\n",
    "                self.reference_panel.iloc[:, self.ref_sample_value_index - 1:].replace(phased_to_unphased_dict,\n",
    "                                                                                       inplace=True)\n",
    "\n",
    "        self.genotype_vals = np.unique(genotype_vals)\n",
    "        self.alleles = get_diploid_alleles(self.genotype_vals) if not self.ref_is_hap else self.genotype_vals\n",
    "        self.allele_count = len(self.alleles)\n",
    "        self.MISSING_VALUE = self.allele_count if self.is_phased else len(self.genotype_vals)\n",
    "\n",
    "        print(f\"DEBUG: self.genotype_vals: {self.genotype_vals}\")\n",
    "        print(f\"DEBUG: self.alleles: {self.alleles}\")\n",
    "        print(f\"DEBUG: is_phased: {self.is_phased}\")\n",
    "\n",
    "        if self.is_phased:\n",
    "            self.hap_map = {str(v): i for i, v in enumerate(list(sorted(self.alleles)))}\n",
    "            self.hap_map.update({\".\": self.MISSING_VALUE})\n",
    "            self.r_hap_map = {i: k for k, i in self.hap_map.items()}\n",
    "            self.map_preds_2_allele = np.vectorize(lambda x: self.r_hap_map[x])\n",
    "            print(f\"DEBUG: hap_map: {self.hap_map}\")\n",
    "        else:\n",
    "            unphased_missing_genotype = \"./.\"\n",
    "            self.replacement_dict = {g: i for i, g in enumerate(list(sorted(self.genotype_vals)))}\n",
    "            self.replacement_dict[unphased_missing_genotype] = self.MISSING_VALUE\n",
    "            self.reverse_replacement_dict = {v: k for k, v in self.replacement_dict.items()}\n",
    "            print(f\"DEBUG: replacement_dict: {self.replacement_dict}\")\n",
    "\n",
    "        self.SEQ_DEPTH = self.allele_count + 1 if self.is_phased else len(self.genotype_vals) + 1\n",
    "        print(f\"DEBUG: self.SEQ_DEPTH: {self.SEQ_DEPTH}\")\n",
    "\n",
    "    def assign_test_set(self, file_path, variants_as_columns=False, delimiter=None,\n",
    "                        file_format=\"infer\", first_column_is_index=True, comments=\"##\") -> None:\n",
    "        \"\"\"Assign test set for imputation\"\"\"\n",
    "        if self.reference_panel is None:\n",
    "            raise RuntimeError(\"First you need to use 'DataReader.assign_training_set(...) to assign a training set.'\")\n",
    "\n",
    "        self.target_file_extension, separator = self.__find_file_extension(file_path, file_format, delimiter)\n",
    "\n",
    "        test_df = self.__read_csv(file_path, is_reference=False, is_vcf=False, separator=separator,\n",
    "                                  first_column_is_index=first_column_is_index,\n",
    "                                  comments=comments) if self.target_file_extension != 'vcf' else self.__read_csv(\n",
    "            file_path, is_reference=False, is_vcf=True, separator='\\t', first_column_is_index=False, comments=\"##\")\n",
    "\n",
    "        if self.target_file_extension != \"vcf\":\n",
    "            if variants_as_columns:\n",
    "                test_df = test_df.transpose()\n",
    "            test_df.reset_index(drop=False, inplace=True)\n",
    "            test_df.rename(columns={test_df.columns[0]: \"ID\"}, inplace=True)\n",
    "        else:\n",
    "            self.target_sample_value_index += 8\n",
    "\n",
    "        self.target_is_hap = not (\"|\" in test_df.iloc[0, self.target_sample_value_index - 1] or \"/\" in\n",
    "                                  test_df.iloc[0, self.target_sample_value_index - 1])\n",
    "        is_phased = \"|\" in test_df.iloc[0, self.target_sample_value_index - 1]\n",
    "        test_var_count = test_df.shape[0]\n",
    "        print(f\"{test_var_count} {'haplotype' if self.target_is_hap else 'diplotype'} variants found!\")\n",
    "\n",
    "        # Validate compatibility\n",
    "        if (self.target_is_hap or is_phased) and not (self.ref_is_phased or self.ref_is_hap):\n",
    "            raise RuntimeError(\"The training set contains unphased data. The target must be unphased as well.\")\n",
    "        if self.ref_is_hap and not (self.target_is_hap or is_phased):\n",
    "            raise RuntimeError(\"The training set contains haploids. Target set should be phased or haploids.\")\n",
    "\n",
    "        # Merge with reference panel to align variants\n",
    "        self.target_set = test_df.merge(right=self.reference_panel[[\"ID\"]], on='ID', how='right')\n",
    "        if self.target_file_extension == \"vcf\" == self.ref_file_extension:\n",
    "            self.target_set[self.reference_panel.columns[:9]] = self.reference_panel[self.reference_panel.columns[:9]]\n",
    "\n",
    "        self.target_set = self.target_set.astype('str')\n",
    "        missing_value = \".\" if self.target_is_hap else \".|.\" if self.is_phased else \"./.\"\n",
    "        self.target_set.fillna(missing_value, inplace=True)\n",
    "        self.target_set.replace(\"nan\", missing_value, inplace=True)\n",
    "        print(\"Target set assignment done!\")\n",
    "\n",
    "    def __map_hap_2_ind_parent_1(self, x) -> int:\n",
    "        return self.hap_map[x.split('|')[0]]\n",
    "\n",
    "    def __map_hap_2_ind_parent_2(self, x) -> int:\n",
    "        return self.hap_map[x.split('|')[1]]\n",
    "\n",
    "    def __map_haps_2_ind(self, x) -> int:\n",
    "        return self.hap_map[x]\n",
    "\n",
    "    def get_ref_set(self, starting_var_index=0, ending_var_index=0) -> np.ndarray:\n",
    "        if 0 <= starting_var_index < ending_var_index:\n",
    "            data = self.reference_panel.iloc[starting_var_index:ending_var_index, self.ref_sample_value_index - 1:]\n",
    "        else:\n",
    "            data = self.reference_panel.iloc[:, self.ref_sample_value_index - 1:]\n",
    "\n",
    "        if self.is_phased:\n",
    "            is_haps = \"|\" not in data.iloc[0, 0]\n",
    "            if not is_haps:\n",
    "                # diploids to hap vecs\n",
    "                _x = np.empty((data.shape[1] * 2, data.shape[0]), dtype=np.int32)\n",
    "                _x[0::2] = self.map_values_1_vec(data.values.T)\n",
    "                _x[1::2] = self.map_values_2_vec(data.values.T)\n",
    "                return _x\n",
    "            else:\n",
    "                return self.map_haps_to_vec(data.values.T)\n",
    "        else:\n",
    "            return data.replace(self.replacement_dict).values.T.astype(np.int32)\n",
    "\n",
    "    def get_target_set(self, starting_var_index=0, ending_var_index=0) -> np.ndarray:\n",
    "        \"\"\"Get target data for imputation\"\"\"\n",
    "        if 0 <= starting_var_index < ending_var_index:\n",
    "            data = self.target_set.iloc[starting_var_index:ending_var_index, self.target_sample_value_index - 1:]\n",
    "        else:\n",
    "            data = self.target_set.iloc[:, self.target_sample_value_index - 1:]\n",
    "\n",
    "        if self.is_phased:\n",
    "            is_haps = \"|\" not in data.iloc[0, 0]\n",
    "            if not is_haps:\n",
    "                # diploids to hap vecs\n",
    "                _x = np.empty((data.shape[1] * 2, data.shape[0]), dtype=np.int32)\n",
    "                _x[0::2] = self.map_values_1_vec(data.values.T)\n",
    "                _x[1::2] = self.map_values_2_vec(data.values.T)\n",
    "                return _x\n",
    "            else:\n",
    "                return self.map_haps_to_vec(data.values.T)\n",
    "        else:\n",
    "            return data.replace(self.replacement_dict).values.T.astype(np.int32)\n",
    "\n",
    "    def __convert_unphased_probs_to_genotypes(self, allele_probs) -> np.ndarray:\n",
    "        \"\"\"Convert unphased probabilities to genotypes\"\"\"\n",
    "        n_samples, n_variants, n_alleles = allele_probs.shape\n",
    "        genotypes = np.zeros((n_samples, n_variants), dtype=object)\n",
    "\n",
    "        for i in tqdm(range(n_samples)):\n",
    "            for j in range(n_variants):\n",
    "                unphased_probs = allele_probs[i, j]\n",
    "                variant_genotypes = np.vectorize(self.reverse_replacement_dict.get)(\n",
    "                    np.argmax(unphased_probs, axis=-1)).flatten()\n",
    "                genotypes[i, j] = variant_genotypes\n",
    "        return genotypes\n",
    "\n",
    "    def __convert_hap_probs_to_diploid_genotypes(self, allele_probs) -> np.ndarray:\n",
    "        \"\"\"Convert haplotype probabilities to diploid genotypes\"\"\"\n",
    "        n_haploids, n_variants, n_alleles = allele_probs.shape\n",
    "\n",
    "        if n_haploids % 2 != 0:\n",
    "            raise ValueError(\"Number of haploids should be even.\")\n",
    "\n",
    "        n_samples = n_haploids // 2\n",
    "        genotypes = np.empty((n_samples, n_variants), dtype=object)\n",
    "        haploids_as_diploids = allele_probs.reshape((n_samples, 2, n_variants, -1))\n",
    "        variant_genotypes = self.map_preds_2_allele(np.argmax(haploids_as_diploids, axis=-1))\n",
    "\n",
    "        def process_variant_in_sample(haps_for_sample_at_variant, variant_genotypes_for_sample_at_variant):\n",
    "            if n_alleles > 2:\n",
    "                return '|'.join(variant_genotypes_for_sample_at_variant)\n",
    "            else:\n",
    "                # Output GP (genotype probabilities)\n",
    "                phased_probs = np.outer(haps_for_sample_at_variant[0], haps_for_sample_at_variant[1]).flatten()\n",
    "                unphased_probs = np.array([phased_probs[0], phased_probs[1] + phased_probs[2], phased_probs[-1]])\n",
    "                unphased_probs_str = \",\".join([f\"{v:.6f}\" for v in unphased_probs])\n",
    "                alt_dosage = np.dot(unphased_probs, [0, 1, 2])\n",
    "                return '|'.join(variant_genotypes_for_sample_at_variant) + f\":{unphased_probs_str}:{alt_dosage:.3f}\"\n",
    "\n",
    "        def process_sample(i):\n",
    "            return np.array([\n",
    "                process_variant_in_sample(haploids_as_diploids[i, :, j, :], variant_genotypes[i, :, j])\n",
    "                for j in range(n_variants)\n",
    "            ])\n",
    "\n",
    "        # Parallel processing\n",
    "        genotypes = Parallel(n_jobs=-1)(delayed(process_sample)(i) for i in tqdm(range(n_samples)))\n",
    "        return np.array(genotypes)\n",
    "\n",
    "    def __convert_hap_probs_to_hap_genotypes(self, allele_probs) -> np.ndarray:\n",
    "        \"\"\"Convert hap probabilities to hap genotypes\"\"\"\n",
    "        return np.argmax(allele_probs, axis=1).astype(str)\n",
    "\n",
    "    def __get_headers_for_output(self, contain_probs, chr=22):\n",
    "        \"\"\"Get VCF headers for output file\"\"\"\n",
    "        headers = [\n",
    "            \"##fileformat=VCFv4.2\",\n",
    "            '''##source=BiMamba v1.0.0''',\n",
    "            '''##INFO=<ID=AF,Number=A,Type=Float,Description=\"Estimated Alternate Allele Frequency\">''',\n",
    "            '''##INFO=<ID=MAF,Number=1,Type=Float,Description=\"Estimated Minor Allele Frequency\">''',\n",
    "            '''##INFO=<ID=AVG_CS,Number=1,Type=Float,Description=\"Average Call Score\">''',\n",
    "            '''##INFO=<ID=IMPUTED,Number=0,Type=Flag,Description=\"Marker was imputed\">''',\n",
    "            '''##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">''',\n",
    "        ]\n",
    "        probs_headers = [\n",
    "            '''##FORMAT=<ID=DS,Number=A,Type=Float,Description=\"Estimated Alternate Allele Dosage : [P(0/1)+2*P(1/1)]\">''',\n",
    "            '''##FORMAT=<ID=GP,Number=G,Type=Float,Description=\"Estimated Posterior Probabilities for Genotypes 0/0, 0/1 and 1/1\">'''\n",
    "        ]\n",
    "        if contain_probs:\n",
    "            headers.extend(probs_headers)\n",
    "        return headers\n",
    "\n",
    "    def __convert_genotypes_to_vcf(self, genotypes, pred_format=\"GT:GP:DS\"):\n",
    "        \"\"\"Convert genotypes to VCF format\"\"\"\n",
    "        new_vcf = self.target_set.copy()\n",
    "        new_vcf[new_vcf.columns[self.target_sample_value_index - 1:]] = genotypes\n",
    "        new_vcf[\"FORMAT\"] = pred_format\n",
    "        new_vcf[\"QUAL\"] = \".\"\n",
    "        new_vcf[\"FILTER\"] = \".\"\n",
    "        new_vcf[\"INFO\"] = \"IMPUTED\"\n",
    "        return new_vcf\n",
    "\n",
    "    def preds_to_genotypes(self, predictions: Union[str, np.ndarray]) -> pd.DataFrame:\n",
    "        \"\"\"Convert predictions to genotypes\"\"\"\n",
    "        if isinstance(predictions, str):\n",
    "            preds = np.load(predictions)\n",
    "        else:\n",
    "            preds = predictions\n",
    "\n",
    "        target_df = self.target_set.copy()\n",
    "        if not self.is_phased:\n",
    "            target_df[\n",
    "                target_df.columns[self.target_sample_value_index - 1:]] = self.__convert_unphased_probs_to_genotypes(\n",
    "                preds).T\n",
    "        elif self.target_is_hap:\n",
    "            target_df[\n",
    "                target_df.columns[self.target_sample_value_index - 1:]] = self.__convert_hap_probs_to_hap_genotypes(\n",
    "                preds).T\n",
    "        else:\n",
    "            pred_format = \"GT:GP:DS\" if preds.shape[-1] == 2 else \"GT\"\n",
    "            target_df = self.__convert_genotypes_to_vcf(self.__convert_hap_probs_to_diploid_genotypes(preds).T,\n",
    "                                                        pred_format)\n",
    "        return target_df\n",
    "\n",
    "    def write_ligated_results_to_file(self, df: pd.DataFrame, file_name: str, compress=True) -> str:\n",
    "        \"\"\"Write results to file\"\"\"\n",
    "        to_write_format = self.ref_file_extension\n",
    "        file_path = f\"{file_name}.{to_write_format}.gz\" if compress else f\"{file_name}.{to_write_format}\"\n",
    "\n",
    "        with gzip.open(file_path, 'wt') if compress else open(file_path, 'wt') as f_out:\n",
    "            # Write headers\n",
    "            if self.ref_file_extension == \"vcf\":\n",
    "                f_out.write(\n",
    "                    \"\\n\".join(self.__get_headers_for_output(contain_probs=\"GP\" in df[\"FORMAT\"].values[0])) + \"\\n\")\n",
    "            else:\n",
    "                f_out.write(\"\\n\".join(self.ref_n_header_lines))\n",
    "\n",
    "        # Append data\n",
    "        df.to_csv(file_path, sep=self.ref_separator, mode='a', index=False)\n",
    "        return file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fbb96506",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GenomicDataset(Dataset):\n",
    "    \"\"\"Dataset class for genomic data with masking for training\"\"\"\n",
    "\n",
    "    def __init__(self, data, targets, seq_depth, offset_before=0, offset_after=0,\n",
    "                 training=True, masking_rates=(0.5, 0.99)):\n",
    "        self.data = data\n",
    "        self.targets = targets\n",
    "        self.seq_depth = seq_depth\n",
    "        self.offset_before = offset_before\n",
    "        self.offset_after = offset_after\n",
    "        self.training = training\n",
    "        self.masking_rates = masking_rates\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.data)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        x = self.data[idx].copy()\n",
    "        y = self.targets[idx]\n",
    "\n",
    "        if self.training:\n",
    "            # Apply masking\n",
    "            seq_len = len(x)\n",
    "            masking_rate = np.random.uniform(*self.masking_rates)\n",
    "            mask_size = int(seq_len * masking_rate)\n",
    "            mask_indices = np.random.choice(seq_len, mask_size, replace=False)\n",
    "            x[mask_indices] = self.seq_depth - 1  # Missing value token\n",
    "\n",
    "        # Convert to one-hot\n",
    "        x_onehot = np.eye(self.seq_depth)[x]\n",
    "        y_onehot = np.eye(self.seq_depth - 1)[y]\n",
    "\n",
    "        return torch.FloatTensor(x_onehot), torch.FloatTensor(y_onehot)\n",
    "\n",
    "class ImputationDataset(Dataset):\n",
    "    \"\"\"Dataset for imputation (no masking needed)\"\"\"\n",
    "\n",
    "    def __init__(self, data, seq_depth):\n",
    "        self.data = data\n",
    "        self.seq_depth = seq_depth\n",
    "\n",
    "    def __len__(self):\n",
    "        return len(self.data)\n",
    "\n",
    "    def __getitem__(self, idx):\n",
    "        x = self.data[idx]\n",
    "        # Convert to one-hot without masking\n",
    "        x_onehot = np.eye(self.seq_depth)[x]\n",
    "        return torch.FloatTensor(x_onehot)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c86e007",
   "metadata": {},
   "source": [
    "## Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5339189",
   "metadata": {},
   "source": [
    "EvoFill 单chunk module 版"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ac21e2b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "class BiMambaBlock(nn.Module):\n",
    "    \"\"\"Bidirectional Mamba block for genomic sequence processing\"\"\"\n",
    "\n",
    "    def __init__(self, d_model, d_state=16, d_conv=4, expand=2):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "\n",
    "        # Forward and backward Mamba blocks\n",
    "        self.mamba_forward = Mamba2(\n",
    "            d_model=d_model,\n",
    "            d_state=d_state,\n",
    "            d_conv=d_conv,\n",
    "            expand=expand\n",
    "        )\n",
    "\n",
    "        self.mamba_backward = Mamba2(\n",
    "            d_model=d_model,\n",
    "            d_state=d_state,\n",
    "            d_conv=d_conv,\n",
    "            expand=expand\n",
    "        )\n",
    "\n",
    "        # Layer normalization\n",
    "        self.norm1 = nn.LayerNorm(d_model)\n",
    "        self.norm2 = nn.LayerNorm(d_model)\n",
    "\n",
    "        # FFN\n",
    "        self.ffn = nn.Sequential(\n",
    "            nn.Linear(d_model * 2, d_model * 4),\n",
    "            nn.GELU(),\n",
    "            nn.Linear(d_model * 4, d_model),\n",
    "            nn.GELU()\n",
    "        )\n",
    "\n",
    "        self.dropout = nn.Dropout(0.1)\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, d_model)\n",
    "        residual = x\n",
    "\n",
    "        # Bidirectional processing\n",
    "        x_norm = self.norm1(x)\n",
    "\n",
    "        # Forward direction\n",
    "        forward_out = self.mamba_forward(x_norm)\n",
    "\n",
    "        # Backward direction (flip sequence)\n",
    "        x_backward = torch.flip(x_norm, dims=[1])\n",
    "        backward_out = self.mamba_backward(x_backward)\n",
    "        backward_out = torch.flip(backward_out, dims=[1])\n",
    "\n",
    "        # Concatenate bidirectional outputs\n",
    "        bi_out = torch.cat([forward_out, backward_out], dim=-1)\n",
    "\n",
    "        # FFN\n",
    "        ffn_out = self.ffn(bi_out)\n",
    "        ffn_out = self.dropout(ffn_out)\n",
    "\n",
    "        # Residual connection\n",
    "        out = self.norm2(residual + ffn_out)\n",
    "\n",
    "        return out\n",
    "\n",
    "class ConvBlock(nn.Module):\n",
    "    \"\"\"Convolutional block for local pattern extraction\"\"\"\n",
    "\n",
    "    def __init__(self, d_model):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "\n",
    "        self.conv1 = nn.Conv1d(d_model, d_model, kernel_size=3, padding=1)\n",
    "        self.conv2 = nn.Conv1d(d_model, d_model, kernel_size=5, padding=2)\n",
    "        self.conv3 = nn.Conv1d(d_model, d_model, kernel_size=7, padding=3)\n",
    "\n",
    "        self.conv_large1 = nn.Conv1d(d_model, d_model, kernel_size=7, padding=3)\n",
    "        self.conv_large2 = nn.Conv1d(d_model, d_model, kernel_size=15, padding=7)\n",
    "\n",
    "        self.conv_final = nn.Conv1d(d_model, d_model, kernel_size=3, padding=1)\n",
    "        self.conv_reduce = nn.Conv1d(d_model, d_model, kernel_size=1)\n",
    "\n",
    "        self.bn1 = nn.BatchNorm1d(d_model)\n",
    "        self.bn2 = nn.BatchNorm1d(d_model)\n",
    "\n",
    "        self.gelu = nn.GELU()\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, d_model)\n",
    "        x = x.transpose(1, 2)  # (batch, d_model, seq_len)\n",
    "\n",
    "        xa = self.gelu(self.conv1(x))\n",
    "\n",
    "        xb = self.gelu(self.conv2(xa))\n",
    "        xb = self.gelu(self.conv3(xb))\n",
    "\n",
    "        xc = self.gelu(self.conv_large1(xa))\n",
    "        xc = self.gelu(self.conv_large2(xc))\n",
    "\n",
    "        xa = xb + xc\n",
    "        xa = self.gelu(self.conv_final(xa))\n",
    "        xa = self.bn1(xa)\n",
    "        xa = self.gelu(self.conv_reduce(xa))\n",
    "        xa = self.bn2(xa)\n",
    "        xa = self.gelu(xa)\n",
    "\n",
    "        return xa.transpose(1, 2)  # (batch, seq_len, d_model)\n",
    "\n",
    "class Mamba2CrossBlock(nn.Module):\n",
    "    \"\"\"\n",
    "    用 Mamba2Simple 替代 MultiheadAttention 的交叉块。\n",
    "    接口与原来 CrossAttentionLayer 保持一致，可直接替换。\n",
    "    \"\"\"\n",
    "    def __init__(\n",
    "        self,\n",
    "        d_model,\n",
    "        d_state=64,\n",
    "        d_conv=4,\n",
    "        expand=2,\n",
    "        headdim=128,\n",
    "        ngroups=1,\n",
    "        chunk_size=256,\n",
    "        dropout=0.0,\n",
    "        device=None,\n",
    "        dtype=None,\n",
    "    ):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "\n",
    "        # 1. 归一化\n",
    "        self.norm1 = nn.LayerNorm(d_model)\n",
    "        self.norm2 = nn.LayerNorm(d_model)\n",
    "        self.norm3 = nn.LayerNorm(d_model)\n",
    "\n",
    "        # 2. Mamba2Simple 作为交叉建模核心\n",
    "        self.ssd = Mamba2Block(\n",
    "            d_model=d_model,\n",
    "            d_state=d_state,\n",
    "            d_conv=d_conv,\n",
    "            expand=expand,\n",
    "            headdim=headdim,\n",
    "            ngroups=ngroups,\n",
    "            chunk_size=chunk_size,\n",
    "            use_mem_eff_path=True,\n",
    "            device=device,\n",
    "            dtype=dtype,\n",
    "        )\n",
    "\n",
    "        # 3. FFN 保持不变\n",
    "        self.ffn = nn.Sequential(\n",
    "            nn.Linear(d_model, d_model // 2),\n",
    "            nn.GELU(),\n",
    "            nn.Linear(d_model // 2, d_model),\n",
    "        )\n",
    "\n",
    "    def forward(self, local_repr, global_repr, start_offset=0, end_offset=0):\n",
    "        \"\"\"\n",
    "        local_repr: [B, L, D]\n",
    "        global_repr: [B, G, D]\n",
    "        输出:       [B, L, D]  （长度与 local_repr 保持一致）\n",
    "        \"\"\"\n",
    "        B, L, D = local_repr.shape\n",
    "\n",
    "        # 1. 归一化\n",
    "        local_norm  = self.norm1(local_repr)\n",
    "        global_norm = self.norm2(global_repr)\n",
    "\n",
    "        # 2. 如果用了 offset，先截断 query 长度\n",
    "        if start_offset > 0 or end_offset > 0:\n",
    "            query = local_norm[:, start_offset:L - end_offset]\n",
    "        else:\n",
    "            query = local_norm\n",
    "\n",
    "        # 3. 拼接：global 在前，local 在后，让 SSD 扫描时 local 能看到 global\n",
    "        x = torch.cat([global_norm, query], dim=1)   # [B, G + L', D]\n",
    "        x = self.ssd(x)                              # [B, G + L', D]\n",
    "        x = x[:, global_norm.shape[1]:, :]           # 只取 local 对应部分 [B, L', D]\n",
    "\n",
    "        # 4. pad 回原始长度（如果之前截断过）\n",
    "        if start_offset > 0 or end_offset > 0:\n",
    "            x = F.pad(x, (0, 0, start_offset, end_offset))  # [B, L, D]\n",
    "\n",
    "        # 5. 残差连接\n",
    "        x = x + local_norm\n",
    "\n",
    "        # 6. FFN\n",
    "        x = self.norm3(x)\n",
    "        x = self.ffn(x) + x\n",
    "\n",
    "        return x\n",
    "\n",
    "class GenoEmbedding(nn.Module):\n",
    "    \"\"\"Genomic embedding layer with positional encoding\"\"\"\n",
    "\n",
    "    def __init__(self, n_alleles, n_snps, d_model):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "        self.n_alleles = n_alleles\n",
    "        self.n_snps = n_snps\n",
    "\n",
    "        # Allele embedding\n",
    "        self.allele_embedding = nn.Parameter(torch.randn(n_alleles, d_model))\n",
    "\n",
    "        # Positional embedding\n",
    "        self.position_embedding = nn.Embedding(n_snps, d_model)\n",
    "\n",
    "        # Initialize parameters\n",
    "        nn.init.xavier_uniform_(self.allele_embedding)\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, n_alleles) - one-hot encoded\n",
    "        batch_size, seq_len, _ = x.shape\n",
    "\n",
    "        # Allele embedding\n",
    "        embedded = torch.einsum('bsn,nd->bsd', x, self.allele_embedding)\n",
    "\n",
    "        # Positional embedding\n",
    "        positions = torch.arange(seq_len, device=x.device)\n",
    "        pos_emb = self.position_embedding(positions).unsqueeze(0)\n",
    "\n",
    "        return embedded + pos_emb\n",
    "\n",
    "class ChunkModule(nn.Module):\n",
    "    \"\"\"Single chunk processing module with BiMamba\"\"\"\n",
    "\n",
    "    def __init__(self, d_model, start_offset=0, end_offset=0, dropout_rate=0.1):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "        self.start_offset = start_offset\n",
    "        self.end_offset = end_offset\n",
    "\n",
    "        # BiMamba block\n",
    "        self.bimamba_block = BiMambaBlock(d_model)\n",
    "\n",
    "        # Convolutional blocks\n",
    "        self.conv_block1 = ConvBlock(d_model)\n",
    "        self.conv_block2 = ConvBlock(d_model)\n",
    "        self.conv_block3 = ConvBlock(d_model)\n",
    "\n",
    "        # Cross attention\n",
    "        # self.cross_attention = CrossAttentionLayer(d_model, n_heads)\n",
    "        self.cross_attention = Mamba2CrossBlock(\n",
    "            d_model=d_model,\n",
    "            d_state=64,\n",
    "            d_conv=4,\n",
    "            expand=2,\n",
    "            headdim=128,\n",
    "            ngroups=1,\n",
    "            chunk_size=256,\n",
    "        )\n",
    "\n",
    "        # Additional layers\n",
    "        self.dense = nn.Linear(d_model, d_model)\n",
    "        self.dropout = nn.Dropout(dropout_rate)\n",
    "        self.gelu = nn.GELU()\n",
    "\n",
    "    def forward(self, x):\n",
    "        # BiMamba processing\n",
    "        xa0 = self.bimamba_block(x)\n",
    "\n",
    "        # First conv block\n",
    "        xa = self.conv_block1(xa0)\n",
    "        xa_skip = self.conv_block2(xa)\n",
    "\n",
    "        # Dense layer\n",
    "        xa = self.gelu(self.dense(xa))\n",
    "        xa = self.conv_block2(xa)\n",
    "\n",
    "        # Cross attention\n",
    "        xa = self.cross_attention(xa, xa0, self.start_offset, self.end_offset)\n",
    "        xa = self.dropout(xa)\n",
    "\n",
    "        # Final conv block\n",
    "        xa = self.conv_block3(xa)\n",
    "\n",
    "        # Concatenate with skip connection\n",
    "        xa = torch.cat([xa_skip, xa], dim=-1)\n",
    "\n",
    "        return xa\n",
    "\n",
    "class EvoFill(nn.Module):\n",
    "    \"\"\"Main BiMamba model for genomic imputation\"\"\"\n",
    "\n",
    "    def __init__(self,\n",
    "                 d_model,\n",
    "                 chunk_size=2048,\n",
    "                 chunk_overlap=64,\n",
    "                 offset_before=0,\n",
    "                 offset_after=0,\n",
    "                 dropout_rate=0.1):\n",
    "        super().__init__()\n",
    "        self.d_model = d_model\n",
    "        self.chunk_size = chunk_size\n",
    "        self.chunk_overlap = chunk_overlap\n",
    "        self.offset_before = offset_before\n",
    "        self.offset_after = offset_after\n",
    "        self.dropout_rate = dropout_rate\n",
    "\n",
    "        # Will be set during build\n",
    "        self.seq_len = None\n",
    "        self.n_alleles = None\n",
    "        self.embedding = None\n",
    "        self.chunk_modules = nn.ModuleList()\n",
    "        self.final_conv = None\n",
    "        self.output_conv = None\n",
    "\n",
    "    def build(self, seq_len, n_alleles):\n",
    "        \"\"\"Build the model with specific sequence length and allele count\"\"\"\n",
    "        self.seq_len = seq_len\n",
    "        self.n_alleles = n_alleles\n",
    "\n",
    "        # Embedding layer\n",
    "        self.embedding = GenoEmbedding(n_alleles, seq_len, self.d_model)\n",
    "\n",
    "        # Calculate chunks\n",
    "        chunk_starts = list(range(0, seq_len, self.chunk_size))\n",
    "        chunk_ends = [min(cs + self.chunk_size, seq_len) for cs in chunk_starts]\n",
    "        mask_starts = [max(0, cs - self.chunk_overlap) for cs in chunk_starts]\n",
    "        mask_ends = [min(ce + self.chunk_overlap, seq_len) for ce in chunk_ends]\n",
    "\n",
    "        # Create chunk modules\n",
    "        self.chunk_module = ChunkModule(\n",
    "            d_model=self.d_model,\n",
    "            start_offset=0,\n",
    "            end_offset=0,\n",
    "            dropout_rate=self.dropout_rate\n",
    "        )\n",
    "\n",
    "        # Store chunk information\n",
    "        self.chunk_starts = chunk_starts\n",
    "        self.chunk_ends = chunk_ends\n",
    "        self.mask_starts = mask_starts\n",
    "        self.mask_ends = mask_ends\n",
    "\n",
    "        # Final layers\n",
    "        self.final_conv = nn.Conv1d(self.d_model * 2, self.d_model // 2,\n",
    "                                    kernel_size=5, padding=2)\n",
    "        self.output_conv = nn.Conv1d(self.d_model // 2, n_alleles - 1,\n",
    "                                     kernel_size=5, padding=2)\n",
    "        # self.output_proj = nn.Linear(self.d_model * 2, n_alleles - 1)\n",
    "\n",
    "        self.gelu = nn.GELU()\n",
    "        self.softmax = nn.Softmax(dim=-1)\n",
    "\n",
    "    def forward(self, x):\n",
    "        # x shape: (batch, seq_len, n_alleles)\n",
    "        if self.embedding is None:\n",
    "            raise RuntimeError(\"Model not built. Call build() first.\")\n",
    "\n",
    "        # Embedding\n",
    "        x_embedded = self.embedding(x)\n",
    "\n",
    "        # Process chunks\n",
    "        chunk_outputs = []\n",
    "        for i in range(len(self.chunk_starts)):\n",
    "            pad_left  = self.chunk_starts[i] - self.mask_starts[i]\n",
    "            pad_right = self.mask_ends[i] - self.chunk_ends[i]\n",
    "            chunk_input = x_embedded[:, self.mask_starts[i]:self.mask_ends[i]]\n",
    "            chunk_output = self.chunk_module(chunk_input)   # 共享权重\n",
    "            if pad_left or pad_right:\n",
    "                chunk_output = F.pad(chunk_output, (0, 0, pad_left, pad_right))\n",
    "            chunk_outputs.append(chunk_output)\n",
    "        # Concatenate chunks along sequence dimension\n",
    "        x_concat = torch.cat(chunk_outputs, dim=1)\n",
    "\n",
    "        # # Final processing\n",
    "        x_concat = x_concat.transpose(1, 2)  # (batch, features, seq_len)\n",
    "        x_final = self.gelu(self.final_conv(x_concat))\n",
    "        x_output = self.output_conv(x_final)\n",
    "        x_output = x_output.transpose(1, 2)  # (batch, seq_len, n_alleles-1)\n",
    "        # x_output = self.output_proj(x_concat) \n",
    "\n",
    "        # Apply offsets\n",
    "        if self.offset_before > 0 or self.offset_after > 0:\n",
    "            x_output = x_output[:, self.offset_before:self.seq_len - self.offset_after]\n",
    "        else:\n",
    "            x_output = x_output[:, :self.seq_len]\n",
    "\n",
    "        x_output = self.softmax(x_output)\n",
    "\n",
    "        return x_output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f053a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_alleles = 4  # 包含missing\n",
    "model = EvoFill(\n",
    "    d_model=256,\n",
    "    chunk_size=5120,\n",
    "    chunk_overlap=64, \n",
    "    offset_before=0,\n",
    "    offset_after=0,\n",
    "    dropout_rate=0.1,\n",
    ").cuda()\n",
    "\n",
    "B, L = 2, 5120\n",
    "model.build(seq_len=L, n_alleles=n_alleles)\n",
    "model = model.cuda()  \n",
    "\n",
    "# 1. 生成输入\n",
    "x = torch.randint(0, n_alleles, (B, L)).long().cuda()   # {0,1,2,3} 3=missing\n",
    "\n",
    "# 2. -1 -> 3，并构造 one-hot（4 维）\n",
    "x_map = x.clone()\n",
    "x_onehot = torch.zeros(B, L, n_alleles, device='cuda')\n",
    "x_onehot.scatter_(2, x_map.unsqueeze(-1), 1)\n",
    "\n",
    "# 3. 前向\n",
    "with torch.no_grad():\n",
    "    probs = model(x_onehot)          # shape: (B,L,3)\n",
    "\n",
    "# 4. 简单校验\n",
    "assert torch.allclose(probs.sum(dim=-1), torch.ones(B, L, device='cuda'), atol=1e-5), \\\n",
    "    \"概率未归一\"\n",
    "print(\"✅ 含缺失数据前向通过，输出形状:\", probs.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ce8c794",
   "metadata": {},
   "source": [
    "## Loss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8f6f2206",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ImputationLoss(nn.Module):\n",
    "    \"\"\"Custom loss function for genomic imputation\"\"\"\n",
    "\n",
    "    def __init__(self, use_r2=True):\n",
    "        super().__init__()\n",
    "        self.use_r2_loss = use_r2\n",
    "        self.ce_loss = nn.CrossEntropyLoss(reduction='sum')\n",
    "        self.kl_loss = nn.KLDivLoss(reduction='sum')\n",
    "\n",
    "    def calculate_minimac_r2(self, pred_alt_allele_probs, gt_alt_af):\n",
    "        \"\"\"Calculate Minimac-style RÂ² metric\"\"\"\n",
    "        mask = torch.logical_or(torch.eq(gt_alt_af, 0.0), torch.eq(gt_alt_af, 1.0))\n",
    "        gt_alt_af = torch.where(mask, 0.5, gt_alt_af)\n",
    "        denom = gt_alt_af * (1.0 - gt_alt_af)\n",
    "        denom = torch.where(denom < 0.01, 0.01, denom)\n",
    "        r2 = torch.mean(torch.square(pred_alt_allele_probs - gt_alt_af), dim=0) / denom\n",
    "        r2 = torch.where(mask, torch.zeros_like(r2), r2)\n",
    "        return r2\n",
    "\n",
    "    def forward(self, y_pred, y_true):\n",
    "        y_true = y_true.float()\n",
    "\n",
    "        # Convert to proper format for losses\n",
    "        y_true_ce = torch.argmax(y_true, dim=-1)  # For CrossEntropy\n",
    "        y_pred_log = torch.log(y_pred + 1e-8)  # For KL divergence\n",
    "\n",
    "        # Basic losses\n",
    "        ce_loss = self.ce_loss(y_pred.view(-1, y_pred.size(-1)), y_true_ce.view(-1))\n",
    "        kl_loss = self.kl_loss(y_pred_log.view(-1, y_pred.size(-1)),\n",
    "                               y_true.view(-1, y_true.size(-1)))\n",
    "\n",
    "        total_loss = ce_loss + kl_loss\n",
    "\n",
    "        if self.use_r2_loss:\n",
    "            batch_size = y_true.size(0)\n",
    "            group_size = 4\n",
    "            num_full_groups = batch_size // group_size\n",
    "\n",
    "            if num_full_groups > 0:\n",
    "                y_true_grouped = y_true[:num_full_groups * group_size].view(\n",
    "                    num_full_groups, group_size, *y_true.shape[1:])\n",
    "                y_pred_grouped = y_pred[:num_full_groups * group_size].view(\n",
    "                    num_full_groups, group_size, *y_pred.shape[1:])\n",
    "\n",
    "                r2_loss = 0.0\n",
    "                for i in range(num_full_groups):\n",
    "                    gt_alt_af = torch.count_nonzero(\n",
    "                        torch.argmax(y_true_grouped[i], dim=-1), dim=0\n",
    "                    ).float() / group_size\n",
    "\n",
    "                    pred_alt_allele_probs = torch.sum(y_pred_grouped[i][:, :, 1:], dim=-1)\n",
    "                    r2_loss += -torch.sum(self.calculate_minimac_r2(\n",
    "                        pred_alt_allele_probs, gt_alt_af)) * group_size\n",
    "\n",
    "                total_loss += r2_loss\n",
    "\n",
    "        return total_loss, None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0d0ff63",
   "metadata": {},
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bb648d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_similar_rows(array):\n",
    "    \"\"\"Remove duplicate haploids from training set\"\"\"\n",
    "    print(\"Finding duplicate haploids in training set.\")\n",
    "    unique_array = np.unique(array, axis=0)\n",
    "    print(f\"Removed {len(array) - len(unique_array)} rows. {len(unique_array)} training samples remaining.\")\n",
    "    return unique_array\n",
    "\n",
    "def create_directories(save_dir, models_dir=\"models\", outputs=\"out\") -> None:\n",
    "    \"\"\"Create necessary directories\"\"\"\n",
    "    for dd in [save_dir, f\"{save_dir}/{models_dir}\", f\"{save_dir}/{outputs}\"]:\n",
    "        if not os.path.exists(dd):\n",
    "            os.makedirs(dd)\n",
    "\n",
    "def clear_dir(path) -> None:\n",
    "    \"\"\"Clear directory if it exists\"\"\"\n",
    "    if os.path.exists(path):\n",
    "        shutil.rmtree(path)\n",
    "\n",
    "def load_chunk_info(save_dir, break_points):\n",
    "    \"\"\"Load chunk training status information\"\"\"\n",
    "    chunk_info = {ww: False for ww in list(range(len(break_points) - 1))}\n",
    "    if os.path.isfile(f\"{save_dir}/models/chunks_info.json\"):\n",
    "        with open(f\"{save_dir}/models/chunks_info.json\", 'r') as f:\n",
    "            loaded_chunks_info = json.load(f)\n",
    "            if isinstance(loaded_chunks_info, dict) and len(loaded_chunks_info) == len(chunk_info):\n",
    "                print(\"Resuming the training...\")\n",
    "                chunk_info = {int(k): v for k, v in loaded_chunks_info.items()}\n",
    "    return chunk_info\n",
    "\n",
    "def save_chunk_status(save_dir, chunk_info) -> None:\n",
    "    \"\"\"Save chunk training status information\"\"\"\n",
    "    with open(f\"{save_dir}/models/chunks_info.json\", \"w\") as outfile:\n",
    "        json.dump(chunk_info, outfile)\n",
    "\n",
    "def create_model(args, seq_len, n_alleles):\n",
    "    \"\"\"Create BiMamba model\"\"\"\n",
    "    model = EvoFill(\n",
    "        d_model=args.embed_dim,\n",
    "        chunk_size=args.cs,\n",
    "        chunk_overlap=args.co,\n",
    "        offset_before=getattr(args, 'offset_before', 0),\n",
    "        offset_after=getattr(args, 'offset_after', 0),\n",
    "        dropout_rate=0.1\n",
    "    )\n",
    "\n",
    "    # Build the model\n",
    "    model.build(seq_len, n_alleles)\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7c302bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "MAF_BINS = [(0.00, 0.05), (0.05, 0.10), (0.10, 0.20),\n",
    "            (0.20, 0.30), (0.30, 0.40), (0.40, 0.50)]\n",
    "\n",
    "def precompute_maf(gts_np, mask_int=-1):\n",
    "    \"\"\"\n",
    "    gts_np: (N, L)  int64\n",
    "    return:\n",
    "        maf: (L,) float32\n",
    "        bin_cnt: list[int] 长度 6，对应 6 个 bin 的位点数量\n",
    "    \"\"\"\n",
    "    L = gts_np.shape[1]\n",
    "    maf = np.zeros(L, dtype=np.float32)\n",
    "    bin_cnt = [0] * 6\n",
    "\n",
    "    for l in range(L):\n",
    "        alleles = gts_np[:, l]\n",
    "        alleles = alleles[alleles != mask_int]   # 去掉缺失\n",
    "        if alleles.size == 0:\n",
    "            maf[l] = 0.0\n",
    "            continue\n",
    "\n",
    "        uniq, cnt = np.unique(alleles, return_counts=True)\n",
    "        total = cnt.sum()\n",
    "        freq = cnt / total\n",
    "        freq[::-1].sort()\n",
    "        maf_val = freq[1] if len(freq) > 1 else 0.0\n",
    "        maf[l] = maf_val\n",
    "\n",
    "        # 统计 bin\n",
    "        for i, (lo, hi) in enumerate(MAF_BINS):\n",
    "            if lo <= maf_val < hi:\n",
    "                bin_cnt[i] += 1\n",
    "                break\n",
    "\n",
    "    return maf, bin_cnt\n",
    "\n",
    "def imputation_maf_accuracy_epoch(all_logits, all_gts, global_maf, mask=None):\n",
    "    \"\"\"\n",
    "    all_logits: (N, L, C)\n",
    "    all_gts:    (N, L, C) one-hot\n",
    "    global_maf: (L,)\n",
    "    mask:       (N, L) 或 None\n",
    "    return:     list[float] 长度 6\n",
    "    \"\"\"\n",
    "    # 1. 预测 vs 真实\n",
    "    all_gts = all_gts.argmax(dim=-1)      # (N, L)\n",
    "    preds   = all_logits.argmax(dim=-1)   # (N, L)\n",
    "\n",
    "    # 2. 如果没有外部 mask，就默认全 1\n",
    "    if mask is None:\n",
    "        mask = torch.ones_like(all_gts, dtype=torch.bool)   # (N, L)\n",
    "    correct = (preds == all_gts) & mask                   # (N, L)\n",
    "\n",
    "    # 3. MAF 条件 -> (1, L) 再广播到 (N, L)\n",
    "    maf = global_maf.unsqueeze(0)                         # (1, L)\n",
    "\n",
    "    # 4. 分 bin 计算\n",
    "    accs = []\n",
    "    for lo, hi in MAF_BINS:\n",
    "        maf_bin = mask & (maf >= lo) & (maf < hi)                # (1, L)\n",
    "        n_cor = (correct & maf_bin).sum()\n",
    "        n_tot = maf_bin.sum()\n",
    "        accs.append(100*(n_cor / n_tot).item() if n_tot > 0 else 0.0)\n",
    "    return accs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d21fb7c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using device: cuda\n",
      "Reading the file...\n",
      "2404 diploid samples with 99314 variants found!\n",
      "DEBUG: Unique genotypes in dataset: ['0|0' '0|1' '1|0' '1|1']...\n",
      "DEBUG: self.genotype_vals: ['0|0' '0|1' '1|0' '1|1']\n",
      "DEBUG: self.alleles: ['0' '1']\n",
      "DEBUG: is_phased: True\n",
      "DEBUG: hap_map: {'0': 0, '1': 1, '.': 2}\n",
      "DEBUG: self.SEQ_DEPTH: 3\n"
     ]
    }
   ],
   "source": [
    "# ---------------- 以下即命令行参数对应的行内变量 ----------------\n",
    "mode                 = 'train'\n",
    "restart_training     = True          # 对应命令行 1\n",
    "ref                  = \"/home/qmtang/GitHub/STICI-HPC/data/training_sets/ALL.chr22.training.samples.100k.any.type.0.01.maf.variants.vcf.gz\"\n",
    "tihp                 = True          # 对应命令行 1\n",
    "which_chunk          = -1            # All chunkss\n",
    "save_dir             = '/home/qmtang/mnt_qmtang/EvoFill/data/251016_ver1_chr22'\n",
    "co                   = 64            # 64 in STICI\n",
    "cs                   = 8192          # 2048 in STICI\n",
    "sites_per_model      = 65536        # 10240 in STICI\n",
    "max_mr               = 0.7\n",
    "min_mr               = 0.3\n",
    "epochs               = 100\n",
    "embed_dim            = 64            # 128 in STICI\n",
    "lr                   = 0.001\n",
    "weight_decay         = 1e-5\n",
    "batch_size_per_gpu   = 8\n",
    "use_r2               = True\n",
    "earlystop_patience   = 9\n",
    "verbose              = 1\n",
    "# -------------------------------------------------------------\n",
    "\n",
    "# 组装成 Namespace\n",
    "args = Namespace(\n",
    "    restart_training=restart_training,\n",
    "    ref=ref,\n",
    "    tihp=tihp,\n",
    "    which_chunk=which_chunk,\n",
    "    save_dir=save_dir,\n",
    "    co=co,\n",
    "    cs=cs,\n",
    "    sites_per_model=sites_per_model,\n",
    "    max_mr=max_mr,\n",
    "    min_mr=min_mr,\n",
    "    epochs=epochs,\n",
    "    embed_dim=embed_dim,\n",
    "    lr=lr,\n",
    "    weight_decay=weight_decay,\n",
    "    batch_size_per_gpu=batch_size_per_gpu,\n",
    "    use_r2=use_r2,\n",
    "    earlystop_patience=earlystop_patience,\n",
    "    verbose=verbose,\n",
    ")\n",
    "\n",
    "assert args.max_mr > 0\n",
    "assert args.min_mr > 0\n",
    "assert args.max_mr >= args.min_mr\n",
    "\n",
    "# Setup device\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "print(f\"Using device: {device}\")\n",
    "\n",
    "# Create directories\n",
    "create_directories(args.save_dir)\n",
    "with open(f\"{args.save_dir}/commandline_args.json\", 'w') as f:\n",
    "    json.dump(vars(args), f, indent=4)\n",
    "\n",
    "# Load data\n",
    "dr = DataReader()\n",
    "dr.assign_training_set(\n",
    "    file_path=args.ref,\n",
    "    target_is_gonna_be_phased_or_haps=args.tihp,\n",
    "    variants_as_columns=getattr(args, 'ref_vac', False),\n",
    "    delimiter=getattr(args, 'ref_sep', None),\n",
    "    file_format=getattr(args, 'ref_file_format', 'infer'),\n",
    "    first_column_is_index=getattr(args, 'ref_fcai', True),\n",
    "    comments=getattr(args, 'ref_comment', '##')\n",
    ")\n",
    "\n",
    "# Split data for validation\n",
    "n_samples = dr.get_ref_set(0, 1).shape[0]\n",
    "val_n_samples = args.batch_size_per_gpu * getattr(args, 'val_n_batches', 8)\n",
    "x_train_indices, x_valid_indices = train_test_split(\n",
    "    range(n_samples),\n",
    "    test_size=val_n_samples,\n",
    "    random_state=getattr(args, 'random_seed', 2022),\n",
    "    shuffle=True\n",
    ")\n",
    "\n",
    "# Process chunks\n",
    "break_points = list(np.arange(0, dr.VARIANT_COUNT, args.sites_per_model)) + [dr.VARIANT_COUNT]\n",
    "chunks_done = load_chunk_info(args.save_dir, break_points)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "006172b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training on chunk 1/2\n",
      "Data shape: (4808, 65664)\n",
      "Finding duplicate haploids in training set.\n",
      "Removed 0 rows. 4744 training samples remaining.\n",
      "Chunk MAF-bin counts: [26838, 9346, 10401, 6907, 6055, 5985]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 1/100, Train Loss: 339426.2052, Val Loss: 368955.4102\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.35 96.89\n",
      "(0.05, 0.10)   9346 91.73 90.44\n",
      "(0.10, 0.20)  10401 83.83 81.62\n",
      "(0.20, 0.30)   6907 73.70 70.98\n",
      "(0.30, 0.40)   6055 64.19 63.86\n",
      "(0.40, 0.50)   5985 58.47 60.71\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 2/100, Train Loss: 279713.9254, Val Loss: 337015.7656\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.60 97.00\n",
      "(0.05, 0.10)   9346 92.60 90.86\n",
      "(0.10, 0.20)  10401 86.01 82.82\n",
      "(0.20, 0.30)   6907 78.22 73.36\n",
      "(0.30, 0.40)   6055 71.86 68.38\n",
      "(0.40, 0.50)   5985 68.45 65.33\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 3/100, Train Loss: 248956.8245, Val Loss: 316151.4648\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.62 96.94\n",
      "(0.05, 0.10)   9346 92.90 91.22\n",
      "(0.10, 0.20)  10401 87.08 84.17\n",
      "(0.20, 0.30)   6907 80.18 75.72\n",
      "(0.30, 0.40)   6055 76.02 71.68\n",
      "(0.40, 0.50)   5985 73.42 69.66\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 4/100, Train Loss: 224431.5869, Val Loss: 272011.7539\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.64 97.20\n",
      "(0.05, 0.10)   9346 93.25 92.12\n",
      "(0.10, 0.20)  10401 88.13 86.42\n",
      "(0.20, 0.30)   6907 81.85 79.32\n",
      "(0.30, 0.40)   6055 78.77 76.13\n",
      "(0.40, 0.50)   5985 76.97 74.31\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 5/100, Train Loss: 204788.9342, Val Loss: 273938.7637\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.68 97.13\n",
      "(0.05, 0.10)   9346 93.53 92.00\n",
      "(0.10, 0.20)  10401 88.86 86.02\n",
      "(0.20, 0.30)   6907 83.42 79.05\n",
      "(0.30, 0.40)   6055 80.82 76.12\n",
      "(0.40, 0.50)   5985 79.37 74.23\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 6/100, Train Loss: 189482.4041, Val Loss: 236391.7871\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.72 97.30\n",
      "(0.05, 0.10)   9346 93.80 92.79\n",
      "(0.10, 0.20)  10401 89.40 87.79\n",
      "(0.20, 0.30)   6907 84.58 81.76\n",
      "(0.30, 0.40)   6055 82.47 79.74\n",
      "(0.40, 0.50)   5985 81.34 78.30\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 7/100, Train Loss: 177348.4993, Val Loss: 238603.1445\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.74 97.40\n",
      "(0.05, 0.10)   9346 94.00 92.88\n",
      "(0.10, 0.20)  10401 89.93 87.75\n",
      "(0.20, 0.30)   6907 85.48 81.56\n",
      "(0.30, 0.40)   6055 83.60 79.68\n",
      "(0.40, 0.50)   5985 82.72 77.99\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 8/100, Train Loss: 168539.5522, Val Loss: 224983.6191\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.77 97.41\n",
      "(0.05, 0.10)   9346 94.15 93.11\n",
      "(0.10, 0.20)  10401 90.39 88.39\n",
      "(0.20, 0.30)   6907 86.13 82.53\n",
      "(0.30, 0.40)   6055 84.36 80.64\n",
      "(0.40, 0.50)   5985 83.59 79.37\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 9/100, Train Loss: 160097.9035, Val Loss: 232167.8203\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.81 97.41\n",
      "(0.05, 0.10)   9346 94.33 92.93\n",
      "(0.10, 0.20)  10401 90.84 87.78\n",
      "(0.20, 0.30)   6907 86.70 81.69\n",
      "(0.30, 0.40)   6055 85.05 80.00\n",
      "(0.40, 0.50)   5985 84.31 78.42\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 10/100, Train Loss: 152783.1892, Val Loss: 209739.0527\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.84 97.41\n",
      "(0.05, 0.10)   9346 94.50 93.34\n",
      "(0.10, 0.20)  10401 91.19 89.06\n",
      "(0.20, 0.30)   6907 87.22 83.79\n",
      "(0.30, 0.40)   6055 85.84 82.29\n",
      "(0.40, 0.50)   5985 84.90 80.71\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 11/100, Train Loss: 147575.9761, Val Loss: 249177.6113\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.86 97.32\n",
      "(0.05, 0.10)   9346 94.63 92.48\n",
      "(0.10, 0.20)  10401 91.46 86.70\n",
      "(0.20, 0.30)   6907 87.63 80.39\n",
      "(0.30, 0.40)   6055 86.26 78.10\n",
      "(0.40, 0.50)   5985 85.36 76.28\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 12/100, Train Loss: 142840.9102, Val Loss: 233011.5039\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.89 97.34\n",
      "(0.05, 0.10)   9346 94.77 92.81\n",
      "(0.10, 0.20)  10401 91.68 87.62\n",
      "(0.20, 0.30)   6907 87.92 81.33\n",
      "(0.30, 0.40)   6055 86.63 79.37\n",
      "(0.40, 0.50)   5985 85.73 77.35\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 13/100, Train Loss: 138894.2188, Val Loss: 190904.9922\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.92 97.42\n",
      "(0.05, 0.10)   9346 94.88 93.63\n",
      "(0.10, 0.20)  10401 91.87 89.84\n",
      "(0.20, 0.30)   6907 88.21 85.04\n",
      "(0.30, 0.40)   6055 86.98 83.97\n",
      "(0.40, 0.50)   5985 86.07 82.16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 14/100, Train Loss: 133692.4189, Val Loss: 185425.2168\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.95 97.54\n",
      "(0.05, 0.10)   9346 95.00 93.99\n",
      "(0.10, 0.20)  10401 92.08 90.28\n",
      "(0.20, 0.30)   6907 88.59 85.78\n",
      "(0.30, 0.40)   6055 87.43 84.68\n",
      "(0.40, 0.50)   5985 86.48 82.81\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 15/100, Train Loss: 130406.2657, Val Loss: 200093.9980\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.97 97.50\n",
      "(0.05, 0.10)   9346 95.11 93.66\n",
      "(0.10, 0.20)  10401 92.25 89.28\n",
      "(0.20, 0.30)   6907 88.88 84.21\n",
      "(0.30, 0.40)   6055 87.71 82.94\n",
      "(0.40, 0.50)   5985 86.74 80.99\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 16/100, Train Loss: 126612.3739, Val Loss: 182315.8887\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.98 97.44\n",
      "(0.05, 0.10)   9346 95.18 93.90\n",
      "(0.10, 0.20)  10401 92.39 90.24\n",
      "(0.20, 0.30)   6907 89.13 86.06\n",
      "(0.30, 0.40)   6055 88.02 84.79\n",
      "(0.40, 0.50)   5985 86.97 82.82\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 17/100, Train Loss: 123726.1458, Val Loss: 204204.2891\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 97.99 97.52\n",
      "(0.05, 0.10)   9346 95.25 93.62\n",
      "(0.10, 0.20)  10401 92.52 88.96\n",
      "(0.20, 0.30)   6907 89.36 83.87\n",
      "(0.30, 0.40)   6055 88.28 82.24\n",
      "(0.40, 0.50)   5985 87.18 80.14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 18/100, Train Loss: 120595.3088, Val Loss: 188890.4395\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.03 97.61\n",
      "(0.05, 0.10)   9346 95.34 94.04\n",
      "(0.10, 0.20)  10401 92.68 89.91\n",
      "(0.20, 0.30)   6907 89.58 85.38\n",
      "(0.30, 0.40)   6055 88.50 84.00\n",
      "(0.40, 0.50)   5985 87.42 81.66\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 19/100, Train Loss: 118276.2873, Val Loss: 154788.9395\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.05 97.75\n",
      "(0.05, 0.10)   9346 95.44 94.82\n",
      "(0.10, 0.20)  10401 92.79 91.76\n",
      "(0.20, 0.30)   6907 89.76 88.17\n",
      "(0.30, 0.40)   6055 88.74 87.30\n",
      "(0.40, 0.50)   5985 87.66 85.51\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 20/100, Train Loss: 115945.4667, Val Loss: 167458.7852\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.07 97.62\n",
      "(0.05, 0.10)   9346 95.49 94.43\n",
      "(0.10, 0.20)  10401 92.91 90.97\n",
      "(0.20, 0.30)   6907 89.90 87.00\n",
      "(0.30, 0.40)   6055 88.89 85.60\n",
      "(0.40, 0.50)   5985 87.83 83.85\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 21/100, Train Loss: 113404.6327, Val Loss: 167513.3535\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.08 97.67\n",
      "(0.05, 0.10)   9346 95.56 94.36\n",
      "(0.10, 0.20)  10401 93.02 90.82\n",
      "(0.20, 0.30)   6907 90.06 86.53\n",
      "(0.30, 0.40)   6055 89.07 85.35\n",
      "(0.40, 0.50)   5985 88.00 83.54\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 22/100, Train Loss: 111348.5079, Val Loss: 154173.7676\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.09 97.67\n",
      "(0.05, 0.10)   9346 95.63 94.73\n",
      "(0.10, 0.20)  10401 93.12 91.69\n",
      "(0.20, 0.30)   6907 90.18 87.94\n",
      "(0.30, 0.40)   6055 89.22 86.87\n",
      "(0.40, 0.50)   5985 88.18 85.05\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 23/100, Train Loss: 109729.5622, Val Loss: 152968.6055\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.13 97.71\n",
      "(0.05, 0.10)   9346 95.67 94.80\n",
      "(0.10, 0.20)  10401 93.21 91.65\n",
      "(0.20, 0.30)   6907 90.29 87.91\n",
      "(0.30, 0.40)   6055 89.34 86.99\n",
      "(0.40, 0.50)   5985 88.29 85.14\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 24/100, Train Loss: 107562.6879, Val Loss: 167042.7617\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.14 97.69\n",
      "(0.05, 0.10)   9346 95.72 94.52\n",
      "(0.10, 0.20)  10401 93.30 90.92\n",
      "(0.20, 0.30)   6907 90.39 86.74\n",
      "(0.30, 0.40)   6055 89.47 85.78\n",
      "(0.40, 0.50)   5985 88.42 83.71\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 25/100, Train Loss: 106226.6729, Val Loss: 180832.3750\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.14 97.63\n",
      "(0.05, 0.10)   9346 95.76 94.18\n",
      "(0.10, 0.20)  10401 93.38 90.33\n",
      "(0.20, 0.30)   6907 90.50 85.58\n",
      "(0.30, 0.40)   6055 89.60 84.26\n",
      "(0.40, 0.50)   5985 88.55 81.97\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 26/100, Train Loss: 104480.5815, Val Loss: 153568.8066\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.16 97.78\n",
      "(0.05, 0.10)   9346 95.81 94.82\n",
      "(0.10, 0.20)  10401 93.46 91.65\n",
      "(0.20, 0.30)   6907 90.61 87.96\n",
      "(0.30, 0.40)   6055 89.75 86.79\n",
      "(0.40, 0.50)   5985 88.66 85.10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 27/100, Train Loss: 102703.5255, Val Loss: 150043.7695\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.19 97.70\n",
      "(0.05, 0.10)   9346 95.87 94.82\n",
      "(0.10, 0.20)  10401 93.56 91.83\n",
      "(0.20, 0.30)   6907 90.73 88.25\n",
      "(0.30, 0.40)   6055 89.86 87.28\n",
      "(0.40, 0.50)   5985 88.79 85.28\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 28/100, Train Loss: 101272.4403, Val Loss: 148596.0273\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.18 97.81\n",
      "(0.05, 0.10)   9346 95.89 94.90\n",
      "(0.10, 0.20)  10401 93.61 91.82\n",
      "(0.20, 0.30)   6907 90.81 88.21\n",
      "(0.30, 0.40)   6055 89.95 87.33\n",
      "(0.40, 0.50)   5985 88.88 85.65\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 29/100, Train Loss: 99887.8950, Val Loss: 126362.6816\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.22 97.90\n",
      "(0.05, 0.10)   9346 95.95 95.46\n",
      "(0.10, 0.20)  10401 93.69 93.12\n",
      "(0.20, 0.30)   6907 90.96 90.11\n",
      "(0.30, 0.40)   6055 90.10 89.27\n",
      "(0.40, 0.50)   5985 89.01 87.84\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 30/100, Train Loss: 97807.7073, Val Loss: 141345.9746\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.22 97.86\n",
      "(0.05, 0.10)   9346 96.00 95.06\n",
      "(0.10, 0.20)  10401 93.77 92.26\n",
      "(0.20, 0.30)   6907 91.07 88.86\n",
      "(0.30, 0.40)   6055 90.22 87.86\n",
      "(0.40, 0.50)   5985 89.12 86.21\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 31/100, Train Loss: 96325.6957, Val Loss: 149700.8613\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.25 97.72\n",
      "(0.05, 0.10)   9346 96.03 94.82\n",
      "(0.10, 0.20)  10401 93.83 91.76\n",
      "(0.20, 0.30)   6907 91.18 88.07\n",
      "(0.30, 0.40)   6055 90.34 86.86\n",
      "(0.40, 0.50)   5985 89.23 85.09\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 32/100, Train Loss: 95586.0966, Val Loss: 128979.4434\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.26 97.94\n",
      "(0.05, 0.10)   9346 96.07 95.43\n",
      "(0.10, 0.20)  10401 93.89 92.84\n",
      "(0.20, 0.30)   6907 91.25 89.62\n",
      "(0.30, 0.40)   6055 90.42 88.71\n",
      "(0.40, 0.50)   5985 89.32 87.05\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 33/100, Train Loss: 94317.2808, Val Loss: 122556.2734\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.27 97.91\n",
      "(0.05, 0.10)   9346 96.09 95.60\n",
      "(0.10, 0.20)  10401 93.95 93.05\n",
      "(0.20, 0.30)   6907 91.32 90.25\n",
      "(0.30, 0.40)   6055 90.51 89.37\n",
      "(0.40, 0.50)   5985 89.39 87.85\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 34/100, Train Loss: 92839.7512, Val Loss: 120566.7793\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.28 97.97\n",
      "(0.05, 0.10)   9346 96.13 95.62\n",
      "(0.10, 0.20)  10401 93.99 93.30\n",
      "(0.20, 0.30)   6907 91.38 90.34\n",
      "(0.30, 0.40)   6055 90.56 89.52\n",
      "(0.40, 0.50)   5985 89.48 88.12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 35/100, Train Loss: 92405.5730, Val Loss: 138317.0645\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.29 97.88\n",
      "(0.05, 0.10)   9346 96.15 95.26\n",
      "(0.10, 0.20)  10401 94.05 92.30\n",
      "(0.20, 0.30)   6907 91.42 88.85\n",
      "(0.30, 0.40)   6055 90.61 87.74\n",
      "(0.40, 0.50)   5985 89.51 86.15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 36/100, Train Loss: 90882.0131, Val Loss: 119383.2090\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.32 97.99\n",
      "(0.05, 0.10)   9346 96.22 95.76\n",
      "(0.10, 0.20)  10401 94.12 93.41\n",
      "(0.20, 0.30)   6907 91.53 90.35\n",
      "(0.30, 0.40)   6055 90.70 89.53\n",
      "(0.40, 0.50)   5985 89.64 88.12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 37/100, Train Loss: 89961.0740, Val Loss: 139849.4355\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.32 97.93\n",
      "(0.05, 0.10)   9346 96.22 95.31\n",
      "(0.10, 0.20)  10401 94.16 92.30\n",
      "(0.20, 0.30)   6907 91.58 88.75\n",
      "(0.30, 0.40)   6055 90.75 87.75\n",
      "(0.40, 0.50)   5985 89.68 85.99\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 38/100, Train Loss: 88921.3577, Val Loss: 111896.9316\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.33 98.03\n",
      "(0.05, 0.10)   9346 96.25 95.78\n",
      "(0.10, 0.20)  10401 94.20 93.66\n",
      "(0.20, 0.30)   6907 91.64 90.93\n",
      "(0.30, 0.40)   6055 90.82 90.15\n",
      "(0.40, 0.50)   5985 89.76 88.58\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 39/100, Train Loss: 87949.1339, Val Loss: 117046.5527\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.35 98.06\n",
      "(0.05, 0.10)   9346 96.29 95.81\n",
      "(0.10, 0.20)  10401 94.26 93.50\n",
      "(0.20, 0.30)   6907 91.74 90.60\n",
      "(0.30, 0.40)   6055 90.89 89.71\n",
      "(0.40, 0.50)   5985 89.82 88.33\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 40/100, Train Loss: 86901.5311, Val Loss: 128758.3809\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.36 97.93\n",
      "(0.05, 0.10)   9346 96.32 95.50\n",
      "(0.10, 0.20)  10401 94.31 92.82\n",
      "(0.20, 0.30)   6907 91.80 89.53\n",
      "(0.30, 0.40)   6055 90.94 88.53\n",
      "(0.40, 0.50)   5985 89.90 86.77\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 41/100, Train Loss: 86250.4485, Val Loss: 105352.1934\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.37 98.11\n",
      "(0.05, 0.10)   9346 96.36 96.09\n",
      "(0.10, 0.20)  10401 94.37 94.14\n",
      "(0.20, 0.30)   6907 91.85 91.46\n",
      "(0.30, 0.40)   6055 91.01 90.74\n",
      "(0.40, 0.50)   5985 89.96 89.33\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 42/100, Train Loss: 84988.1208, Val Loss: 118022.1367\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.38 98.02\n",
      "(0.05, 0.10)   9346 96.37 95.74\n",
      "(0.10, 0.20)  10401 94.41 93.44\n",
      "(0.20, 0.30)   6907 91.90 90.40\n",
      "(0.30, 0.40)   6055 91.06 89.48\n",
      "(0.40, 0.50)   5985 90.00 87.94\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 43/100, Train Loss: 84225.6233, Val Loss: 111071.7656\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.40 98.17\n",
      "(0.05, 0.10)   9346 96.40 95.95\n",
      "(0.10, 0.20)  10401 94.45 93.77\n",
      "(0.20, 0.30)   6907 91.95 90.96\n",
      "(0.30, 0.40)   6055 91.09 90.05\n",
      "(0.40, 0.50)   5985 90.07 88.56\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 44/100, Train Loss: 83700.3693, Val Loss: 116804.1953\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.40 98.05\n",
      "(0.05, 0.10)   9346 96.42 95.78\n",
      "(0.10, 0.20)  10401 94.49 93.52\n",
      "(0.20, 0.30)   6907 91.98 90.47\n",
      "(0.30, 0.40)   6055 91.12 89.51\n",
      "(0.40, 0.50)   5985 90.09 87.78\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 45/100, Train Loss: 82540.2240, Val Loss: 134918.5273\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.41 97.90\n",
      "(0.05, 0.10)   9346 96.45 95.28\n",
      "(0.10, 0.20)  10401 94.54 92.44\n",
      "(0.20, 0.30)   6907 92.05 89.04\n",
      "(0.30, 0.40)   6055 91.19 87.89\n",
      "(0.40, 0.50)   5985 90.16 86.09\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 46/100, Train Loss: 78592.5146, Val Loss: 113713.7734\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.49 98.08\n",
      "(0.05, 0.10)   9346 96.60 95.93\n",
      "(0.10, 0.20)  10401 94.76 93.60\n",
      "(0.20, 0.30)   6907 92.30 90.65\n",
      "(0.30, 0.40)   6055 91.45 89.71\n",
      "(0.40, 0.50)   5985 90.43 88.18\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 47/100, Train Loss: 77983.6057, Val Loss: 112906.5879\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.49 98.08\n",
      "(0.05, 0.10)   9346 96.62 95.92\n",
      "(0.10, 0.20)  10401 94.77 93.57\n",
      "(0.20, 0.30)   6907 92.32 90.67\n",
      "(0.30, 0.40)   6055 91.49 89.62\n",
      "(0.40, 0.50)   5985 90.47 88.10\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 48/100, Train Loss: 77422.9461, Val Loss: 108816.5723\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.50 98.14\n",
      "(0.05, 0.10)   9346 96.63 96.05\n",
      "(0.10, 0.20)  10401 94.78 93.80\n",
      "(0.20, 0.30)   6907 92.34 90.90\n",
      "(0.30, 0.40)   6055 91.51 89.98\n",
      "(0.40, 0.50)   5985 90.51 88.40\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 49/100, Train Loss: 76738.4916, Val Loss: 117968.4707\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.51 98.06\n",
      "(0.05, 0.10)   9346 96.66 95.72\n",
      "(0.10, 0.20)  10401 94.83 93.25\n",
      "(0.20, 0.30)   6907 92.38 90.22\n",
      "(0.30, 0.40)   6055 91.52 89.08\n",
      "(0.40, 0.50)   5985 90.53 87.41\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 1, Epoch 50/100, Train Loss: 74885.7914, Val Loss: 116146.0938\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  26838 98.54 98.12\n",
      "(0.05, 0.10)   9346 96.73 95.87\n",
      "(0.10, 0.20)  10401 94.92 93.43\n",
      "(0.20, 0.30)   6907 92.51 90.40\n",
      "(0.30, 0.40)   6055 91.64 89.41\n",
      "(0.40, 0.50)   5985 90.68 87.69\n",
      "Early stopping triggered\n",
      "Training on chunk 2/2\n",
      "Data shape: (4808, 33906)\n",
      "Finding duplicate haploids in training set.\n",
      "Removed 0 rows. 4744 training samples remaining.\n",
      "Chunk MAF-bin counts: [15310, 4771, 5026, 3324, 2778, 2568]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 1/100, Train Loss: 141885.6191, Val Loss: 188965.8633\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 97.48 96.71\n",
      "(0.05, 0.10)   4771 92.73 91.18\n",
      "(0.10, 0.20)   5026 86.25 82.64\n",
      "(0.20, 0.30)   3324 77.74 72.44\n",
      "(0.30, 0.40)   2778 70.16 64.40\n",
      "(0.40, 0.50)   2568 67.72 61.90\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 2/100, Train Loss: 95721.4525, Val Loss: 181615.1172\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 97.73 96.72\n",
      "(0.05, 0.10)   4771 93.79 91.39\n",
      "(0.10, 0.20)   5026 90.27 83.78\n",
      "(0.20, 0.30)   3324 84.90 74.19\n",
      "(0.30, 0.40)   2778 82.21 67.45\n",
      "(0.40, 0.50)   2568 81.55 64.65\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 3/100, Train Loss: 82304.9804, Val Loss: 161071.6406\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 97.82 96.87\n",
      "(0.05, 0.10)   4771 94.29 92.23\n",
      "(0.10, 0.20)   5026 91.62 85.76\n",
      "(0.20, 0.30)   3324 87.13 77.52\n",
      "(0.30, 0.40)   2778 85.14 71.98\n",
      "(0.40, 0.50)   2568 84.36 70.20\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 4/100, Train Loss: 74322.7191, Val Loss: 138523.2480\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 97.90 97.04\n",
      "(0.05, 0.10)   4771 94.82 93.02\n",
      "(0.10, 0.20)   5026 92.42 87.40\n",
      "(0.20, 0.30)   3324 88.42 80.09\n",
      "(0.30, 0.40)   2778 86.66 75.64\n",
      "(0.40, 0.50)   2568 85.98 73.50\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 5/100, Train Loss: 67720.9967, Val Loss: 120361.6475\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 97.96 97.19\n",
      "(0.05, 0.10)   4771 95.26 93.87\n",
      "(0.10, 0.20)   5026 92.99 89.18\n",
      "(0.20, 0.30)   3324 89.56 83.00\n",
      "(0.30, 0.40)   2778 87.90 79.88\n",
      "(0.40, 0.50)   2568 87.21 76.47\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 6/100, Train Loss: 62392.0000, Val Loss: 134846.6270\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.02 97.02\n",
      "(0.05, 0.10)   4771 95.55 93.34\n",
      "(0.10, 0.20)   5026 93.47 87.63\n",
      "(0.20, 0.30)   3324 90.53 81.18\n",
      "(0.30, 0.40)   2778 88.96 76.97\n",
      "(0.40, 0.50)   2568 88.18 74.29\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 7/100, Train Loss: 57963.4783, Val Loss: 136112.6250\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.07 97.06\n",
      "(0.05, 0.10)   4771 95.83 93.36\n",
      "(0.10, 0.20)   5026 93.89 87.92\n",
      "(0.20, 0.30)   3324 91.24 81.23\n",
      "(0.30, 0.40)   2778 89.77 76.83\n",
      "(0.40, 0.50)   2568 89.06 73.76\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 8/100, Train Loss: 54440.4075, Val Loss: 123425.8672\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.11 97.22\n",
      "(0.05, 0.10)   4771 96.06 94.01\n",
      "(0.10, 0.20)   5026 94.25 89.41\n",
      "(0.20, 0.30)   3324 91.82 83.46\n",
      "(0.30, 0.40)   2778 90.35 79.89\n",
      "(0.40, 0.50)   2568 89.65 77.15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                            \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 9/100, Train Loss: 51333.7401, Val Loss: 118775.8867\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.17 97.27\n",
      "(0.05, 0.10)   4771 96.25 94.21\n",
      "(0.10, 0.20)   5026 94.52 89.51\n",
      "(0.20, 0.30)   3324 92.30 84.00\n",
      "(0.30, 0.40)   2778 90.89 80.74\n",
      "(0.40, 0.50)   2568 90.21 77.65\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 10/100, Train Loss: 48865.3718, Val Loss: 110070.4639\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.21 97.36\n",
      "(0.05, 0.10)   4771 96.41 94.67\n",
      "(0.10, 0.20)   5026 94.75 90.31\n",
      "(0.20, 0.30)   3324 92.70 85.26\n",
      "(0.30, 0.40)   2778 91.31 82.10\n",
      "(0.40, 0.50)   2568 90.69 79.19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 11/100, Train Loss: 46754.4411, Val Loss: 113185.3701\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.25 97.38\n",
      "(0.05, 0.10)   4771 96.52 94.65\n",
      "(0.10, 0.20)   5026 94.90 90.35\n",
      "(0.20, 0.30)   3324 92.98 84.45\n",
      "(0.30, 0.40)   2778 91.64 81.15\n",
      "(0.40, 0.50)   2568 91.05 79.11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 12/100, Train Loss: 44841.8714, Val Loss: 120365.6562\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.30 97.29\n",
      "(0.05, 0.10)   4771 96.64 94.26\n",
      "(0.10, 0.20)   5026 95.06 89.64\n",
      "(0.20, 0.30)   3324 93.24 83.77\n",
      "(0.30, 0.40)   2778 91.97 80.15\n",
      "(0.40, 0.50)   2568 91.45 77.84\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 13/100, Train Loss: 43126.6244, Val Loss: 119793.2891\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.34 97.32\n",
      "(0.05, 0.10)   4771 96.76 94.41\n",
      "(0.10, 0.20)   5026 95.23 89.83\n",
      "(0.20, 0.30)   3324 93.53 84.18\n",
      "(0.30, 0.40)   2778 92.26 80.14\n",
      "(0.40, 0.50)   2568 91.80 78.27\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 14/100, Train Loss: 41553.4054, Val Loss: 111555.4004\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.38 97.41\n",
      "(0.05, 0.10)   4771 96.86 94.68\n",
      "(0.10, 0.20)   5026 95.38 90.33\n",
      "(0.20, 0.30)   3324 93.71 85.44\n",
      "(0.30, 0.40)   2778 92.47 82.10\n",
      "(0.40, 0.50)   2568 92.02 79.90\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 15/100, Train Loss: 37707.3298, Val Loss: 94291.7959\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.50 97.58\n",
      "(0.05, 0.10)   4771 97.13 95.42\n",
      "(0.10, 0.20)   5026 95.75 91.65\n",
      "(0.20, 0.30)   3324 94.21 87.48\n",
      "(0.30, 0.40)   2778 93.03 84.29\n",
      "(0.40, 0.50)   2568 92.60 82.30\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 16/100, Train Loss: 36732.2677, Val Loss: 94551.0986\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.52 97.60\n",
      "(0.05, 0.10)   4771 97.17 95.34\n",
      "(0.10, 0.20)   5026 95.82 91.57\n",
      "(0.20, 0.30)   3324 94.32 87.05\n",
      "(0.30, 0.40)   2778 93.13 84.21\n",
      "(0.40, 0.50)   2568 92.74 82.35\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 17/100, Train Loss: 35941.9555, Val Loss: 96412.6211\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.54 97.59\n",
      "(0.05, 0.10)   4771 97.21 95.29\n",
      "(0.10, 0.20)   5026 95.89 91.62\n",
      "(0.20, 0.30)   3324 94.40 87.24\n",
      "(0.30, 0.40)   2778 93.21 84.31\n",
      "(0.40, 0.50)   2568 92.83 82.19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 18/100, Train Loss: 35075.7530, Val Loss: 91357.5391\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.56 97.66\n",
      "(0.05, 0.10)   4771 97.26 95.52\n",
      "(0.10, 0.20)   5026 95.96 91.97\n",
      "(0.20, 0.30)   3324 94.51 87.76\n",
      "(0.30, 0.40)   2778 93.32 85.09\n",
      "(0.40, 0.50)   2568 92.94 83.19\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 19/100, Train Loss: 34599.9255, Val Loss: 100997.5117\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.58 97.56\n",
      "(0.05, 0.10)   4771 97.28 95.07\n",
      "(0.10, 0.20)   5026 96.02 91.16\n",
      "(0.20, 0.30)   3324 94.57 86.67\n",
      "(0.30, 0.40)   2778 93.40 83.48\n",
      "(0.40, 0.50)   2568 93.03 81.38\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 20/100, Train Loss: 33819.5711, Val Loss: 113357.4404\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.60 97.51\n",
      "(0.05, 0.10)   4771 97.34 94.85\n",
      "(0.10, 0.20)   5026 96.09 90.38\n",
      "(0.20, 0.30)   3324 94.67 85.42\n",
      "(0.30, 0.40)   2778 93.48 81.38\n",
      "(0.40, 0.50)   2568 93.14 79.89\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 21/100, Train Loss: 33275.5576, Val Loss: 100851.0684\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.62 97.59\n",
      "(0.05, 0.10)   4771 97.37 95.23\n",
      "(0.10, 0.20)   5026 96.15 91.38\n",
      "(0.20, 0.30)   3324 94.73 87.05\n",
      "(0.30, 0.40)   2778 93.55 83.60\n",
      "(0.40, 0.50)   2568 93.24 81.55\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 22/100, Train Loss: 32664.3652, Val Loss: 100908.1074\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.63 97.61\n",
      "(0.05, 0.10)   4771 97.42 95.23\n",
      "(0.10, 0.20)   5026 96.23 91.36\n",
      "(0.20, 0.30)   3324 94.82 86.75\n",
      "(0.30, 0.40)   2778 93.66 83.70\n",
      "(0.40, 0.50)   2568 93.34 82.00\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 23/100, Train Loss: 30756.9461, Val Loss: 97878.2607\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.70 97.66\n",
      "(0.05, 0.10)   4771 97.53 95.37\n",
      "(0.10, 0.20)   5026 96.42 91.68\n",
      "(0.20, 0.30)   3324 95.06 87.04\n",
      "(0.30, 0.40)   2778 93.90 84.11\n",
      "(0.40, 0.50)   2568 93.61 82.33\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 24/100, Train Loss: 30301.8877, Val Loss: 101856.7461\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.71 97.64\n",
      "(0.05, 0.10)   4771 97.57 95.23\n",
      "(0.10, 0.20)   5026 96.48 91.43\n",
      "(0.20, 0.30)   3324 95.09 86.88\n",
      "(0.30, 0.40)   2778 93.94 83.59\n",
      "(0.40, 0.50)   2568 93.68 81.80\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 25/100, Train Loss: 29586.7699, Val Loss: 96576.7344\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.72 97.68\n",
      "(0.05, 0.10)   4771 97.58 95.46\n",
      "(0.10, 0.20)   5026 96.50 91.89\n",
      "(0.20, 0.30)   3324 95.14 87.73\n",
      "(0.30, 0.40)   2778 93.98 84.68\n",
      "(0.40, 0.50)   2568 93.73 82.82\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 26/100, Train Loss: 29557.9002, Val Loss: 109991.4023\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.74 97.59\n",
      "(0.05, 0.10)   4771 97.60 95.12\n",
      "(0.10, 0.20)   5026 96.53 90.94\n",
      "(0.20, 0.30)   3324 95.18 86.15\n",
      "(0.30, 0.40)   2778 94.05 82.66\n",
      "(0.40, 0.50)   2568 93.78 80.91\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                             \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chunk 2, Epoch 27/100, Train Loss: 28346.5804, Val Loss: 96585.2510\n",
      "     MAF_bin Counts Train   Val\n",
      "(0.00, 0.05)  15310 98.76 97.67\n",
      "(0.05, 0.10)   4771 97.69 95.42\n",
      "(0.10, 0.20)   5026 96.65 91.83\n",
      "(0.20, 0.30)   3324 95.31 87.47\n",
      "(0.30, 0.40)   2778 94.17 84.33\n",
      "(0.40, 0.50)   2568 93.90 82.60\n",
      "Early stopping triggered\n"
     ]
    }
   ],
   "source": [
    "for w in range(len(break_points) - 1):\n",
    "    if chunks_done[w]:\n",
    "        print(f\"Skipping chunk {w + 1}/{len(break_points) - 1} due to previous training.\")\n",
    "        continue\n",
    "    if args.which_chunk != -1 and w + 1 != args.which_chunk:\n",
    "        print(f\"Skipping chunk {w + 1}/{len(break_points) - 1} due to your request using --which-chunk.\")\n",
    "        continue\n",
    "\n",
    "    print(f\"Training on chunk {w + 1}/{len(break_points) - 1}\")\n",
    "\n",
    "    # Calculate chunk boundaries\n",
    "    final_start_pos = max(0, break_points[w] - 2 * args.co)\n",
    "    final_end_pos = min(dr.VARIANT_COUNT, break_points[w + 1] + 2 * args.co)\n",
    "    offset_before = break_points[w] - final_start_pos\n",
    "    offset_after = final_end_pos - break_points[w + 1]\n",
    "\n",
    "    # Get data for this chunk\n",
    "    ref_set = dr.get_ref_set(final_start_pos, final_end_pos).astype(np.int32)\n",
    "    print(f\"Data shape: {ref_set.shape}\")\n",
    "\n",
    "    # Remove duplicates from training\n",
    "    ref_set_train = remove_similar_rows(ref_set[x_train_indices])\n",
    "    ref_set_val = ref_set[x_valid_indices]\n",
    "\n",
    "    # MAF bins counts\n",
    "    valid_slice = slice(offset_before,\n",
    "                        ref_set_train.shape[1] - offset_after)\n",
    "    chunk_maf, chunk_bin_cnt = precompute_maf(\n",
    "        ref_set_train[:, valid_slice], \n",
    "        mask_int=-1\n",
    "    )\n",
    "    chunk_maf = torch.from_numpy(chunk_maf).to(device)          # (L_chunk,)\n",
    "    if args.verbose:\n",
    "        print('Chunk MAF-bin counts:', chunk_bin_cnt)\n",
    "\n",
    "    # Create targets (same as input for reconstruction)\n",
    "    target_train = ref_set_train[:, offset_before:ref_set_train.shape[1] - offset_after]\n",
    "    target_val = ref_set_val[:, offset_before:ref_set_val.shape[1] - offset_after]\n",
    "\n",
    "    # Create datasets\n",
    "    train_dataset = GenomicDataset(\n",
    "        ref_set_train, target_train, dr.SEQ_DEPTH,\n",
    "        offset_before, offset_after, training=True,\n",
    "        masking_rates=(args.min_mr, args.max_mr)\n",
    "    )\n",
    "\n",
    "    val_dataset = GenomicDataset(\n",
    "        ref_set_val, target_val, dr.SEQ_DEPTH,\n",
    "        offset_before, offset_after, training=False,\n",
    "        masking_rates=(args.min_mr, args.max_mr)\n",
    "    )\n",
    "\n",
    "    # Create data loaders\n",
    "    train_loader = DataLoader(train_dataset, batch_size=args.batch_size_per_gpu,\n",
    "                                shuffle=True, num_workers=4, pin_memory=True)\n",
    "    val_loader = DataLoader(val_dataset, batch_size=args.batch_size_per_gpu,\n",
    "                            shuffle=False, num_workers=4, pin_memory=True)\n",
    "\n",
    "    # Create model\n",
    "    seq_len = ref_set.shape[1]\n",
    "    model = create_model(args, seq_len, dr.SEQ_DEPTH)\n",
    "    model.offset_before = offset_before\n",
    "    model.offset_after = offset_after\n",
    "    model.to(device)\n",
    "\n",
    "    # Loss and optimizer\n",
    "    criterion = ImputationLoss(use_r2=getattr(args, 'use_r2', True))\n",
    "    optimizer = torch.optim.AdamW(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)\n",
    "    # optimizer = Lamb(model.parameters(), lr=args.lr, weight_decay=args.weight_decay)\n",
    "    \n",
    "    scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(\n",
    "        optimizer, mode='min', factor=0.5, patience=3, min_lr=1e-7\n",
    "    )\n",
    "\n",
    "    # Training loop\n",
    "    best_loss = float('inf')\n",
    "    patience = args.earlystop_patience\n",
    "    patience_counter = 0\n",
    "\n",
    "    for epoch in range(args.epochs):\n",
    "        model.train()\n",
    "        train_loss = 0.0\n",
    "        train_logits, train_gts, train_mask = [], [], []\n",
    "\n",
    "        train_pbar = tqdm(train_loader, desc=f'Epoch {epoch + 1}/{args.epochs}', leave=False)\n",
    "        for batch_idx, (data, target) in enumerate(train_pbar):\n",
    "            data, target = data.to(device), target.to(device)\n",
    "\n",
    "            optimizer.zero_grad()\n",
    "            output = model(data)\n",
    "            loss, logs = criterion(output, target)\n",
    "            loss.backward()\n",
    "            optimizer.step()\n",
    "\n",
    "            train_loss += loss.item()\n",
    "            train_pbar.set_postfix({'loss': loss.item()})\n",
    "\n",
    "            # === 收集训练结果 ===\n",
    "            mask = data[..., -1].bool()         # 只关心被 mask 的位点\n",
    "            train_logits.append(output.detach())\n",
    "            train_gts.append(target.detach())\n",
    "            train_mask.append(mask)\n",
    "\n",
    "        # 训练集 MAF-acc\n",
    "        train_logits = torch.cat(train_logits, dim=0)\n",
    "        train_gts    = torch.cat(train_gts,    dim=0)\n",
    "        train_mask   = torch.cat(train_mask,   dim=0)\n",
    "        # 只保留有效位点（去掉 offset  padding）\n",
    "        if model.offset_before > 0 or model.offset_after > 0:\n",
    "            train_mask   = train_mask  [:, model.offset_before : train_mask.shape[1]  -model.offset_after]\n",
    "        train_maf_accs = imputation_maf_accuracy_epoch(train_logits, train_gts, chunk_maf, mask=train_mask)\n",
    "\n",
    "        # ----------- 验证循环同理 ------------\n",
    "        model.eval()\n",
    "        val_loss = 0.0\n",
    "        val_logits, val_gts = [], []\n",
    "        with torch.no_grad():\n",
    "            for data, target in val_loader:\n",
    "                data, target = data.to(device), target.to(device)\n",
    "                output = model(data)\n",
    "                loss, logs = criterion(output, target)\n",
    "                val_loss += loss.item()\n",
    "\n",
    "                mask = data[..., -1].bool()\n",
    "                val_logits.append(output)\n",
    "                val_gts.append(target)\n",
    "\n",
    "        val_logits = torch.cat(val_logits, dim=0)\n",
    "        val_gts    = torch.cat(val_gts,    dim=0)\n",
    "        val_maf_accs = imputation_maf_accuracy_epoch(\n",
    "            val_logits, val_gts, chunk_maf,  mask=None,)\n",
    "\n",
    "        avg_train_loss = train_loss / len(train_loader)\n",
    "        avg_val_loss   = val_loss   / len(val_loader)\n",
    "\n",
    "        if args.verbose >= 1:\n",
    "            avg_train_loss = train_loss / len(train_loader)\n",
    "            avg_val_loss   = val_loss   / len(val_loader)\n",
    "            print(f'Chunk {w + 1}, Epoch {epoch + 1}/{args.epochs}, '\n",
    "                  f'Train Loss: {avg_train_loss:.4f}, Val Loss: {avg_val_loss:.4f}')\n",
    "\n",
    "            # 用 DataFrame 打印 MAF-bin 结果\n",
    "            maf_df = pd.DataFrame({\n",
    "                'MAF_bin': ['(0.00, 0.05)', '(0.05, 0.10)', '(0.10, 0.20)',\n",
    "                            '(0.20, 0.30)', '(0.30, 0.40)', '(0.40, 0.50)'],\n",
    "                'Counts':  [f\"{c}\" for c in chunk_bin_cnt],\n",
    "                'Train':   [f\"{acc:.2f}\" for acc in train_maf_accs],\n",
    "                'Val':     [f\"{acc:.2f}\" for acc in val_maf_accs]\n",
    "            })\n",
    "            print(maf_df.to_string(index=False))\n",
    "\n",
    "        scheduler.step(avg_val_loss)\n",
    "\n",
    "        # Early stopping\n",
    "        if avg_val_loss < best_loss:\n",
    "            best_loss = avg_val_loss\n",
    "            patience_counter = 0\n",
    "            # Save best model\n",
    "            import os\n",
    "            os.makedirs(f'{args.save_dir}/models', exist_ok=True)\n",
    "            torch.save(model.state_dict(), f'{args.save_dir}/models/w_{w}.pth')\n",
    "        else:\n",
    "            patience_counter += 1\n",
    "            if patience_counter >= patience:\n",
    "                if args.verbose >= 1:\n",
    "                    print('Early stopping triggered')\n",
    "                break\n",
    "\n",
    "    # Mark chunk as done\n",
    "    chunks_done[w] = True\n",
    "    save_chunk_status(args.save_dir, chunks_done)\n",
    "\n",
    "    # Clean up\n",
    "    del ref_set, train_dataset, val_dataset, train_loader, val_loader, model\n",
    "    torch.cuda.empty_cache() if torch.cuda.is_available() else None"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b4b1c54",
   "metadata": {},
   "source": [
    "复制以上输出到文件，然后执行"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d1e2457",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\"\"\"\n",
    "把训练日志中的 chunk/epoch/loss 以及 MAF 六区间统计\n",
    "整理成一张宽表（csv 或 DataFrame）。\n",
    "用法：\n",
    "    python parse_log.py train.log  ->  在当前目录生成 train.log.csv\n",
    "    也可以在脚本里直接改 LOG_PATH。\n",
    "\"\"\"\n",
    "\n",
    "# --------------------- 需要改的地方 ---------------------\n",
    "LOG_PATH = '/home/qmtang/mnt_qmtang/EvoFill/logs/251005.logs'          # 日志路径\n",
    "OUT_PATH = None                 # 输出 csv 路径，None 则自动生成\n",
    "# ------------------------------------------------------\n",
    "\n",
    "if OUT_PATH is None:\n",
    "    OUT_PATH = Path(LOG_PATH).with_suffix('.csv')\n",
    "\n",
    "# 正则：抓取 Chunk x, Epoch y/100, Train Loss: ..., Val Loss: ...\n",
    "epoch_re = re.compile(\n",
    "    r'Chunk\\s+(\\d+),\\s+Epoch\\s+(\\d+)/\\d+,\\s+'\n",
    "    r'Train Loss:\\s+([\\d\\.]+),\\s+Val Loss:\\s+([\\d\\.]+)'\n",
    ")\n",
    "\n",
    "# 正则：抓取 MAF 表格里每一行  (0.00, 0.05)   7421 0.974 0.997\n",
    "maf_re = re.compile(\n",
    "    r'\\(\\d+\\.\\d+,\\s*\\d+\\.\\d+\\)\\s+(\\d+)\\s+([\\d\\.]+)\\s+([\\d\\.]+)'\n",
    ")\n",
    "\n",
    "records = []\n",
    "current = {}\n",
    "\n",
    "with open(LOG_PATH, encoding='utf-8') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "\n",
    "        # 遇到新的 epoch 行\n",
    "        m = epoch_re.search(line)\n",
    "        if m:\n",
    "            # 如果上一个 epoch 已经抓完 MAF，就保存\n",
    "            if current and 'maf_0_counts' in current:\n",
    "                records.append(current)\n",
    "\n",
    "            current = {\n",
    "                'chunk': int(m.group(1)),\n",
    "                'epoch': int(m.group(2)),\n",
    "                'train_loss': float(m.group(3)),\n",
    "                'val_loss': float(m.group(4))\n",
    "            }\n",
    "            continue\n",
    "\n",
    "        # 在 MAF 表格区域内\n",
    "        if current:\n",
    "            m = maf_re.search(line)\n",
    "            if m:\n",
    "                idx = len([k for k in current.keys() if k.startswith('maf_')]) // 3\n",
    "                current[f'maf_{idx}_counts'] = int(m.group(1))\n",
    "                current[f'maf_{idx}_train'] = float(m.group(2))\n",
    "                current[f'maf_{idx}_val'] = float(m.group(3))\n",
    "\n",
    "# 别忘了最后一条\n",
    "if current and 'maf_0_counts' in current:\n",
    "    records.append(current)\n",
    "\n",
    "# 拼表\n",
    "df = pd.DataFrame(records)\n",
    "\n",
    "# 让列顺序好看一点\n",
    "cols = ['chunk', 'epoch', 'train_loss', 'val_loss']\n",
    "maf_cols = sorted([c for c in df.columns if c.startswith('maf_')])\n",
    "df = df[cols + maf_cols]\n",
    "\n",
    "# 保存 & 打印\n",
    "df.to_csv(OUT_PATH, index=False, float_format='%.6f')\n",
    "print(f'已解析 {len(df)} 条记录，保存为 {OUT_PATH}')\n",
    "print(df.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b400ce5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# 1. 读取日志\n",
    "logs_set = {\n",
    "    \"n·(MB+CA)\": \"/home/qmtang/mnt_qmtang/EvoFill/logs/251003.csv\",\n",
    "    \"n·(MB+MB)\": \"/home/qmtang/mnt_qmtang/EvoFill/logs/251004.csv\",\n",
    "    \"1·(MB+MB)\": \"/home/qmtang/mnt_qmtang/EvoFill/logs/251005.csv\"\n",
    "}\n",
    "\n",
    "dfs = {}\n",
    "for m, p in logs_set.items():\n",
    "    dfs[m] = pd.read_csv(p)\n",
    "\n",
    "# 2. 统一列名检查（可选）\n",
    "# for m, df in dfs.items():\n",
    "#     print(m, df.columns)\n",
    "\n",
    "# 3. 画图：train_loss\n",
    "n_chunk = 7\n",
    "fig, axes = plt.subplots(2, 4, figsize=(16, 9))\n",
    "axes = axes.flatten()\n",
    "for idx, chunk in enumerate(range(1, n_chunk + 1)):\n",
    "    ax = axes[idx]\n",
    "    for m in dfs:\n",
    "        df = dfs[m]\n",
    "        df_c = df[df['chunk'] == chunk]\n",
    "        sns.lineplot(data=df_c, x='epoch', y='train_loss', label=m, ax=ax, marker='o')\n",
    "    ax.set_title(f'Chunk {chunk} – train_loss')\n",
    "    ax.set_xlabel('Epoch')\n",
    "    ax.set_ylabel('train_loss')\n",
    "    ax.legend()\n",
    "# 隐藏多余子图\n",
    "for j in range(n_chunk, len(axes)):\n",
    "    axes[j].set_visible(False)\n",
    "plt.suptitle('Train Loss across Chunks', fontsize=16)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# 4. 画图：val_loss\n",
    "fig, axes = plt.subplots(2, 4, figsize=(16, 9))\n",
    "axes = axes.flatten()\n",
    "for idx, chunk in enumerate(range(1, n_chunk + 1)):\n",
    "    ax = axes[idx]\n",
    "    for m in dfs:\n",
    "        df = dfs[m]\n",
    "        df_c = df[df['chunk'] == chunk]\n",
    "        sns.lineplot(data=df_c, x='epoch', y='val_loss', label=m, ax=ax, marker='o')\n",
    "    ax.set_title(f'Chunk {chunk} – val_loss')\n",
    "    ax.set_xlabel('Epoch')\n",
    "    ax.set_ylabel('val_loss')\n",
    "    ax.legend()\n",
    "for j in range(n_chunk, len(axes)):\n",
    "    axes[j].set_visible(False)\n",
    "plt.suptitle('Validation Loss across Chunks', fontsize=16)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ba06a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "logs_set = {\n",
    "    \"n·(MB+CA)\": \"/home/qmtang/mnt_qmtang/EvoFill/logs/251003.csv\",\n",
    "    \"n·(MB+MB)\": \"/home/qmtang/mnt_qmtang/EvoFill/logs/251004.csv\",\n",
    "    \"1·(MB+MB)\": \"/home/qmtang/mnt_qmtang/EvoFill/logs/251005.csv\"\n",
    "}\n",
    "\n",
    "dfs = {m: pd.read_csv(p) for m, p in logs_set.items()}\n",
    "\n",
    "# 2. 每个模型每个 chunk 取最大 epoch 的一行\n",
    "best_rows = []\n",
    "for m, df in dfs.items():\n",
    "    best = df.loc[df.groupby('chunk')['epoch'].idxmax()].copy()\n",
    "    best['model'] = m\n",
    "    best_rows.append(best)\n",
    "best_df = pd.concat(best_rows, ignore_index=True)\n",
    "\n",
    "maf_cols = [f'maf_{i}_{s}' for i in range(6) for s in ['train', 'val']]\n",
    "best_df.loc[best_df['model'] == \"1·(MB+MB)\", maf_cols] /= 100\n",
    "\n",
    "maf_labels = [\n",
    "    'MAF(0.00,0.05)',\n",
    "    'MAF(0.05,0.10)',\n",
    "    'MAF(0.10,0.20)',\n",
    "    'MAF(0.20,0.30)',\n",
    "    'MAF(0.30,0.40)',\n",
    "    'MAF(0.40,0.50)'\n",
    "]\n",
    "\n",
    "# ---------- maf_0~5_train ----------\n",
    "train_cols = [f'maf_{i}_train' for i in range(6)]\n",
    "fig, axes = plt.subplots(2, 3, figsize=(15, 8))\n",
    "axes = axes.flatten()\n",
    "for i, col in enumerate(train_cols):\n",
    "    sns.barplot(data=best_df, x='chunk', y=col, hue='model', ax=axes[i])\n",
    "    axes[i].set_ylim(0.75, 1.0)\n",
    "    axes[i].set_ylabel('Accuracy')\n",
    "    axes[i].set_title(maf_labels[i])\n",
    "plt.suptitle('Best Epoch – Training Accuracy', fontsize=16)\n",
    "plt.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "# ---------- maf_0~5_val ----------\n",
    "val_cols = [f'maf_{i}_val' for i in range(6)]\n",
    "fig, axes = plt.subplots(2, 3, figsize=(15, 8))\n",
    "axes = axes.flatten()\n",
    "for i, col in enumerate(val_cols):\n",
    "    sns.barplot(data=best_df, x='chunk', y=col, hue='model', ax=axes[i])\n",
    "    axes[i].set_ylim(0.75, 1.0)\n",
    "    axes[i].set_ylabel('Accuracy')\n",
    "    axes[i].set_title(maf_labels[i])\n",
    "plt.suptitle('Best Epoch – Validation Accuracy', fontsize=16)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
