{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "928e399c",
   "metadata": {},
   "source": [
    "# EvoFill working demo\n",
    "\n",
    "ver 4.   new imputation loss with evo loss\n",
    "\n",
    "ver 4.1  long range modules integrated in stage1 training\n",
    "\n",
    "ver 4.2  stage 3 fine tuning with under-reprensted population samples.\n",
    "\n",
    "last update: 2025/11/13"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed77817",
   "metadata": {},
   "source": [
    "## 0. Dependency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "23548ca5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python ver:   3.10.19 | packaged by conda-forge | (main, Oct 13 2025, 14:08:27) [GCC 14.3.0]\n",
      "Pytorch ver:  2.8.0+cu129\n",
      "Mamba ver:    2.2.5\n",
      "GPU in use:   NVIDIA H100 PCIe\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"1\" \n",
    "\n",
    "import sys\n",
    "import json\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from tqdm import tqdm\n",
    "import torch\n",
    "import mamba_ssm\n",
    "\n",
    "print('Python ver:  ', sys.version)\n",
    "print('Pytorch ver: ', torch.__version__)\n",
    "print('Mamba ver:   ', mamba_ssm.__version__)\n",
    "print('GPU in use:  ', torch.cuda.get_device_name(torch.cuda.current_device()))\n",
    "\n",
    "os.chdir('/mnt/qmtang/EvoFill/')\n",
    "from src.data import GenotypeEncoder, ImputationDataset\n",
    "from src.model import EvoFill\n",
    "from src.utils import setup_workdir, precompute_maf, metrics_by_maf_with95ci, print_maf_stat_df_with95ci"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb7064af",
   "metadata": {},
   "source": [
    "## 1. Encoding all vcfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1c4b1516",
   "metadata": {},
   "outputs": [],
   "source": [
    "work_dir = Path('/mnt/qmtang/EvoFill_data/20251211_chr22/')\n",
    "os.chdir(work_dir)\n",
    "setup_workdir(work_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29a0a57f",
   "metadata": {},
   "source": [
    "### 1.1 1kGP cohorts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d332f86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DATA] 总计 190,184 个位点  \n",
      "[DATA] EvoMat shape: (3009, 3009)\n",
      "[DATA] 结果已写入 /mnt/qmtang/EvoFill_data/20251211_chr22/train\n",
      "[DATA] 位点矩阵 = (3009, 190184)，稀疏度 = 26.30%\n",
      "[DATA] 位点字典 = {'0|1': 1, '0|0': 0, '1|1': 2, '.|.': -1}，字典深度 = 4\n",
      "[DATA] 3,009 Samples\n",
      "[DATA] 190,184 Variants Sites\n",
      "[DATA] 4 seq_depth\n"
     ]
    }
   ],
   "source": [
    "gt_enc = GenotypeEncoder(phased = False, gts012 = False, save2disk = True, save_dir = Path(work_dir / \"train\"))\n",
    "gt_enc = gt_enc.encode_new(vcf_path   = \"data/major_pops_train.vcf.gz\" ,\n",
    "                           evo_mat    = \"data/evo_mat_major_pops_train.tsv\")\n",
    "\n",
    "print(f\"[DATA] {gt_enc.n_samples:,} Samples\")\n",
    "print(f\"[DATA] {gt_enc.n_variants:,} Variants Sites\")\n",
    "print(f\"[DATA] {gt_enc.seq_depth} seq_depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff71b241",
   "metadata": {},
   "source": [
    "### 1.2 aDNA cohorts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "258362bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DATA] 总计 15,517 个位点  \n",
      "[DATA] EvoMat shape: (8953, 8953)\n",
      "[DATA] 结果已写入 /mnt/qmtang/EvoFill_data/20251211_chr22/augment\n",
      "[DATA] 位点矩阵 = (8953, 15517)，稀疏度 = 64.16%，缺失率 = 49.41%\n",
      "[DATA] 位点字典 = {'0|0': 0, '0|1': 1, '1|1': 2, '.|.': -1, '0/0': 0, '0/1': 1, '1/1': 2, './.': -1}，字典深度 = 4\n",
      "[DATA] 8,953 Samples\n",
      "[DATA] 15,517 Variants Sites\n",
      "[DATA] 4 seq_depth\n"
     ]
    }
   ],
   "source": [
    "gt_enc_aug = GenotypeEncoder(phased=False, gts012=False, save2disk = True, save_dir = Path(work_dir / \"augment\"))\n",
    "gt_enc_aug = gt_enc_aug.encode_ref(\n",
    "        ref_meta_json = work_dir/\"train\"/\"gt_enc_meta.json\",   # 与 Stage1 同构\n",
    "        vcf_path      = \"augment/hg38_chr22.AADRl_plus_Shimao.renamed.vcf.gz\",\n",
    "        evo_mat       = \"augment/evo_mat_aDNA.tsv\",)\n",
    "\n",
    "print(f\"[DATA] {gt_enc_aug.n_samples:,} Samples\")\n",
    "print(f\"[DATA] {gt_enc_aug.n_variants:,} Variants Sites\")\n",
    "print(f\"[DATA] {gt_enc_aug.seq_depth} seq_depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf94ad6b",
   "metadata": {},
   "source": [
    "将1240K中的位点映射到千人基因组"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ecc3581",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "古DNA VCF 位点数: 15517\n",
      "在参考 VCF 中命中: 14713\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from cyvcf2 import VCF\n",
    "\n",
    "ref_vcf_path   = \"data/major_pops_train.vcf.gz\"\n",
    "aDNA_vcf_path = 'augment/hg38_chr22.AADRl_plus_Shimao.renamed.vcf.gz'\n",
    "\n",
    "# 1. 大 VCF 建索引  fingerprint -> row_index\n",
    "big_index = {}\n",
    "for idx, variant in enumerate(VCF(ref_vcf_path)):\n",
    "    # 用 CHROM:POS:REF:ALT 当 key，和 bcftools 脚本保持一致\n",
    "    key = f'{variant.CHROM}:{variant.POS}:{variant.REF}:{\",\".join(variant.ALT)}'\n",
    "    big_index[key] = idx          # 0-based 行号\n",
    "\n",
    "# 2. 小 VCF 生成 mapping 数组\n",
    "mapping = []\n",
    "for variant in VCF(aDNA_vcf_path):\n",
    "    key = f'{variant.CHROM}:{variant.POS}:{variant.REF}:{\",\".join(variant.ALT)}'\n",
    "    mapping.append(big_index.get(key, -1))\n",
    "\n",
    "mapping = np.array(mapping, dtype=np.int32)\n",
    "print('古DNA VCF 位点数:', len(mapping))\n",
    "print('在参考面板中命中:', np.sum(mapping != -1))\n",
    "\n",
    "# 保存\n",
    "np.save('augment/aDNA-1kGP_sitesmap.npy', mapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92e19437",
   "metadata": {},
   "source": [
    "### 1.4 validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ae1b2e23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DATA] 总计 190,184 个位点  \n",
      "[DATA] 结果已写入 /mnt/qmtang/EvoFill_data/20251211_chr22/impute_in\n",
      "[DATA] 位点矩阵 = (93, 190184)，稀疏度 = 92.38%，缺失率 = 89.93%\n",
      "[DATA] 位点字典 = {'0|0': 0, '0|1': 1, '1|1': 2, '.|.': -1, '0/0': 0, '0/1': 1, '1/1': 2, './.': -1}，字典深度 = 4\n",
      "[INFER] 93 samples, 190184 variants\n"
     ]
    }
   ],
   "source": [
    "gt_enc_imp = GenotypeEncoder(phased=False, gts012=False, save2disk=True, save_dir = Path(work_dir / \"impute_in\"))\n",
    "gt_enc_imp = gt_enc_imp.encode_ref(\n",
    "        ref_meta_json = work_dir/\"train\"/\"gt_enc_meta.json\",   # 与 Stage1 同构\n",
    "        vcf_path      = \"/mnt/qmtang/EvoFill_data/20251211_chr22/data/minor_pops_all.mask90p.vcf.gz\" )\n",
    "\n",
    "print(f'[INFER] {gt_enc_imp.n_samples} samples, {gt_enc_imp.n_variants} variants')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ebe0aebe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[DATA] 总计 190,184 个位点  \n",
      "[DATA] 结果已写入 /mnt/qmtang/EvoFill_data/20251211_chr22/impute_out\n",
      "[DATA] 位点矩阵 = (93, 190184)，稀疏度 = 24.31%，缺失率 = 0.00%\n",
      "[DATA] 位点字典 = {'0|0': 0, '0|1': 1, '1|1': 2, '.|.': -1, '0/0': 0, '0/1': 1, '1/1': 2, './.': -1}，字典深度 = 4\n",
      "[INFER] 93 samples, 190184 variants\n"
     ]
    }
   ],
   "source": [
    "gt_enc_imp = GenotypeEncoder(phased=False, gts012=False, save2disk=True, save_dir = Path(work_dir / \"impute_out\"))\n",
    "gt_enc_imp = gt_enc_imp.encode_ref(\n",
    "        ref_meta_json = work_dir/\"train\"/\"gt_enc_meta.json\",   # 与 Stage1 同构\n",
    "        vcf_path      = \"/mnt/qmtang/EvoFill_data/20251211_chr22/data/minor_pops_all.vcf.gz\" )\n",
    "\n",
    "print(f'[INFER] {gt_enc_imp.n_samples} samples, {gt_enc_imp.n_variants} variants')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fef1699",
   "metadata": {},
   "source": [
    "## 2. Training with multi-GPUs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79db71c0",
   "metadata": {},
   "source": [
    "### 2.1 Pre-training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d02a9d8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /mnt/qmtang/EvoFill/\n",
    "nohup env OMP_NUM_THREADS=4 CUDA_VISIBLE_DEVICES=1 \\\n",
    "  accelerate launch --config_file ds_zero3.yaml \\\n",
    "  stage1_training_ds.py \\\n",
    "  > logs/pre_HLA_251206.log 2>&1 &\n",
    "\n",
    "nohup env OMP_NUM_THREADS=8 \\\n",
    "  accelerate launch --config_file ds_zero3.yaml \\\n",
    "  stage1_training_ds.py \\\n",
    "  > logs/pre_chr22_251211-2.log 2>&1 &\n",
    "\n",
    "tail -f logs/pre_chr22_251211-2.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc271a60",
   "metadata": {},
   "source": [
    "### 2.2 Augmentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "816c3a6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /mnt/qmtang/EvoFill/\n",
    "nohup env OMP_NUM_THREADS=4 accelerate launch --config_file ds_zero3.yaml stage2_augmentation_ds.py > logs/aug_chr22_2512011.log 2>&1 &"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9ea5d12",
   "metadata": {},
   "source": [
    "### 2.3 Merge weight files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8bc9ff6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd /mnt/qmtang/EvoFill_data/20251204_chr22/models/checkpoint-stage1/\n",
    "python zero_to_fp32.py . ./"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b718f797",
   "metadata": {},
   "source": [
    "## 3. Imputation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4034fdcb",
   "metadata": {},
   "source": [
    "### 3.1 Load the trained model\n",
    "\n",
    "Choose a path where including `<work_dir>/model` and have trained model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df4c2e39",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Work Dir: /mnt/qmtang/EvoFill_data/20251211_chr22\n",
      "[INF] Model[hg38_chr22] loaded.\n",
      "[INF] Total params: 39,503,723\n"
     ]
    }
   ],
   "source": [
    "work_dir = Path('/mnt/qmtang/EvoFill_data/20251211_chr22')\n",
    "os.chdir(work_dir)\n",
    "print(f\"Work Dir: {work_dir}\")\n",
    "\n",
    "# ---- 1. 加载模型 ----\n",
    "device = 'cuda' if torch.cuda.is_available() else 'cpu'\n",
    "\n",
    "json_path = f\"{work_dir}/models/model_meta.json\"\n",
    "meta = json.load(open(json_path))\n",
    "model = EvoFill(\n",
    "    n_alleles=int(meta[\"alleles\"]),\n",
    "    total_sites=int(meta[\"total_sites\"]),\n",
    "    chunk_size=int(meta[\"chunk_size\"]),\n",
    "    chunk_overlap=int(meta[\"overlap\"]),\n",
    "    d_model=int(meta[\"d_model\"]),\n",
    "    d_state=int(meta[\"d_state\"]),\n",
    "    headdim=int(meta[\"headdim\"]),\n",
    "    bimamba_layers=int(meta[\"bimamba_layers\"]),\n",
    "    stack_mamba_layers=int(meta[\"stack_mamba_layers\"])\n",
    ").to(device)\n",
    "\n",
    "\n",
    "# state_dict = torch.load(f\"{work_dir}/models/pytorch_model_stage1.bin\", map_location=\"cpu\")\n",
    "state_dict = torch.load(f\"{work_dir}/models/pytorch_model_stage2.bin\", map_location=\"cpu\")\n",
    "\n",
    "model.load_state_dict(state_dict)\n",
    "total_params = sum(p.numel() for p in model.parameters())\n",
    "model.eval()\n",
    "print(f'[INF] Model[{meta[\"model_name\"]}] loaded.')\n",
    "print(f\"[INF] Total params: {total_params:,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "809730fd",
   "metadata": {},
   "source": [
    "### 3.2 Inferring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "34d5223c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INFER] 93 samples, 190184 variants\n",
      "[ImputationDataset] 93 samples, missing rate = 89.93%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Imputing: 100%|██████████| 93/93 [00:51<00:00,  1.82it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INF] 概率矩阵已保存 → /mnt/qmtang/EvoFill_data/20251211_chr22/impute_out/impute_prob.npy with shape = (93, 190184, 3) \n"
     ]
    }
   ],
   "source": [
    "gt_enc_imp = GenotypeEncoder.loadfromdisk(Path(work_dir / \"impute_in\"))\n",
    "\n",
    "print(f'[INFER] {gt_enc_imp.n_samples} samples, {gt_enc_imp.n_variants} variants')\n",
    "\n",
    "# ---- 2. 构建推理 Dataset / Loader ----\n",
    "imp_dataset = ImputationDataset(\n",
    "    x_gts_sparse=gt_enc_imp.X_gt,\n",
    "    seq_depth=gt_enc_imp.seq_depth,\n",
    "    indices=None                 # 可传入指定样本索引\n",
    ")\n",
    "imp_dataset.print_missing_stat()          # 查看原始缺失比例\n",
    "\n",
    "def collate_fn(batch):\n",
    "    x_onehot = torch.stack([item[0] for item in batch])\n",
    "    real_idx_list = [item[1] for item in batch]\n",
    "    return x_onehot, real_idx_list   # 无 y\n",
    "\n",
    "imp_loader = torch.utils.data.DataLoader(\n",
    "    imp_dataset,\n",
    "    batch_size=1,\n",
    "    shuffle=False,\n",
    "    num_workers=4,\n",
    "    pin_memory=True,\n",
    "    collate_fn=collate_fn\n",
    ")\n",
    "\n",
    "y_prob = []\n",
    "y_mask = []\n",
    "with torch.no_grad():\n",
    "    for x_onehot, real_idx in tqdm(imp_loader, desc='Imputing'):\n",
    "        x_onehot = x_onehot.to(device)\n",
    "        _, prob, _ = model(x_onehot)\n",
    "        miss_mask = x_onehot[..., -1].bool()\n",
    "        y_prob.append(prob)\n",
    "        y_mask.append(miss_mask)\n",
    "y_prob = torch.cat(y_prob, dim=0).cpu().numpy()\n",
    "y_mask = torch.cat(y_mask, dim=0).cpu().numpy()\n",
    "# 4. 保存\n",
    "out_dir = os.path.join(work_dir, 'impute_out')\n",
    "os.makedirs(out_dir, exist_ok=True)\n",
    "np.save(os.path.join(out_dir, 'impute_prob.npy'), y_prob)\n",
    "np.save(os.path.join(out_dir, 'impute_mask.npy'), y_mask)\n",
    "print(f'[INF] 概率矩阵已保存 → {out_dir}/impute_prob.npy '\n",
    "      f'with shape = {y_prob.shape} ')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da27d710",
   "metadata": {},
   "source": [
    "### 3.4 Evaluating the imputation results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b435b8ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     MAF_bin Counts val_Acc val_INFO val_IQS val_MaCH   val_Acc_CI95  val_INFO_CI95   val_IQS_CI95  val_MaCH_CI95\n",
      "(0.00, 0.05) 109870   0.994    0.145   0.142    0.752 [0.994, 0.994] [0.143, 0.147] [0.140, 0.144] [0.749, 0.754]\n",
      "(0.05, 0.10)  15765   0.952    0.644   0.656    0.733 [0.951, 0.953] [0.638, 0.650] [0.650, 0.663] [0.726, 0.739]\n",
      "(0.10, 0.20)  22521   0.934    0.717   0.725    0.806 [0.932, 0.935] [0.713, 0.722] [0.720, 0.730] [0.801, 0.810]\n",
      "(0.20, 0.30)  14896   0.915    0.751   0.749    0.841 [0.913, 0.917] [0.746, 0.757] [0.743, 0.754] [0.836, 0.846]\n",
      "(0.30, 0.40)  13774   0.911    0.772   0.764    0.857 [0.909, 0.914] [0.766, 0.777] [0.759, 0.770] [0.852, 0.863]\n",
      "(0.40, 0.50)  12984   0.910    0.777   0.761    0.857 [0.907, 0.912] [0.772, 0.783] [0.756, 0.767] [0.852, 0.863]\n"
     ]
    }
   ],
   "source": [
    "gt_enc_true = GenotypeEncoder.loadfromdisk(Path(work_dir / \"impute_out\"))\n",
    "y_true = gt_enc_true.X_gt.toarray()\n",
    "maf, bin_cnt = precompute_maf(y_true)\n",
    "y_true_oh = np.eye(gt_enc_true.seq_depth - 1)[y_true]\n",
    "bins_metrics   = metrics_by_maf_with95ci(y_prob, y_true_oh, hap_map = gt_enc_true.hap_map, maf_vec = maf, mask=y_mask)\n",
    "print_maf_stat_df_with95ci(bin_cnt,{'val': bins_metrics})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39543bcb",
   "metadata": {},
   "source": [
    "测试集：90% 少数族裔样本（CDX），90% masked 位点\n",
    "\n",
    "训练集：hg38 chr22 190184 variants\n",
    "\n",
    "**= 1kGP ONLY =**\n",
    "\n",
    "|      MAF_bin | Counts | val_Acc | val_INFO | val_IQS | val_MaCH |\n",
    "| :----------: | :----: | ------: | -------: | ------: | -------: |\n",
    "| (0.00, 0.05) |   1546 |   0.972 |    0.060 |   0.068 |    0.166 |\n",
    "| (0.05, 0.10) |   1246 |   0.931 |    0.198 |   0.216 |    0.367 |\n",
    "| (0.10, 0.20) |   2205 |   0.875 |    0.332 |   0.353 |    0.545 |\n",
    "| (0.20, 0.30) |   3112 |   0.798 |    0.487 |   0.495 |    0.754 |\n",
    "| (0.30, 0.40) |   5193 |   0.755 |    0.550 |   0.552 |    0.853 |\n",
    "| (0.40, 0.50) |   1565 |   0.739 |    0.548 |   0.545 |    0.845 |\n",
    "\n",
    "     MAF_bin Counts val_Acc val_INFO val_IQS val_MaCH   val_Acc_CI95  val_INFO_CI95   val_IQS_CI95  val_MaCH_CI95\n",
    "(0.00, 0.05) 109870   0.994    0.145   0.141    0.757 [0.993, 0.994] [0.143, 0.147] [0.139, 0.143] [0.755, 0.760]\n",
    "(0.05, 0.10)  15765   0.951    0.643   0.653    0.732 [0.950, 0.952] [0.636, 0.649] [0.646, 0.659] [0.725, 0.738]\n",
    "(0.10, 0.20)  22521   0.932    0.719   0.723    0.807 [0.931, 0.934] [0.714, 0.723] [0.718, 0.728] [0.802, 0.812]\n",
    "(0.20, 0.30)  14896   0.915    0.756   0.749    0.842 [0.913, 0.917] [0.751, 0.762] [0.743, 0.754] [0.837, 0.848]\n",
    "(0.30, 0.40)  13774   0.909    0.778   0.761    0.858 [0.907, 0.912] [0.772, 0.783] [0.755, 0.767] [0.852, 0.863]\n",
    "(0.40, 0.50)  12984   0.909    0.782   0.760    0.856 [0.906, 0.912] [0.777, 0.788] [0.754, 0.766] [0.851, 0.862]\n",
    "\n",
    "     MAF_bin Counts val_Acc val_INFO val_IQS val_MaCH   val_Acc_CI95  val_INFO_CI95   val_IQS_CI95  val_MaCH_CI95\n",
    "(0.00, 0.05)  83276   0.976    0.400   0.419    0.557 [0.976, 0.976] [0.397, 0.403] [0.416, 0.422] [0.554, 0.560]\n",
    "(0.05, 0.10)  29657   0.954    0.640   0.656    0.779 [0.953, 0.954] [0.636, 0.644] [0.651, 0.660] [0.775, 0.784]\n",
    "(0.10, 0.20)  27650   0.931    0.711   0.719    0.843 [0.930, 0.932] [0.707, 0.715] [0.715, 0.723] [0.839, 0.847]\n",
    "(0.20, 0.30)  19387   0.916    0.757   0.755    0.874 [0.914, 0.917] [0.753, 0.762] [0.750, 0.759] [0.870, 0.878]\n",
    "(0.30, 0.40)  15843   0.909    0.769   0.761    0.877 [0.907, 0.911] [0.765, 0.774] [0.756, 0.766] [0.872, 0.882]\n",
    "(0.40, 0.50)  14047   0.897    0.767   0.751    0.879 [0.895, 0.900] [0.762, 0.772] [0.746, 0.756] [0.874, 0.884]\n",
    "\n",
    "**= aDNA -> 1kGP =** \n",
    "\n",
    "|      MAF_bin | Counts | val_Acc | val_INFO | val_IQS | val_MaCH |\n",
    "| :----------: | :----: | ------: | -------: | ------: | -------: |\n",
    "| (0.00, 0.05) |   1546 |   0.972 |    0.089 |   0.085 |    0.193 |\n",
    "| (0.05, 0.10) |   1246 |   0.935 |    0.280 |   0.283 |    0.463 |\n",
    "| (0.10, 0.20) |   2205 |   0.890 |    0.452 |   0.456 |    0.685 |\n",
    "| (0.20, 0.30) |   3112 |   0.824 |    0.592 |   0.574 |    0.850 |\n",
    "| (0.30, 0.40) |   5193 |   0.787 |    0.645 |   0.614 |    0.906 |\n",
    "| (0.40, 0.50) |   1565 |   0.770 |    0.634 |   0.599 |    0.894 |\n",
    "\n",
    "**= 1kGP -> aDNA =** \n",
    "\n",
    "|      MAF_bin | Counts | val_Acc | val_INFO | val_IQS | val_MaCH |\n",
    "| :----------: | :----: | ------: | -------: | ------: | -------: |\n",
    "| (0.00, 0.05) |   1546 |   0.938 |    0.109 |   0.227 |    0.734 |\n",
    "| (0.05, 0.10) |   1246 |   0.893 |    0.174 |   0.396 |    0.633 |\n",
    "| (0.10, 0.20) |   2205 |   0.810 |    0.196 |   0.435 |    0.545 |\n",
    "| (0.20, 0.30) |   3112 |   0.690 |    0.144 |   0.433 |    0.499 |\n",
    "| (0.30, 0.40) |   5193 |   0.640 |    0.101 |   0.429 |    0.503 |\n",
    "| (0.40, 0.50) |   1565 |   0.604 |    0.102 |   0.398 |    0.464 |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de94e3af",
   "metadata": {},
   "source": [
    "### 3.5 Saving to .vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a3c07302",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[INF] 缺失位点填充完成 → /mnt/qmtang/EvoFill_data/20251107_ver4/impute_out/imputed.vcf.gz\n"
     ]
    }
   ],
   "source": [
    "from cyvcf2 import VCF, Writer\n",
    "\n",
    "# 0. 路径\n",
    "ref_vcf = \"/mnt/qmtang/EvoFill_data/20251107_ver4/minor_pops.masked30p.vcf.gz\"\n",
    "out_vcf = os.path.join(out_dir, 'imputed.vcf.gz')\n",
    "\n",
    "n_site = gt_enc.n_variants\n",
    "n_samp = gt_enc.n_samples\n",
    "n_alleles = gt_enc.seq_depth - 1\n",
    "assert y_prob.shape == (n_samp, n_site, n_alleles)\n",
    "\n",
    "# 2. 反向映射  idx -> '0|0' / '0|1' / ...\n",
    "rev_hap_map = {v: k for k, v in gt_enc.hap_map.items()}\n",
    "\n",
    "samp2idx = {sid: i for i, sid in enumerate(gt_enc.sample_ids)}\n",
    "\n",
    "# 4. 打开参考 VCF\n",
    "invcf = VCF(ref_vcf)\n",
    "tmpl  = invcf\n",
    "tmpl.set_samples(gt_enc.sample_ids)   # 替换样本列\n",
    "\n",
    "out = Writer(out_vcf, tmpl, mode='wz')\n",
    "\n",
    "for rec_idx, rec in enumerate(invcf):\n",
    "    # 当前位点全部样本的 GT\n",
    "    gt_int_pairs = []\n",
    "    for samp_idx, sample_id in enumerate(gt_enc.sample_ids):\n",
    "        old_gt = rec.genotypes[samp_idx]          # [allele1, allele2, phased]\n",
    "        if old_gt[0] == -1 or old_gt[1] == -1:    # 缺失\n",
    "            prob_vec = y_prob[samp_idx, rec_idx, :].ravel()\n",
    "            best_idx = int(prob_vec.argmax())\n",
    "            gt_str   = rev_hap_map[best_idx]\n",
    "            alleles  = list(map(int, gt_str.split('|')))\n",
    "            phased   = old_gt[2] if old_gt[2] != -1 else 1\n",
    "            gt_int_pairs.append([alleles[0], alleles[1], phased])\n",
    "        else:                                       # 非缺失，保持原样\n",
    "            gt_int_pairs.append(old_gt)\n",
    "\n",
    "    # 转成 int8 二维数组  (n_sample, 3)  last dim = [a1,a2,phased]\n",
    "    gt_array = np.array(gt_int_pairs, dtype=np.int8)\n",
    "    rec.set_format('GT', gt_array)\n",
    "    out.write_record(rec)\n",
    "\n",
    "invcf.close()\n",
    "out.close()\n",
    "\n",
    "# 5. tabix\n",
    "os.system(f'tabix -fp vcf {out_vcf}')\n",
    "print(f'[INF] 缺失位点填充完成 → {out_vcf}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mamba",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
